---
title: "count"
output: html_document
---

#0. Prepare the final annotated VCFs for splitting:
##Individuals:
```{R, engine='bash'}

#The final VCFs had a few triallelic sites (the biallelic filter was applied per population, so when they were subsequently combined, some sites ended up being triallelic).

cd /share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/

grep '^#' Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.vcf > Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.filtered.vcf
grep -v '^#' Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.vcf | awk '!($5 ~ /,/) { print }' >> Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.filtered.vcf

grep '^#' Alllines-Allcomb_xhap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.vcf > Alllines-Allcomb_xhap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.filtered.vcf
grep -v '^#' Alllines-Allcomb_xhap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.vcf | awk '!($5 ~ /,/) { print }' >> Alllines-Allcomb_xhap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.filtered.vcf

```

##Pools:
###Fix gen 140 pools VCF:
####README:
```{R, engine='bash'}

#When extracting counts I realised that the base population gen 140 was wrong (all counts were 0). In the end the issue is that during the calling with CRISP the wrong bam was used as input ("c1" which is gen 5 from the base pop, instead of "C1" which is gen 140 from the base pop), and on top of this, that bam was corrupted and lacked information, but somehow CRISP didn't complain and processed it as 0s. #To fix it, I followed these subsequent steps:

```

####Prepare files:
```{R, engine='bash'}

#First I removed the bam files derived from the problematic one:
cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/5processed
rm autosomic_isolated/c1_gen140_bwa_iso_rmdup_mq30_rg_multirealigned_autosomic.bam
rm x_isolated/c1_gen140_bwa_iso_rmdup_mq30_rg_multirealigned_x.bam

#Next I moved all previous VCF related files to the following folders:
cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/not_masked
mkdir obsolete_wrong
mv * obsolete_wrong
/share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked
mkdir obsolete_wrong
mv * obsolete_wrong

```

####Isolate chromosomes:
```{R, engine='bash'}

#Next, isolate autosomes and the Xchr from the correct bam:
cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/5processed
qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/Scripts/samtools_view_autosomic.sh C1_bwa_iso_rmdup_mq30_rg_multirealigned.bam autosomic_isolated/C1_bwa_iso_rmdup_mq30_rg_multirealigned_autosomic.bam
qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/Scripts/samtools_view_onlyX.sh C1_bwa_iso_rmdup_mq30_rg_multirealigned.bam x_isolated/C1_bwa_iso_rmdup_mq30_rg_multirealigned_x.bam

```

####Perform the CRISP calling:
#####crisp_all_autosomic_dani.sh
```{R, engine='bash'}

#Launch CRISP with all pools from the gen 140, for autosomes:
module load crisp/1.0

cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/not_masked/
BAMLIST=$(ls -v /share/rdata/ramon.pouso/POOLS/gen140/gen140/5processed/autosomic_isolated/*autosomic.bam) #make sure that only the desired BAMs are in the folder
CRISP --ref /share/rdata/ramon.pouso/reference/indexed_reference/dmel-all-chromosome-r6.14.fasta --VCF $1 \
$(for bam in $BAMLIST; do echo --bam ${bam}" "; done) -p 160 > all_autosomic.log

#Launch this as follows:
## cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/not_masked/
## qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/POOLS/Scripts/crisp_all_autosomic_dani.sh crisp_all_gen140_autosomic.vcf

```

#####crisp_all_x_dani.sh
```{R, engine='bash'}

#Launch CRISP with all pools from the gen 140, for autosomes:
module load crisp/1.0

cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/not_masked/
BAMLIST=$(ls -v /share/rdata/ramon.pouso/POOLS/gen140/gen140/5processed/x_isolated/*x.bam) #make sure that only the desired BAMs are in the folder
CRISP --ref /share/rdata/ramon.pouso/reference/indexed_reference/dmel-all-chromosome-r6.14.fasta --VCF $1 \
$(for bam in $BAMLIST; do echo --bam ${bam}" "; done) -p 80 > all_x.log

#Launch this as follows:
## cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/not_masked/
## qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/POOLS/Scripts/crisp_all_x_dani.sh crisp_all_gen140_x.vcf

```

####Mask variant files:
#####Autosomic:
```{R, engine='bash'}

#Launch this as follows:
cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked
qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/Scripts/vcf_mask_bed.sh /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/not_masked/crisp_all_gen140_autosomic.vcf crisp_all_gen140_autosomic_masked /share/rdata/ramon.pouso/reference/indexed_reference/dmel-r6.14_mask.bed

```

#####Xchr:
```{R, engine='bash'}

#Launch this as follows:
cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked
qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/Scripts/vcf_mask_bed.sh /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/not_masked/crisp_all_gen140_x.vcf crisp_all_gen140_x_masked /share/rdata/ramon.pouso/reference/indexed_reference/dmel-r6.14_mask.bed

```

####Keep SNPs:
#####Autosomic:
```{R, engine='bash'}

cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked
module load vcftools/0.1.17
vcftools --vcf crisp_all_gen140_autosomic_masked.recode.vcf --remove-indels --recode --recode-INFO-all --out crisp_all_gen140_autosomic_masked.recode_snps
mv crisp_all_gen140_autosomic_masked.recode_snps.recode.vcf crisp_all_gen140_autosomic_masked.recode_snps.vcf

```

#####Xchr:
```{R, engine='bash'}

cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked
module load vcftools/0.1.17
vcftools --vcf crisp_all_gen140_x_masked.recode.vcf --remove-indels --recode --recode-INFO-all --out crisp_all_gen140_x_masked.recode_snps
mv crisp_all_gen140_x_masked.recode_snps.recode.vcf crisp_all_gen140_x_masked.recode_snps.vcf

```

####Perform the remaining steps:
```{R, engine='bash'}

#Now that I've reobtained the files that were my starting point, my entire pipeline has to be rerun on these files.

#First go to the polarisation.Rmd script, section #8 (Polarise the VCFs), subsection ##Obtain final VCFs, and run it for the gen 140 pools.

```

###Perform additional calling with all pools (gen 0 - 140) for target sites:
####README:
```{R, engine='bash'}

#Monomorphic sites in gen 140 are missing from the VCF, and from these some are ancestral but some are derived, and are missing from the counts. It's easy to determine which are which because they are already polarized in the gens0-40 VCF. However, because we are dealing with pools and not individuals, we can't simply add two copies for each of these sites and call it a day; instead we need to retrieve the actual number of reads supporting each allele, as provided by CRISP. Thus, the best solution is to perform a joint calling (including all pools from gens0-40 and gen 140) for the subset of sites missing from the gen140 VCF which are fixed for the derived allele. For the SFS or to calculate the proportion of derived/total reads, it's enough to know the number of lost (AF=0) and fixed (AF=1) sites per category.

#However, the -bed flag in CRISP is not working, which means it's not possible to launch the calling for the target sites. So instead, I'll launch it for all sites (for each chromosome separately), as detailed in the next section.

#Eventually this functionality was apparently included after I opened an issue on their github, but I didn't test it.

```

####Extract target sites:
#####Extract all sites missing in the gen 140 VCF:
```{R, engine='bash'}

#Extract list of sites missing in the VCF from gen140 which should be fixed for the derived allele instead of the ancestral one.

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/

#Extract sites missing from gen140:
bedtools subtract -a crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf -header > crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf

```

#####Extract the fixed subset:
```{R, engine='bash'}

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/

#Then extract those which were polarized in gens 0-40 (i.e. which should be fixed in gen140), for autosomes:
bedtools intersect -a /share/rdata/ramon.pouso/polarisation/estsfs/output/final_snp_set.orthologs.sum_gen0_gen20_gen30_allele_counts.extrapolated_to_100.ancestral_vs_derived_complete.confident_inconsistent.rate6.pval.bed -b crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf | awk '$1!="X"' | cut -f-3 > /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.autosomes.bed
#And generate the VCF:
bedtools intersect -a crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.autosomes.bed > crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.missing_from_gen140_fixed_in_gen140.annovar.dm6_multianno.autosomes.vcf

#Then extract those which were polarized in gens 0-40 (i.e. which should be fixed in gen140), for the Xchr:
bedtools intersect -a /share/rdata/ramon.pouso/polarisation/estsfs/output/final_snp_set.orthologs.sum_gen0_gen20_gen30_allele_counts.extrapolated_to_100.ancestral_vs_derived_complete.confident_inconsistent.rate6.pval.bed -b crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf | awk '$1=="X"' | cut -f-3 > /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.Xchr.bed
#And generate the VCF:
bedtools intersect -a crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.Xchr.bed > crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.missing_from_gen140_fixed_in_gen140.annovar.dm6_multianno.Xchr.vcf

```

#####Extract the lost subset:
```{R, engine='bash'}

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/

#Then extract the rest (i.e. those which should be lost in gen140), for autosomes:
bedtools subtract -a crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.autosomes.bed | awk '$1!="X" {printf("%s\t%s\t%s\n",$1,$2-1,$2)}' > /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.autosomes.bed
#And generate the VCF:
bedtools intersect -a crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.autosomes.bed > crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.missing_from_gen140_lost_in_gen140.annovar.dm6_multianno.autosomes.vcf

#Then extract the rest (i.e. those which should be lost in gen140), for the Xchr:
bedtools subtract -a crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.Xchr.bed | awk '$1=="X" {printf("%s\t%s\t%s\n",$1,$2-1,$2)}' > /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.Xchr.bed
#And generate the VCF:
bedtools intersect -a crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.Xchr.bed > crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.missing_from_gen140_lost_in_gen140.annovar.dm6_multianno.Xchr.vcf

```

####Count sites per category:
#####Fixed in gen140:
```{R, engine='bash'}

SUBSET="Xchr" #"autosomes" or "Xchr"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/

p=crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.missing_from_gen140_fixed_in_gen140.annovar.dm6_multianno.${SUBSET}.vcf
TOTAL_V=$(wc -l < ${p})
INTRONIC_V=$(grep ';Func.refGene=intronic;' ${p} | wc -l)
SYNONYMOUS_V=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${p} | wc -l)
FOURFOLD_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed | wc -l)
MISSENSE_V=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${p} | wc -l)
MISTOL_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed | wc -l)
MISDEL_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed | wc -l)
STOPGAIN_V=$(grep ';ExonicFunc.refGene=stopgain;' ${p} | wc -l)
STOPLOSS_V=$(grep ';ExonicFunc.refGene=stoploss;' ${p} | wc -l)
echo -e "total_V\t$TOTAL_V\nintronic_V\t$INTRONIC_V\nsynonymous_V\t$SYNONYMOUS_V\nfourfold_V\t$FOURFOLD_V\nmissense_V\t$MISSENSE_V\ntolerated_V\t$MISTOL_V\ndeleterious_V\t$MISDEL_V\nstopgain_V\t$STOPGAIN_V\nstoploss_V\t$STOPLOSS_V" > /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.${SUBSET}_summary.txt

#Combine them:
cd /share/rdata/ramon.pouso/counts/
paste sites_missing_from_gen140_fixed_in_gen140.autosomes_summary.txt sites_missing_from_gen140_fixed_in_gen140.Xchr_summary.txt | awk '{printf ("%s\t%s\n",$1,$2+$4)}' > sites_missing_from_gen140_fixed_in_gen140.all_summary.txt

```

#####Lost in gen140:
```{R, engine='bash'}

SUBSET="autosomes" #"autosomes" or "Xchr"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/

p=crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.missing_from_gen140_lost_in_gen140.annovar.dm6_multianno.${SUBSET}.vcf
TOTAL_V=$(wc -l < ${p})
INTRONIC_V=$(grep ';Func.refGene=intronic;' ${p} | wc -l)
SYNONYMOUS_V=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${p} | wc -l)
FOURFOLD_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed | wc -l)
MISSENSE_V=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${p} | wc -l)
MISTOL_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed | wc -l)
MISDEL_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed | wc -l)
STOPGAIN_V=$(grep ';ExonicFunc.refGene=stopgain;' ${p} | wc -l)
STOPLOSS_V=$(grep ';ExonicFunc.refGene=stoploss;' ${p} | wc -l)
echo -e "total_V\t$TOTAL_V\nintronic_V\t$INTRONIC_V\nsynonymous_V\t$SYNONYMOUS_V\nfourfold_V\t$FOURFOLD_V\nmissense_V\t$MISSENSE_V\ntolerated_V\t$MISTOL_V\ndeleterious_V\t$MISDEL_V\nstopgain_V\t$STOPGAIN_V\nstoploss_V\t$STOPLOSS_V" > /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.${SUBSET}_summary.txt

#Combine them:
cd /share/rdata/ramon.pouso/counts/
paste sites_missing_from_gen140_lost_in_gen140.autosomes_summary.txt sites_missing_from_gen140_lost_in_gen140.Xchr_summary.txt | awk '{printf ("%s\t%s\n",$1,$2+$4)}' > sites_missing_from_gen140_lost_in_gen140.all_summary.txt

```

####Perform the CRISP calling:
#####crisp_gen140missing_autosomes_dani.sh
```{R, engine='bash'}

#Launch CRISP with all pools from all generations, for autosomes:
module load crisp/1.0

cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/not_masked/
BAMLIST=$(cat <(ls -v /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/4separate/multi_autosomic/*.bam) <(ls -v /share/rdata/ramon.pouso/POOLS/gen140/gen140/5processed/autosomic_isolated/*.bam)) #make sure that only the desired BAMs are in the folder
CRISP --ref /share/rdata/ramon.pouso/reference/indexed_reference/dmel-all-chromosome-r6.14.fasta --VCF crisp_all_gen0-140_sites_missing_from_gen140_fixed_in_gen140_autosomes.vcf \
--bed /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.autosomes.bed \
$(for bam in $BAMLIST; do echo --bam ${bam}" "; done) -p 160 > all_gen0-140_sites_missing_from_gen140_fixed_in_gen140_autosomes.log

#Launch this as follows:
## cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/
## qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/POOLS/Scripts/crisp_gen140missing_autosomes_dani.sh 

```

#####crisp_gen140missing_Xchr_dani.sh
```{R, engine='bash'}

#Launch CRISP with all pools from all generations, for Xchr:
module load crisp/1.0

cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/not_masked/
BAMLIST=$(cat <(ls -v /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/4separate/multi_x_chrom/*.bam) <(ls -v /share/rdata/ramon.pouso/POOLS/gen140/gen140/5processed/x_isolated/*.bam)) #make sure that only the desired BAMs are in the folder
CRISP --ref /share/rdata/ramon.pouso/reference/indexed_reference/dmel-all-chromosome-r6.14.fasta --VCF $1 \
--bed /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.Xchr.bed \
$(for bam in $BAMLIST; do echo --bam ${bam}" "; done) -p 80 > all_gen0-140_sites_missing_from_gen140_fixed_in_gen140_Xchr.log

#Launch this as follows:
## cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/
## qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/POOLS/Scripts/crisp_gen140missing_Xchr_dani.sh crisp_all_gen0-140_sites_missing_from_gen140_fixed_in_gen140_Xchr.vcf

```

###Perform additional calling with all pools (gen 0 - 140) for all sites:
####README:
```{R, engine='bash'}

#The -bed flag in CRISP is not working, which means it's not possible to launch the calling for the target sites. So instead, I'll launch it for all sites (for each chromosome separately).

#Eventually this functionality was apparently included after I opened an issue on their github, but I didn't test it.

```

####Generate BAMs with isolated autosomes:
#####Gens0-40: samtools_isolate_chr_gen0-40.sh
```{R, engine='bash'}

CHR=$1

module load samtools/1.4.1
BAMLIST=$(ls -v *multirealigned.bam)
mkdir -p /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/4separate/multi_${CHR}
for bam in ${BAMLIST[@]}
  do
  echo "${bam}"
  samtools view -o /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/4separate/multi_${CHR}/${bam/.bam/_${CHR}.bam} ${bam} ${CHR}
  done

#Launch this as follows:
## cd /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/3processed/
## qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/POOLS/Scripts/samtools_isolate_chr_gen0-40.sh CHR #where CHR should be replaced with 2L, 2R, 3L, 3R or 4

```

#####Gen140: samtools_isolate_chr_gen140.sh
```{R, engine='bash'}

CHR=$1

module load samtools/1.4.1
BAMLIST=$(ls -v *multirealigned.bam)
mkdir -p /share/rdata/ramon.pouso/POOLS/gen140/gen140/5processed/${CHR}_isolated
for bam in ${BAMLIST[@]}
  do
  echo "${bam}"
  samtools view -o /share/rdata/ramon.pouso/POOLS/gen140/gen140/5processed/${CHR}_isolated/${bam/.bam/_${CHR}.bam} ${bam} ${CHR}
  done

#Launch this as follows:
## cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/5processed/
## qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/POOLS/Scripts/samtools_isolate_chr_gen140.sh CHR #where CHR should be replaced with 2L, 2R, 3L, 3R or 4

```

####Perform the CRISP calling:
#####crisp_all_pools_chr_dani.sh
```{R, engine='bash'}

CHR=$1

module load crisp/1.0
cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/not_masked/
BAMLIST=$(cat <(ls -v /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/4separate/multi_${CHR}/*${CHR}.bam) <(ls -v /share/rdata/ramon.pouso/POOLS/gen140/gen140/5processed/${CHR}_isolated/*${CHR}.bam)) #make sure that only the desired BAMs are in the folder
CRISP --ref /share/rdata/ramon.pouso/reference/indexed_reference/dmel-all-chromosome-r6.14.fasta --VCF crisp_all_pools_gen0-140_${CHR}.vcf \
$(for bam in $BAMLIST; do echo --bam ${bam}" "; done) -p 160 > crisp_all_pools_gen0-140_${CHR}.log

#Launch this as follows:
## cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/not_masked/
## qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/POOLS/Scripts/crisp_all_pools_chr_dani.sh CHR #where CHR should be replaced with 2L, 2R, 3L, 3R or 4

```

#####crisp_all_pools_autosomes_dani.sh
```{R, engine='bash'}

#Launch CRISP with all pools from all generations, for autosomes:
module load crisp/1.0

cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/not_masked/
BAMLIST=$(cat <(ls -v /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/4separate/multi_autosomic/*autosomic.bam) <(ls -v /share/rdata/ramon.pouso/POOLS/gen140/gen140/5processed/autosomic_isolated/*autosomic.bam)) #make sure that only the desired BAMs are in the folder
CRISP --ref /share/rdata/ramon.pouso/reference/indexed_reference/dmel-all-chromosome-r6.14.fasta --VCF crisp_all_pools_gen0-140_autosomes.vcf \
$(for bam in $BAMLIST; do echo --bam ${bam}" "; done) -p 160 > crisp_all_pools_gen0-140_autosomes.log

#Launch this as follows:
## cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/not_masked/
## qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/POOLS/Scripts/crisp_all_pools_autosomes_dani.sh 

```

#####crisp_all_pools_Xchr_dani.sh
```{R, engine='bash'}

#Launch CRISP with all pools from all generations, for Xchr:
module load crisp/1.0

cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/not_masked/
BAMLIST=$(cat <(ls -v /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/4separate/multi_x_chrom/*x.bam) <(ls -v /share/rdata/ramon.pouso/POOLS/gen140/gen140/5processed/x_isolated/*x.bam)) #make sure that only the desired BAMs are in the folder
CRISP --ref /share/rdata/ramon.pouso/reference/indexed_reference/dmel-all-chromosome-r6.14.fasta --VCF crisp_all_pools_gen0-140_Xchr.vcf \
$(for bam in $BAMLIST; do echo --bam ${bam}" "; done) -p 80 > crisp_all_pools_gen0-140_Xchr.log

#Launch this as follows:
## cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/not_masked/
## qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/POOLS/Scripts/crisp_all_pools_Xchr_dani.sh

```

####Combine, process, polarise and annotate the VCFs:
#####Pb-000 variable sites only:
```{bash}

module load bcftools/1.9
module load gcc/7.2.0
module add gcc/7.2.0
module load vcftools/0.1.17
module load annovar/4.19

#I moved everything to the following folder:
cd /share/rdata/ramon.pouso/POOLS/all_gens_0_140/variants

#First, extract the common part of the sample names:
bcftools query -l crisp_all_pools_gen0-140_autosomes.vcf | rev | cut -d'_' -f2- | rev > sample_names.list

#Next, fix the sample names, and compress and index the VCFs:
##2L:
bcftools reheader --samples sample_names.list crisp_all_pools_gen0-140_2L.vcf | /DATA/APPS/freebayes/25.03.19/SeqLib/htslib/bgzip -c > crisp_all_pools_gen0-140_2L.vcf.gz
/DATA/APPS/freebayes/25.03.19/SeqLib/htslib/tabix -p vcf crisp_all_pools_gen0-140_2L.vcf.gz
##2R:
bcftools reheader --samples sample_names.list crisp_all_pools_gen0-140_2R.vcf | /DATA/APPS/freebayes/25.03.19/SeqLib/htslib/bgzip -c > crisp_all_pools_gen0-140_2R.vcf.gz
/DATA/APPS/freebayes/25.03.19/SeqLib/htslib/tabix -p vcf crisp_all_pools_gen0-140_2R.vcf.gz
##3L:
bcftools reheader --samples sample_names.list crisp_all_pools_gen0-140_3L.vcf | /DATA/APPS/freebayes/25.03.19/SeqLib/htslib/bgzip -c > crisp_all_pools_gen0-140_3L.vcf.gz
/DATA/APPS/freebayes/25.03.19/SeqLib/htslib/tabix -p vcf crisp_all_pools_gen0-140_3L.vcf.gz
##3R:
bcftools reheader --samples sample_names.list crisp_all_pools_gen0-140_3R.vcf | /DATA/APPS/freebayes/25.03.19/SeqLib/htslib/bgzip -c > crisp_all_pools_gen0-140_3R.vcf.gz
/DATA/APPS/freebayes/25.03.19/SeqLib/htslib/tabix -p vcf crisp_all_pools_gen0-140_3R.vcf.gz
##X:
bcftools reheader --samples sample_names.list crisp_all_pools_gen0-140_Xchr.vcf | /DATA/APPS/freebayes/25.03.19/SeqLib/htslib/bgzip -c > crisp_all_pools_gen0-140_Xchr.vcf.gz
/DATA/APPS/freebayes/25.03.19/SeqLib/htslib/tabix -p vcf crisp_all_pools_gen0-140_Xchr.vcf.gz

#Next, concatenate all files and keep only the previously selected orthologous high confidence SNPs.
bcftools concat crisp_all_pools_gen0-140_2L.vcf.gz crisp_all_pools_gen0-140_2R.vcf.gz crisp_all_pools_gen0-140_3L.vcf.gz crisp_all_pools_gen0-140_3R.vcf.gz crisp_all_pools_gen0-140_Xchr.vcf.gz | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/sites_variable_in_Pb_gen0.bed -header > crisp_all_pools_gen0-140.Pb_0_variable_set.orthologs.confident.vcf

#Keep only biallelic SNPs:
vcftools --vcf crisp_all_pools_gen0-140.Pb_0_variable_set.orthologs.confident.vcf --remove-indels --recode --recode-INFO-all --out crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.vcf
grep '^#' crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.vcf.recode.vcf > crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.vcf
grep -v '^#' crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.vcf.recode.vcf | awk '!($5 ~ /,/) { print }' >> crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.vcf

#And edit some problematic spaces in the header (otherwise vcftools won't be able to deal with them later on):
sed -i 's/Description=" >/Description=">/g' crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.vcf

#Fill the VCF with the ancestral alleles:
/share/rdata/ramon.pouso/Scripts/fill-aa.sh crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.vcf

#Polarize the VCF:
FILE=crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.aafilled.vcf
N_IND=$(bcftools query -l $FILE | wc -l)
ID_FIRST=$(bcftools query -l $FILE | head -n1)
N_FIRST=$(grep -v '##' $FILE | grep '#' | tr "\t" "\n" | grep -n $ID_FIRST | cut -d':' -f1)
grep -v '#' $FILE | awk -F"\t|AA=" '$4==$9' > ${FILE/.vcf/.consistent.vcf}
grep -v '#' $FILE | awk -F"\t|AA=" '$5==$9' | awk -F";AF=|;EMstats=" '{printf ("%s;AF=%s;EMstats=%s\n", $1,1-$2,$3)}' > ${FILE/.vcf/.inconsistent.vcf}
cut -f-$((N_FIRST-1)) ${FILE/.vcf/.inconsistent.vcf} | awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2,$3,$5,$4,$6,$7,$8,$9,$10)}' > ${FILE/.vcf/.inconsistent.polarized.vcf}
for col in $(seq $N_FIRST $((N_FIRST+N_IND-1)))
  do
  echo "processing individual in column number" $col
  paste ${FILE/.vcf/.inconsistent.polarized.vcf} <(cut -f$col ${FILE/.vcf/.inconsistent.vcf} | awk -F":|," '{printf ("%s:%s:%s:%s,%s:%s,%s:%s,%s\n", $1,$2,$3,$5,$4,$7,$6,$9,$8)}') > ${FILE/.vcf/.inconsistent.temp}
  mv ${FILE/.vcf/.inconsistent.temp} ${FILE/.vcf/.inconsistent.polarized.vcf}
  done
cat <(grep '#' $FILE) ${FILE/.vcf/.consistent.vcf} ${FILE/.vcf/.inconsistent.polarized.vcf} | bcftools sort > ${FILE/.aafilled.vcf/.polarized.vcf}

#Annotate the VCF, non-redundant:
##First convert from VCF to table (annovar can't deal with the particular format of the pool VCFs). This was already created in the complete version, so there is no need to run it again.
FILE=crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.vcf
convert2annovar.pl -format vcf4old --outfile ${FILE/.vcf/.annovar} $FILE
##Next, annotate the table. The resulting file is automatically labelled as .vcf by the programme, even though it really isn't a VCF.
table_annovar.pl ${FILE/.vcf/.annovar} /share/rdata/ramon.pouso/reference/indexed_reference/annovar_database/nr_ancestral_dm6 -vcfinput --outfile ${FILE/.vcf/.nr_annovar} -buildver dm6nr --protocol refGene --operation g
##Convert the annotated table to a .bed.
mv ${FILE/.vcf/.nr_annovar}.dm6nr_multianno.vcf ${FILE/.vcf/.nr_annovar}.dm6nr_multianno.tab
awk -F"\t|;ANNOVAR_DATE" '{printf ("%s\t%s\t%s\tANNOVAR_DATE%s\n"),$1,$2-1,$3,$9}' ${FILE/.vcf/.nr_annovar}.dm6nr_multianno.tab > ${FILE/.vcf/.nr_annovar}.dm6nr_multianno.bed
##Then intersect the polarized VCF with the annotated BED and edit it to obtain the annotated VCF:
grep '^#' $FILE > ${FILE/.vcf/.nr_annovar}.dm6nr_multianno.vcf
bedtools intersect -a $FILE -b ${FILE/.vcf/.nr_annovar}.dm6nr_multianno.bed -wb | awk -F"\t" '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s;%s\n"),$1,$2-1,$2,$3,$4,$5,$6,$7,$8,$NF}' | bedtools intersect -a stdin -b $FILE -wb | cut -f1,3-9,18- >> ${FILE/.vcf/.nr_annovar}.dm6nr_multianno.vcf

#Annotate the VCF, complete:
##Next, annotate the table. The resulting file is automatically labelled as .vcf by the programme, even though it really isn't a VCF.
table_annovar.pl ${FILE/.vcf/.annovar} /share/rdata/ramon.pouso/reference/indexed_reference/annovar_database/ancestral_dm6 -vcfinput --outfile ${FILE/.vcf/.annovar} -buildver dm6 --protocol refGene --operation g
##Convert the annotated table to a .bed.
mv ${FILE/.vcf/.annovar}.dm6_multianno.vcf ${FILE/.vcf/.annovar}.dm6_multianno.tab
awk -F"\t|;ANNOVAR_DATE" '{printf ("%s\t%s\t%s\tANNOVAR_DATE%s\n"),$1,$2-1,$3,$9}' ${FILE/.vcf/.annovar}.dm6_multianno.tab > ${FILE/.vcf/.annovar}.dm6_multianno.bed
##Then intersect the polarized VCF with the annotated BED and edit it to obtain the annotated VCF:
grep '^#' $FILE > ${FILE/.vcf/.annovar}.dm6_multianno.vcf
bedtools intersect -a $FILE -b ${FILE/.vcf/.annovar}.dm6_multianno.bed -wb | awk -F"\t" '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s;%s\n"),$1,$2-1,$2,$3,$4,$5,$6,$7,$8,$NF}' | bedtools intersect -a stdin -b $FILE -wb | cut -f1,3-9,18- >> ${FILE/.vcf/.annovar}.dm6_multianno.vcf

```

#####All sites:
```{bash}

module load bcftools/1.9
module load gcc/7.2.0
module add gcc/7.2.0
module load vcftools/0.1.17
module load annovar/4.19

#I moved everything to the following folder:
cd /share/rdata/ramon.pouso/POOLS/all_gens_0_140/variants

#First, extract the common part of the sample names:
bcftools query -l crisp_all_pools_gen0-140_autosomes.vcf | rev | cut -d'_' -f2- | rev > sample_names.list

#Next, fix the sample names, and compress and index the VCFs:
##2L:
bcftools reheader --samples sample_names.list crisp_all_pools_gen0-140_2L.vcf | /DATA/APPS/freebayes/25.03.19/SeqLib/htslib/bgzip -c > crisp_all_pools_gen0-140_2L.vcf.gz
/DATA/APPS/freebayes/25.03.19/SeqLib/htslib/tabix -p vcf crisp_all_pools_gen0-140_2L.vcf.gz
##2R:
bcftools reheader --samples sample_names.list crisp_all_pools_gen0-140_2R.vcf | /DATA/APPS/freebayes/25.03.19/SeqLib/htslib/bgzip -c > crisp_all_pools_gen0-140_2R.vcf.gz
/DATA/APPS/freebayes/25.03.19/SeqLib/htslib/tabix -p vcf crisp_all_pools_gen0-140_2R.vcf.gz
##3L:
bcftools reheader --samples sample_names.list crisp_all_pools_gen0-140_3L.vcf | /DATA/APPS/freebayes/25.03.19/SeqLib/htslib/bgzip -c > crisp_all_pools_gen0-140_3L.vcf.gz
/DATA/APPS/freebayes/25.03.19/SeqLib/htslib/tabix -p vcf crisp_all_pools_gen0-140_3L.vcf.gz
##3R:
bcftools reheader --samples sample_names.list crisp_all_pools_gen0-140_3R.vcf | /DATA/APPS/freebayes/25.03.19/SeqLib/htslib/bgzip -c > crisp_all_pools_gen0-140_3R.vcf.gz
/DATA/APPS/freebayes/25.03.19/SeqLib/htslib/tabix -p vcf crisp_all_pools_gen0-140_3R.vcf.gz
##X:
bcftools reheader --samples sample_names.list crisp_all_pools_gen0-140_Xchr.vcf | /DATA/APPS/freebayes/25.03.19/SeqLib/htslib/bgzip -c > crisp_all_pools_gen0-140_Xchr.vcf.gz
/DATA/APPS/freebayes/25.03.19/SeqLib/htslib/tabix -p vcf crisp_all_pools_gen0-140_Xchr.vcf.gz

#Next, concatenate all files and keep only those sites within the previously selected orthologous regions.
bcftools concat crisp_all_pools_gen0-140_2L.vcf.gz crisp_all_pools_gen0-140_2R.vcf.gz crisp_all_pools_gen0-140_3L.vcf.gz crisp_all_pools_gen0-140_3R.vcf.gz crisp_all_pools_gen0-140_Xchr.vcf.gz | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/outgroups/orthologs/dmel_dsim_dyak_orthologs.db_coord_sorted_merged.bed -header > crisp_all_pools_gen0-140.all_sites.orthologs.confident.vcf

#Keep only biallelic SNPs:
vcftools --vcf crisp_all_pools_gen0-140.all_sites.orthologs.confident.vcf --remove-indels --recode --recode-INFO-all --out crisp_all_pools_gen0-140.recode_snps.gen0-140_all_sites.orthologs.confident.vcf
grep '^#' crisp_all_pools_gen0-140.recode_snps.gen0-140_all_sites.orthologs.confident.vcf.recode.vcf > crisp_all_pools_gen0-140.recode_snps.gen0-140_all_sites.orthologs.confident.vcf
grep -v '^#' crisp_all_pools_gen0-140.recode_snps.gen0-140_all_sites.orthologs.confident.vcf.recode.vcf | awk '!($5 ~ /,/) { print }' >> crisp_all_pools_gen0-140.recode_snps.gen0-140_all_sites.orthologs.confident.vcf

#And edit some problematic spaces in the header (otherwise vcftools won't be able to deal with them later on):
sed -i 's/Description=" >/Description=">/g' crisp_all_pools_gen0-140.recode_snps.gen0-140_all_sites.orthologs.confident.vcf

#Mask variant files to remove repetitive regions:
qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/Scripts/vcf_mask_bed.sh crisp_all_pools_gen0-140.recode_snps.gen0-140_all_sites.orthologs.confident.vcf crisp_all_pools_gen0-140_masked.recode_snps.gen0-140_all_sites.orthologs.confident /share/rdata/ramon.pouso/reference/indexed_reference/dmel-r6.14_mask.bed

mv crisp_all_pools_gen0-140_masked.recode_snps.gen0-140_all_sites.orthologs.confident.recode.vcf crisp_all_pools_gen0-140_masked.recode_snps.gen0-140_all_sites.orthologs.confident.vcf


COPY THIS SECTION TO A NEW SCRIPT INSTEAD.

#First we'll retrieve all sites with depth higher than the average + 3SD for each of the following pools: gen0, gen20, and gen30 of the base population. Then we'll retrieve sites with the flag EMfail. Finally we'll filter them out from the VCF.

cd /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/
rm CRISP_autosomic/masked/crisp_multi_all_masked.recode_snps.DP_EM_filter_raw.bed
touch CRISP_autosomic/masked/crisp_multi_all_masked.recode_snps.DP_EM_filter_raw.bed
##DP Gen 0:
AVG=$(grep "mean coverageData" /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/qualitycontrol/LBT0_new_gen0_bwa_iso_rmdup_mq30_rg_multirealigned_autosomic_stats/genome_results.txt | awk -F"= |X" '{printf ("%s\n",$2)}')
SD=$(grep "std coverageData" /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/qualitycontrol/LBT0_new_gen0_bwa_iso_rmdup_mq30_rg_multirealigned_autosomic_stats/genome_results.txt | awk -F"= |X" '{printf ("%s\n",$2)}')
FILTER=$(echo $AVG + 3*$SD | bc)
grep -v '#' CRISP_autosomic/masked/crisp_multi_all_masked.recode_snps.vcf | cut -f-2,12 | awk -v filter="$FILTER" -F"\t|:" '{if ($5 >= filter) printf ("%s\t%s\t%s\n", $1,$2-1,$2)}' >> CRISP_autosomic/masked/crisp_multi_all_masked.recode_snps.DP_EM_filter_raw.bed
##DP Gen 20:
AVG=$(grep "mean coverageData" /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/qualitycontrol/BT20_new_gen20_bwa_iso_rmdup_mq30_rg_multirealigned_autosomic_stats/genome_results.txt | awk -F"= |X" '{printf ("%s\n",$2)}')
SD=$(grep "std coverageData" /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/qualitycontrol/BT20_new_gen20_bwa_iso_rmdup_mq30_rg_multirealigned_autosomic_stats/genome_results.txt | awk -F"= |X" '{printf ("%s\n",$2)}')
FILTER=$(echo $AVG + 3*$SD | bc)
grep -v '#' CRISP_autosomic/masked/crisp_multi_all_masked.recode_snps.vcf | cut -f-2,13 | awk -v filter="$FILTER" -F"\t|:" '{if ($5 >= filter) printf ("%s\t%s\t%s\n", $1,$2-1,$2)}' >> CRISP_autosomic/masked/crisp_multi_all_masked.recode_snps.DP_EM_filter_raw.bed
##DP Gen 30:
AVG=$(grep "mean coverageData" /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/qualitycontrol/BT30_new_gen30_bwa_iso_rmdup_mq30_rg_multirealigned_autosomic_stats/genome_results.txt | awk -F"= |X" '{printf ("%s\n",$2)}')
SD=$(grep "std coverageData" /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/qualitycontrol/BT30_new_gen30_bwa_iso_rmdup_mq30_rg_multirealigned_autosomic_stats/genome_results.txt | awk -F"= |X" '{printf ("%s\n",$2)}')
FILTER=$(echo $AVG + 3*$SD | bc)
grep -v '#' CRISP_autosomic/masked/crisp_multi_all_masked.recode_snps.vcf | cut -f-2,14 | awk -v filter="$FILTER" -F"\t|:" '{if ($5 >= filter) printf ("%s\t%s\t%s\n", $1,$2-1,$2)}' >> CRISP_autosomic/masked/crisp_multi_all_masked.recode_snps.DP_EM_filter_raw.bed
##EMfail:
grep -v '#' CRISP_autosomic/masked/crisp_multi_all_masked.recode_snps.vcf | grep "EMfail" | awk -F"\t" '{printf ("%s\t%s\t%s\n", $1,$2-1,$2)}' >> CRISP_autosomic/masked/crisp_multi_all_masked.recode_snps.DP_EM_filter_raw.bed

#Sort sites and remove duplicates:
sort -k1,1 -k2,2n CRISP_autosomic/masked/crisp_multi_all_masked.recode_snps.DP_EM_filter_raw.bed | uniq > CRISP_autosomic/masked/crisp_multi_all_masked.recode_snps.DP_EM_filter_sorted.bed

#Filter the VCF:
module load gcc/7.2.0
module add gcc/7.2.0
bedtools subtract -a CRISP_autosomic/masked/crisp_multi_all_masked.recode_snps.vcf -b CRISP_autosomic/masked/crisp_multi_all_masked.recode_snps.DP_EM_filter_sorted.bed -header > CRISP_autosomic/masked/crisp_multi_all_masked.recode_snps.DP_EM_filtered.vcf


THEN SEE DP_EM_filtered in counts (the genome_results.txt files look to be valid), then check if est-sfs needs to be rerun

#Fill the VCF with the ancestral alleles:
/share/rdata/ramon.pouso/Scripts/fill-aa.sh crisp_all_pools_gen0-140.recode_snps.gen0-140_all_sites.orthologs.confident.vcf

#Polarize the VCF:
FILE=crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.aafilled.vcf
N_IND=$(bcftools query -l $FILE | wc -l)
ID_FIRST=$(bcftools query -l $FILE | head -n1)
N_FIRST=$(grep -v '##' $FILE | grep '#' | tr "\t" "\n" | grep -n $ID_FIRST | cut -d':' -f1)
grep -v '#' $FILE | awk -F"\t|AA=" '$4==$9' > ${FILE/.vcf/.consistent.vcf}
grep -v '#' $FILE | awk -F"\t|AA=" '$5==$9' | awk -F";AF=|;EMstats=" '{printf ("%s;AF=%s;EMstats=%s\n", $1,1-$2,$3)}' > ${FILE/.vcf/.inconsistent.vcf}
cut -f-$((N_FIRST-1)) ${FILE/.vcf/.inconsistent.vcf} | awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2,$3,$5,$4,$6,$7,$8,$9,$10)}' > ${FILE/.vcf/.inconsistent.polarized.vcf}
for col in $(seq $N_FIRST $((N_FIRST+N_IND-1)))
  do
  echo "processing individual in column number" $col
  paste ${FILE/.vcf/.inconsistent.polarized.vcf} <(cut -f$col ${FILE/.vcf/.inconsistent.vcf} | awk -F":|," '{printf ("%s:%s:%s:%s,%s:%s,%s:%s,%s\n", $1,$2,$3,$5,$4,$7,$6,$9,$8)}') > ${FILE/.vcf/.inconsistent.temp}
  mv ${FILE/.vcf/.inconsistent.temp} ${FILE/.vcf/.inconsistent.polarized.vcf}
  done
cat <(grep '#' $FILE) ${FILE/.vcf/.consistent.vcf} ${FILE/.vcf/.inconsistent.polarized.vcf} | bcftools sort > ${FILE/.aafilled.vcf/.polarized.vcf}

#Annotate the VCF, non-redundant:
##First convert from VCF to table (annovar can't deal with the particular format of the pool VCFs). This was already created in the complete version, so there is no need to run it again.
FILE=crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.vcf
convert2annovar.pl -format vcf4old --outfile ${FILE/.vcf/.annovar} $FILE
##Next, annotate the table. The resulting file is automatically labelled as .vcf by the programme, even though it really isn't a VCF.
table_annovar.pl ${FILE/.vcf/.annovar} /share/rdata/ramon.pouso/reference/indexed_reference/annovar_database/nr_ancestral_dm6 -vcfinput --outfile ${FILE/.vcf/.nr_annovar} -buildver dm6nr --protocol refGene --operation g
##Convert the annotated table to a .bed.
mv ${FILE/.vcf/.nr_annovar}.dm6nr_multianno.vcf ${FILE/.vcf/.nr_annovar}.dm6nr_multianno.tab
awk -F"\t|;ANNOVAR_DATE" '{printf ("%s\t%s\t%s\tANNOVAR_DATE%s\n"),$1,$2-1,$3,$9}' ${FILE/.vcf/.nr_annovar}.dm6nr_multianno.tab > ${FILE/.vcf/.nr_annovar}.dm6nr_multianno.bed
##Then intersect the polarized VCF with the annotated BED and edit it to obtain the annotated VCF:
grep '^#' $FILE > ${FILE/.vcf/.nr_annovar}.dm6nr_multianno.vcf
bedtools intersect -a $FILE -b ${FILE/.vcf/.nr_annovar}.dm6nr_multianno.bed -wb | awk -F"\t" '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s;%s\n"),$1,$2-1,$2,$3,$4,$5,$6,$7,$8,$NF}' | bedtools intersect -a stdin -b $FILE -wb | cut -f1,3-9,18- >> ${FILE/.vcf/.nr_annovar}.dm6nr_multianno.vcf

#Annotate the VCF, complete:
##Next, annotate the table. The resulting file is automatically labelled as .vcf by the programme, even though it really isn't a VCF.
table_annovar.pl ${FILE/.vcf/.annovar} /share/rdata/ramon.pouso/reference/indexed_reference/annovar_database/ancestral_dm6 -vcfinput --outfile ${FILE/.vcf/.annovar} -buildver dm6 --protocol refGene --operation g
##Convert the annotated table to a .bed.
mv ${FILE/.vcf/.annovar}.dm6_multianno.vcf ${FILE/.vcf/.annovar}.dm6_multianno.tab
awk -F"\t|;ANNOVAR_DATE" '{printf ("%s\t%s\t%s\tANNOVAR_DATE%s\n"),$1,$2-1,$3,$9}' ${FILE/.vcf/.annovar}.dm6_multianno.tab > ${FILE/.vcf/.annovar}.dm6_multianno.bed
##Then intersect the polarized VCF with the annotated BED and edit it to obtain the annotated VCF:
grep '^#' $FILE > ${FILE/.vcf/.annovar}.dm6_multianno.vcf
bedtools intersect -a $FILE -b ${FILE/.vcf/.annovar}.dm6_multianno.bed -wb | awk -F"\t" '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s;%s\n"),$1,$2-1,$2,$3,$4,$5,$6,$7,$8,$NF}' | bedtools intersect -a stdin -b $FILE -wb | cut -f1,3-9,18- >> ${FILE/.vcf/.annovar}.dm6_multianno.vcf


```


###Fix headers (obsolete):
```{R, engine='bash'}

#The final VCFs had a few triallelic sites (the biallelic filter was applied per population, so when they were subsequently combined, some sites ended up being triallelic).

cd /share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/

grep '^#' Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.vcf > Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.filtered.vcf
grep -v '^#' Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.vcf | awk '!($5 ~ /,/) { print }' >> Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.filtered.vcf


cd /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/
awk '{gsub("##contig=<ID=2L>","##contig=<ID=2L,length=23513712>");gsub("##contig=<ID=2R>","##contig=<ID=2R,length=25286936>");gsub("##contig=<ID=3L>","##contig=<ID=3L,length=28110227>");gsub("##contig=<ID=3R>","##contig=<ID=3R,length=32079331>");gsub("##contig=<ID=X>","##contig=<ID=X,length=23542271>");print}' crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf > crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.good_headers.vcf

```

#1. Split the VCFs at the individual/pool level:
##Original set of sites:
###Individuals:
####split_individuals.sh
```{R, engine='bash'}

cd /share/rdata/ramon.pouso/counts
module load java/jre/1.8.0_73
module load GATK/4.1.4
module load bcftools/1.9

VCF=$1
if [ $VCF == "/share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.filtered.vcf" ]
  then
  SUBSET="autosomes"
elif [ $VCF == "/share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_xhap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.filtered.vcf" ]
  then
  SUBSET="Xchr"
fi
echo $SUBSET

mkdir -p individuals_counts/$SUBSET/
POPULATIONS=$(bcftools query -l "${VCF}" | rev | cut -d'-' -f2- | rev | sort | uniq)
POP=$(echo "$POPULATIONS" | sed -n ${SGE_TASK_ID}p)
INDIVIDUALS=$(bcftools query -l "${VCF}" | awk -v pop="$POP" '$1 ~ pop {print}' | sort | uniq)
for i in ${INDIVIDUALS[@]}
  do
  echo "${i}"
  ID=$(echo "${i}")
  java -XX:MaxMetaspaceSize=1g -XX:+UseG1GC -XX:+UseStringDeduplication -Xmx4g -jar /share/apps/GATK/4.1.4/gatk-package-4.1.4.0-local.jar SelectVariants \
  -R /share/rdata/ramon.pouso/reference/indexed_reference/ancestral_dmel-all-chromosome-r6.14.fa \
  -V "${VCF}" \
  -O individuals_counts/$SUBSET/"${i}"_"${SUBSET}"_individual.vcf \
  --exclude-non-variants \
  -sn $ID
  done

```

####Send the array job:
```{R, engine='bash'}

mkdir -p /share/rdata/ramon.pouso/counts/individuals_counts/
cd /share/rdata/ramon.pouso/counts/individuals_counts/

module load bcftools/1.9
VCF=/share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_xhap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.filtered.vcf
 #/share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.filtered.vcf #/share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_xhap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.filtered.vcf
 
N_POPS=$(bcftools query -l "${VCF}" | rev | cut -d'-' -f2- | rev | sort | uniq | wc -l)
qsub -cwd -l h=compute-0-9 -t 1-$N_POPS /share/rdata/ramon.pouso/counts/Scripts/split_individuals.sh $VCF

```

###Pools:
####Exclude non-variants:
#####split_pools.sh
```{R, engine='bash'}

cd /share/rdata/ramon.pouso/counts
module load java/jre/1.8.0_73
module load GATK/4.1.4
module load bcftools/1.9

VCF=$1
if [ $VCF == "/share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf" ]
  then
  SUBSET="gen0-40"
elif [ $VCF == "/share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf" ]
  then
  SUBSET="gen140"
fi
echo $SUBSET

mkdir -p pools_counts/$SUBSET/
POOLS=$(bcftools query -l "${VCF}")
for p in ${POOLS[@]}
  do
  echo "${p}"
  ID=$(echo "${p}")
  POOL_NAME=$(echo "${p}" | cut -d'_' -f1 | sort | uniq)
  FIELD_NUMBER=$(grep "^#" $VCF | grep -v '##' | tr "\t" "\n" | grep -nw $p | cut -d':' -f1)
  paste <(grep -v '^#' $VCF | cut -f1-8) <(grep -v '^#' $VCF | cut -f$FIELD_NUMBER | awk -F":|," '{printf ("%s\t%s\n", $4+$6+$8,$5+$7+$9)}') | awk '$10 > 0' > pools_counts/$SUBSET/"${POOL_NAME}"_"${SUBSET}"_pool.txt
  done

```

#####Send the array job:
```{R, engine='bash'}

mkdir -p /share/rdata/ramon.pouso/counts/pools_counts/
cd /share/rdata/ramon.pouso/counts/pools_counts/

VCF=/share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf #/share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf #/share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf
 
qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/counts/Scripts/split_pools.sh $VCF

```

####Keep non-variants:
#####split_pools_keep_nv.sh
```{R, engine='bash'}

cd /share/rdata/ramon.pouso/counts
module load java/jre/1.8.0_73
module load GATK/4.1.4
module load bcftools/1.9

VCF=$1
if [ $VCF == "/share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf" ]
  then
  SUBSET="gen0-40"
elif [ $VCF == "/share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf" ]
  then
  SUBSET="gen140"
fi
echo $SUBSET

mkdir -p pools_counts/$SUBSET/
POOLS=$(bcftools query -l "${VCF}")
for p in ${POOLS[@]}
  do
  echo "${p}"
  ID=$(echo "${p}")
  POOL_NAME=$(echo "${p}" | cut -d'_' -f1 | sort | uniq)
  FIELD_NUMBER=$(grep "^#" $VCF | grep -v '##' | tr "\t" "\n" | grep -nw $p | cut -d':' -f1)
  paste <(grep -v '^#' $VCF | cut -f1-8) <(grep -v '^#' $VCF | cut -f$FIELD_NUMBER | awk -F":|," '{printf ("%s\t%s\n", $4+$6+$8,$5+$7+$9)}') > pools_counts/$SUBSET/"${POOL_NAME}"_"${SUBSET}"_pool.nv.txt
  done

```

#####Send the array job:
```{R, engine='bash'}

mkdir -p /share/rdata/ramon.pouso/counts/pools_counts/
cd /share/rdata/ramon.pouso/counts/pools_counts/

VCF=/share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf #/share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf #/share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf
 
qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/counts/Scripts/split_pools_keep_nv.sh $VCF

```

#####Filter sites with 10 or more reads:
```{R, engine='bash'}

SUBSET="gen0-40" #SUBSET should be replaced with either "gen0-40" or "gen140"
cd /share/rdata/ramon.pouso/counts/pools_counts/$SUBSET/
POOL_LIST=($(ls -rt `find . -name '*_'${SUBSET}'_pool.nv.txt' -print`))
for pool in "${POOL_LIST[@]}"
  do
  echo $pool
  awk '($9+$10) >= 10' ${pool} > ${pool/.nv.txt/_d10.nv.txt}
  done

```

##Only sites with variation in Pb-000:
###All sites:
####Extract list of sites:
```{bash}

cd /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned

grep -v '^#' crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf | cut -f1-2,4-5,12 | awk -F"\t|:|," '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$8+$10+$12,$9+$11+$13)}' | awk '($6>0) && ($7>0)' > /share/rdata/ramon.pouso/counts/sites_variable_in_Pb_gen0.bed

```

####Filter VCFs:
```{bash}

module load gcc/7.2.0 
module add gcc/7.2.0

#Pools, gen0-40:
bedtools intersect -a /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_variable_in_Pb_gen0.bed -header > /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf
#Pools, gen140:
bedtools intersect -a /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_variable_in_Pb_gen0.bed -header > /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf
#Individuals, autosomes:
bedtools intersect -a /share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.filtered.vcf -b /share/rdata/ramon.pouso/counts/sites_variable_in_Pb_gen0.bed -header > /share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.Pb_0_variable_set.polarized.annovar.dm6_multianno.filtered.vcf
#Individuals, Xchr:
bedtools intersect -a /share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_xhap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.filtered.vcf -b /share/rdata/ramon.pouso/counts/sites_variable_in_Pb_gen0.bed -header > /share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_xhap_genotyped_snps_masked_hf_pass.Pb_0_variable_set.polarized.annovar.dm6_multianno.filtered.vcf

```

####Split pools:
#####Keep non-variants:
######split_pools_keep_nv.sh
```{R, engine='bash'}

cd /share/rdata/ramon.pouso/counts
module load bcftools/1.9

VCF=$1
if [ $VCF == "/share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf" ]
  then
  SUBSET="gen0-40"
elif [ $VCF == "/share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf" ]
  then
  SUBSET="gen140"
fi
echo $SUBSET

mkdir -p pools_counts/$SUBSET/
POOLS=$(bcftools query -l "${VCF}")
for p in ${POOLS[@]}
  do
  echo "${p}"
  ID=$(echo "${p}")
  POOL_NAME=$(echo "${p}" | cut -d'_' -f1 | sort | uniq)
  FIELD_NUMBER=$(grep "^#" $VCF | grep -v '##' | tr "\t" "\n" | grep -nw $p | cut -d':' -f1)
  paste <(grep -v '^#' $VCF | cut -f1-8) <(grep -v '^#' $VCF | cut -f$FIELD_NUMBER | awk -F":|," '{printf ("%s\t%s\n", $4+$6+$8,$5+$7+$9)}') > pools_counts/$SUBSET/"${POOL_NAME}"_"${SUBSET}"_pool.nv.txt
  done

```

######Send the array job:
```{R, engine='bash'}

mkdir -p /share/rdata/ramon.pouso/counts/pools_counts/
cd /share/rdata/ramon.pouso/counts/pools_counts/

VCF=/share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf #/share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf #/share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf
 
qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/counts/Scripts/split_pools_keep_nv.sh $VCF

```

######Filter sites with 10 or more reads:
```{R, engine='bash'}

SUBSET="gen140" #SUBSET should be replaced with either "gen0-40" or "gen140"
cd /share/rdata/ramon.pouso/counts/pools_counts/$SUBSET/
POOL_LIST=($(ls -rt `find . -name '*_'${SUBSET}'_pool.nv.txt' -print`))
for pool in "${POOL_LIST[@]}"
  do
  echo $pool
  awk '($9+$10) >= 10' ${pool} > ${pool/.nv.txt/_d10.nv.txt}
  done

```

###Subset in gen0-140 VCF:
####Extract list of sites:
```{bash}

cd /share/rdata/ramon.pouso/POOLS/all_gens_0_140/variants/

grep -v '^#' crisp_all_pools_gen0-140.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf | cut -f1-2,4-5,12 | awk -F"\t|:|," '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$8+$10+$12,$9+$11+$13)}' | awk '($6>0) && ($7>0)' > /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_in_VCF0-140_variable_in_Pb_gen0.bed

```

####Filter separate VCFs:
#####Pools:
```{bash}

module load gcc/7.2.0 
module add gcc/7.2.0

#Pools, gen0-40:
bedtools intersect -a /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_in_VCF0-140_variable_in_Pb_gen0.bed -header > /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/crisp_multi_all_masked.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf
#Pools, gen140:
bedtools intersect -a /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.final_snp_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_in_VCF0-140_variable_in_Pb_gen0.bed -header > /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf

```

#####Individuals:
```{bash}

module load gcc/7.2.0 
module add gcc/7.2.0

#Individuals, autosomes:
bedtools intersect -a /share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.filtered.vcf -b /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_in_VCF0-140_variable_in_Pb_gen0.bed -header > /share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.gen0-140_Pb_0_variable_set.polarized.annovar.dm6_multianno.filtered.vcf
#Individuals, Xchr:
bedtools intersect -a /share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_xhap_genotyped_snps_masked_hf_pass.final_snp_set.polarized.annovar.dm6_multianno.filtered.vcf -b /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_in_VCF0-140_variable_in_Pb_gen0.bed -header > /share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_xhap_genotyped_snps_masked_hf_pass.gen0-140_Pb_0_variable_set.polarized.annovar.dm6_multianno.filtered.vcf

```

####Split pools:
#####Exclude non-variants:
######Split the gen0-140 VCF:
```{R, engine='bash'}

module load bcftools/1.9
mkdir -p /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/

cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/

VCF="/share/rdata/ramon.pouso/POOLS/all_gens_0_140/variants/crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf"
SUBSET="gen0-140"
echo $SUBSET

POOLS=$(bcftools query -l "${VCF}")
for p in ${POOLS[@]}
  do
  echo "${p}"
  ID=$(echo "${p}")
  POOL_NAME=$(echo "${p}")
  SAMPLE=$(echo "${p}" | cut -d'_' -f1 | sort | uniq)
  FIELD_NUMBER=$(grep "^#" $VCF | grep -v '##' | tr "\t" "\n" | grep -nw $p | cut -d':' -f1)
  if [[ "$SAMPLE" == "sample"* ]]
    then
    GEN=$(echo "${p}" | tr "_" "\n" | grep "gen")
    paste <(grep -v '^#' $VCF | cut -f1-8) <(grep -v '^#' $VCF | cut -f$FIELD_NUMBER | awk -F":|," '{printf ("%s\t%s\n", $4+$6+$8,$5+$7+$9)}') | awk '$10 > 0' > pools_counts/$SUBSET/"${SAMPLE}"_"${GEN}"_"${SUBSET}"_pool.txt
    else
    paste <(grep -v '^#' $VCF | cut -f1-8) <(grep -v '^#' $VCF | cut -f$FIELD_NUMBER | awk -F":|," '{printf ("%s\t%s\n", $4+$6+$8,$5+$7+$9)}') | awk '$10 > 0' > pools_counts/$SUBSET/"${SAMPLE}"_"${SUBSET}"_pool.txt
  fi
  done

mv LBT0_gen0-140_pool.txt LBT0_gen0_gen0-140_pool.txt
mv c1_gen0-140_pool.txt c1_gen5_gen0-140_pool.txt
mv BT20_gen0-140_pool.txt BT20_gen20_gen0-140_pool.txt
mv BT30_gen0-140_pool.txt BT30_gen30_gen0-140_pool.txt
mv c2_gen0-140_pool.txt c2_gen40_gen0-140_pool.txt
mv C1_gen0-140_pool.txt C1_gen140_gen0-140_pool.txt

```

######Low recombination sites, gen0-140 VCF:
```{R, engine='bash'}

module load gcc/7.2.0 
module add gcc/7.2.0

cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/gen0-140/

POOLS=$(ls *_pool.txt)
for p in ${POOLS[@]}
  do
  echo "${p}"
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/POOLS/all_gens_0_140/variants/crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.nr_annovar.dm6nr_multianno.low_recombination.bed > ${p/.txt/.low_recombination.txt}
  done

```

######High recombination sites, gen0-140 VCF:
```{R, engine='bash'}

module load gcc/7.2.0 
module add gcc/7.2.0

cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/gen0-140/

POOLS=$(ls *_pool.txt)
for p in ${POOLS[@]}
  do
  echo "${p}"
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/POOLS/all_gens_0_140/variants/crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.nr_annovar.dm6nr_multianno.high_recombination.bed > ${p/.txt/.high_recombination.txt}
  done

```

######Split the separate VCFs:
```{R, engine='bash'}

VCF="/share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf" #/share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf #/share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf

module load bcftools/1.9
mkdir -p /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/

cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/

if [ $VCF == "/share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf" ]
  then
  SUBSET="gen0-40"
elif [ $VCF == "/share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf" ]
  then
  SUBSET="gen140"
fi
echo $SUBSET

POOLS=$(bcftools query -l "${VCF}")
for p in ${POOLS[@]}
  do
  echo "${p}"
  ID=$(echo "${p}")
  POOL_NAME=$(echo "${p}")
  SAMPLE=$(echo "${p}" | cut -d'_' -f1 | sort | uniq)
  FIELD_NUMBER=$(grep "^#" $VCF | grep -v '##' | tr "\t" "\n" | grep -nw $p | cut -d':' -f1)
  if [[ "$SAMPLE" == "sample"* ]] && [[ "$SUBSET" == "gen0-40" ]]
    then
    GEN="gen40"
    paste <(grep -v '^#' $VCF | cut -f1-8) <(grep -v '^#' $VCF | cut -f$FIELD_NUMBER | awk -F":|," '{printf ("%s\t%s\n", $4+$6+$8,$5+$7+$9)}') | awk '$10 > 0' > pools_counts/$SUBSET/"${SAMPLE}"_"${GEN}"_"${SUBSET}"_pool.txt
  elif [[ "$SAMPLE" == "sample"* ]] && [[ "$SUBSET" == "gen140" ]]
    then
    GEN="gen140"
    paste <(grep -v '^#' $VCF | cut -f1-8) <(grep -v '^#' $VCF | cut -f$FIELD_NUMBER | awk -F":|," '{printf ("%s\t%s\n", $4+$6+$8,$5+$7+$9)}') | awk '$10 > 0' > pools_counts/$SUBSET/"${SAMPLE}"_"${GEN}"_"${SUBSET}"_pool.txt
    else
    paste <(grep -v '^#' $VCF | cut -f1-8) <(grep -v '^#' $VCF | cut -f$FIELD_NUMBER | awk -F":|," '{printf ("%s\t%s\n", $4+$6+$8,$5+$7+$9)}') | awk '$10 > 0' > pools_counts/$SUBSET/"${SAMPLE}"_"${SUBSET}"_pool.txt
  fi
  done

cd pools_counts/$SUBSET/
mv LBT0_${SUBSET}_pool.txt LBT0_gen0_${SUBSET}_pool.txt
mv c1_${SUBSET}_pool.txt c1_gen5_${SUBSET}_pool.txt
mv BT20_${SUBSET}_pool.txt BT20_gen20_${SUBSET}_pool.txt
mv BT30_${SUBSET}_pool.txt BT30_gen30_${SUBSET}_pool.txt
mv c2_${SUBSET}_pool.txt c2_gen40_${SUBSET}_pool.txt
mv C1_${SUBSET}_pool.txt C1_gen140_${SUBSET}_pool.txt

```

#####Keep non-variants:
######Split the gen0-140 VCF:
```{R, engine='bash'}

module load bcftools/1.9
mkdir -p /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/

cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/

VCF="/share/rdata/ramon.pouso/POOLS/all_gens_0_140/variants/crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf"
SUBSET="gen0-140"
echo $SUBSET

POOLS=$(bcftools query -l "${VCF}")
for p in ${POOLS[@]}
  do
  echo "${p}"
  ID=$(echo "${p}")
  POOL_NAME=$(echo "${p}")
  SAMPLE=$(echo "${p}" | cut -d'_' -f1 | sort | uniq)
  FIELD_NUMBER=$(grep "^#" $VCF | grep -v '##' | tr "\t" "\n" | grep -nw $p | cut -d':' -f1)
  if [[ "$SAMPLE" == "sample"* ]]
    then
    GEN=$(echo "${p}" | tr "_" "\n" | grep "gen")
    paste <(grep -v '^#' $VCF | cut -f1-8) <(grep -v '^#' $VCF | cut -f$FIELD_NUMBER | awk -F":|," '{printf ("%s\t%s\n", $4+$6+$8,$5+$7+$9)}') > pools_counts/$SUBSET/"${SAMPLE}"_"${GEN}"_"${SUBSET}"_pool.nv.txt
    else
    paste <(grep -v '^#' $VCF | cut -f1-8) <(grep -v '^#' $VCF | cut -f$FIELD_NUMBER | awk -F":|," '{printf ("%s\t%s\n", $4+$6+$8,$5+$7+$9)}') > pools_counts/$SUBSET/"${SAMPLE}"_"${SUBSET}"_pool.nv.txt
  fi
  done

mv LBT0_gen0-140_pool.nv.txt LBT0_gen0_gen0-140_pool.nv.txt
mv c1_gen0-140_pool.nv.txt c1_gen5_gen0-140_pool.nv.txt
mv BT20_gen0-140_pool.nv.txt BT20_gen20_gen0-140_pool.nv.txt
mv BT30_gen0-140_pool.nv.txt BT30_gen30_gen0-140_pool.nv.txt
mv c2_gen0-140_pool.nv.txt c2_gen40_gen0-140_pool.nv.txt
mv C1_gen0-140_pool.nv.txt C1_gen140_gen0-140_pool.nv.txt

```

######Filter sites with 10 or more reads:
```{R, engine='bash'}

SUBSET="gen0-140" #SUBSET should be replaced with either "gen0-40" or "gen140" or "gen0-140"
cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/$SUBSET/
POOL_LIST=($(ls -rt `find . -name '*_'${SUBSET}'_pool.nv.txt' -print`))
for pool in "${POOL_LIST[@]}"
  do
  echo $pool
  awk '($9+$10) >= 10' ${pool} > ${pool/.nv.txt/_d10.nv.txt}
  done

```

####Split individuals:
#####split_individuals.sh
```{R, engine='bash'}

#Save it as /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/Scripts/split_individuals.sh and run it following the instructions in the subsequent chunk.

cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/

module load java/jre/1.8.0_73
module load GATK/4.1.4
module load bcftools/1.9

VCF=$1
if [ $VCF == "/share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.gen0-140_Pb_0_variable_set.polarized.annovar.dm6_multianno.filtered.vcf" ]
  then
  SUBSET="autosomes"
elif [ $VCF == "/share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_xhap_genotyped_snps_masked_hf_pass.gen0-140_Pb_0_variable_set.polarized.annovar.dm6_multianno.filtered.vcf" ]
  then
  SUBSET="Xchr"
fi
echo $SUBSET

mkdir -p /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/individuals_counts/$SUBSET/
POPULATIONS=$(bcftools query -l "${VCF}" | rev | cut -d'-' -f2- | rev | sort | uniq)
POP=$(echo "$POPULATIONS" | sed -n ${SGE_TASK_ID}p)
INDIVIDUALS=$(bcftools query -l "${VCF}" | awk -v pop="$POP" '$1 ~ pop {print}' | sort -V | uniq)
for i in ${INDIVIDUALS[@]}
  do
  echo "${i}"
  ID=$(echo "${i}")
  java -XX:MaxMetaspaceSize=1g -XX:+UseG1GC -XX:+UseStringDeduplication -Xmx4g -jar /share/apps/GATK/4.1.4/gatk-package-4.1.4.0-local.jar SelectVariants \
  -R /share/rdata/ramon.pouso/reference/indexed_reference/ancestral_dmel-all-chromosome-r6.14.fa \
  -V "${VCF}" \
  -O individuals_counts/$SUBSET/"${i}"_gen140_gen0-140_individual_"${SUBSET}".vcf \
  --exclude-non-variants \
  -sn $ID
  done

```

#####Send the array job:
```{R, engine='bash'}

mkdir -p /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/individuals_counts/
cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/

module load bcftools/1.9
VCF=/share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_xhap_genotyped_snps_masked_hf_pass.gen0-140_Pb_0_variable_set.polarized.annovar.dm6_multianno.filtered.vcf
#/share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_hap_genotyped_snps_masked_hf_pass.gen0-140_Pb_0_variable_set.polarized.annovar.dm6_multianno.filtered.vcf #/share/rdata/ramon.pouso/INDIVIDUOS/gen40/gen40/6variants/Alllines-Allcomb_xhap_genotyped_snps_masked_hf_pass.gen0-140_Pb_0_variable_set.polarized.annovar.dm6_multianno.filtered.vcf
 
N_POPS=$(bcftools query -l "${VCF}" | rev | cut -d'-' -f2- | rev | sort | uniq | wc -l)
qsub -cwd -l h=compute-0-9 -t 1-$N_POPS /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/Scripts/split_individuals.sh $VCF

```

###Key categories, subset in gen0-140 VCF:
####Split pools:
#####Exclude non-variants:
######Split the gen0-140 VCF:
```{R, engine='bash'}

module load bcftools/1.9
SUBSET="gen0-140"
echo $SUBSET

mkdir -p /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/$SUBSET/key_categories
cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/$SUBSET/key_categories

VCF="/share/rdata/ramon.pouso/POOLS/all_gens_0_140/variants/crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.custom.vcf"
POOLS=$(bcftools query -l "${VCF}")
for p in ${POOLS[@]}
  do
  echo "${p}"
  #ID=$(echo "${p}")
  #POOL_NAME=$(echo "${p}")
  SAMPLE=$(echo "${p}" | cut -d'_' -f1 | sort | uniq)
  FIELD_NUMBER=$(grep "^#" $VCF | grep -v '##' | tr "\t" "\n" | grep -nw $p | cut -d':' -f1)
  if [[ "$SAMPLE" == "sample"* ]]
    then
    GEN=$(echo "${p}" | tr "_" "\n" | grep "gen")
    paste <(grep -v '^#' $VCF | awk -F"\t" '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8)}') <(grep -v '^#' $VCF | cut -f$FIELD_NUMBER | awk -F":|," '{printf ("%s\t%s\n", $4+$6+$8,$5+$7+$9)}') | awk '$11 > 0' > "${SAMPLE}"_"${GEN}"_"${SUBSET}"_pool.txt
    else
    paste <(grep -v '^#' $VCF | awk -F"\t" '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8)}') <(grep -v '^#' $VCF | cut -f$FIELD_NUMBER | awk -F":|," '{printf ("%s\t%s\n", $4+$6+$8,$5+$7+$9)}') | awk '$11 > 0' > "${SAMPLE}"_"${SUBSET}"_pool.txt
  fi
  done

mv LBT0_gen0-140_pool.txt LBT0_gen0_gen0-140_pool.txt
mv c1_gen0-140_pool.txt c1_gen5_gen0-140_pool.txt
mv BT20_gen0-140_pool.txt BT20_gen20_gen0-140_pool.txt
mv BT30_gen0-140_pool.txt BT30_gen30_gen0-140_pool.txt
mv c2_gen0-140_pool.txt c2_gen40_gen0-140_pool.txt
mv C1_gen0-140_pool.txt C1_gen140_gen0-140_pool.txt

```


#2. Retrieve counts.
##Original set of sites:
###Individuals:
####individual_counts.sh:
```{R, engine='bash'}

#Launch it as follows:
#cd /share/rdata/ramon.pouso/counts/individuals_counts/
#qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/counts/Scripts/individual_counts.sh SUBSET
##Where SUBSET should be replaced with either "autosomes" or "Xchr"

SUBSET=$1

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/individuals_counts/$SUBSET/
rm counts_individual_${SUBSET}_summary.txt
echo -e "sample\ttotal_V\ttotal_D\tintronic_V\tintronic_D\tsynonymous_V\tsynonymous_D\tfourfold_V\tfourfold_D\tmissense_V\tmissense_D\ttolerated_V\ttolerated_D\tdeleterious_V\tdeleterious_D\tstopgain_V\tstopgain_D\tstoploss_V\tstoploss_D" > counts_individual_${SUBSET}_summary.txt
INDLIST=($(ls `find . -name '*_'${SUBSET}'_individual.vcf' -print`))
for i in "${INDLIST[@]}"
  do
  #echo "${i}"
  SAMPLE=$(echo "${i}" | cut -d'/' -f2 | cut -d'_' -f1)
  echo $SAMPLE
  TOTAL_V=$(grep -v '#' ${i} | wc -l)
  TOTAL_D=$(grep -v '#' ${i} | cut -f8 | awk -F";AC=|;AF=" '{printf ("%s\n"),$2}' | paste -sd+ | bc)
  INTRONIC_V=$(grep ';Func.refGene=intronic;' ${i} | wc -l)
  INTRONIC_D=$(grep ';Func.refGene=intronic;' ${i} | cut -f8 | awk -F";AC=|;AF=" '{printf ("%s\n"),$2}' | paste -sd+ | bc)
  SYNONYMOUS_V=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${i} | wc -l)
  SYNONYMOUS_D=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${i} | cut -f8 | awk -F";AC=|;AF=" '{printf ("%s\n"),$2}' | paste -sd+ | bc)
  bedtools intersect -a ${i} -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed > ${SAMPLE}_synonymous_4fold_${SUBSET}.txt
  FOURFOLD_V=$(wc -l < ${SAMPLE}_synonymous_4fold_${SUBSET}.txt)
  FOURFOLD_D=$(cut -f8 ${SAMPLE}_synonymous_4fold_${SUBSET}.txt | awk -F";AC=|;AF=" '{printf ("%s\n"),$2}' | paste -sd+ | bc)
  MISSENSE_V=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${i} | wc -l)
  MISSENSE_D=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${i} | cut -f8 | awk -F";AC=|;AF=" '{printf ("%s\n"),$2}' | paste -sd+ | bc)
  bedtools intersect -a ${i} -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed > ${SAMPLE}_missense_tolerated_${SUBSET}.txt
  MISTOL_V=$(wc -l < ${SAMPLE}_missense_tolerated_${SUBSET}.txt)
  MISTOL_D=$(cut -f8 ${SAMPLE}_missense_tolerated_${SUBSET}.txt | awk -F";AC=|;AF=" '{printf ("%s\n"),$2}' | paste -sd+ | bc)
  bedtools intersect -a ${i} -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed > ${SAMPLE}_missense_deleterious_${SUBSET}.txt
  MISDEL_V=$(wc -l < ${SAMPLE}_missense_deleterious_${SUBSET}.txt)
  MISDEL_D=$(cut -f8 ${SAMPLE}_missense_deleterious_${SUBSET}.txt | awk -F";AC=|;AF=" '{printf ("%s\n"),$2}' | paste -sd+ | bc)
  STOPGAIN_V=$(grep ';ExonicFunc.refGene=stopgain;' ${i} | wc -l)
  if [ "$STOPGAIN_V" -eq 0 ]; then STOPGAIN_D=0; else STOPGAIN_D=$(grep ';ExonicFunc.refGene=stopgain;' ${i} | cut -f8 | awk -F";AC=|;AF=" '{printf ("%s\n"),$2}' | paste -sd+ | bc); fi
  STOPLOSS_V=$(grep ';ExonicFunc.refGene=stoploss;' ${i} | wc -l)
  if [ "$STOPLOSS_V" -eq 0 ]; then STOPLOSS_D=0; else STOPLOSS_D=$(grep ';ExonicFunc.refGene=stoploss;' ${i} | cut -f8 | awk -F";AC=|;AF=" '{printf ("%s\n"),$2}' | paste -sd+ | bc); fi
  echo -e "$SAMPLE\t$TOTAL_V\t$TOTAL_D\t$INTRONIC_V\t$INTRONIC_D\t$SYNONYMOUS_V\t$SYNONYMOUS_D\t$FOURFOLD_V\t$FOURFOLD_D\t$MISSENSE_V\t$MISSENSE_D\t$MISTOL_V\t$MISTOL_D\t$MISDEL_V\t$MISDEL_D\t$STOPGAIN_V\t$STOPGAIN_D\t$STOPLOSS_V\t$STOPLOSS_D" >> counts_individual_${SUBSET}_summary.txt
  done

```

###Pools:
####Exclude non-variants VCFs:
#####All sites:
######pool_counts_all.sh:
```{R, engine='bash'}

#Launch it as follows:
#cd /share/rdata/ramon.pouso/counts/pools_counts/
#qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/counts/Scripts/pool_counts_all.sh SUBSET
##Where SUBSET should be replaced with either "gen0-40" or "gen140"

SUBSET=$1

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/pools_counts/$SUBSET/
rm counts_pool_${SUBSET}_all_summary.txt
echo -e "sample\ttotal_V\ttotal_D\ttotal_propD\tintronic_V\tintronic_D\tintronic_propD\tsynonymous_V\tsynonymous_D\tsynonymous_propD\tfourfold_V\tfourfold_D\tfourfold_propD\tmissense_V\tmissense_D\tmissense_propD\ttolerated_V\ttolerated_D\ttolerated_propD\tdeleterious_V\tdeleterious_D\tdeleterious_propD\tstopgain_V\tstopgain_D\tstopgain_propD\tstoploss_V\tstoploss_D\tstoploss_propD" > counts_pool_${SUBSET}_all_summary.txt
POOL_LIST=($(ls -rt `find . -name '*_'${SUBSET}'_pool.txt' -print`))
for p in "${POOL_LIST[@]}"
  do
  #echo "${p}"
  SAMPLE=$(echo "${p}" | cut -d'/' -f2 | cut -d'_' -f1)
  echo $SAMPLE
  TOTAL_V=$(wc -l < ${p})
  TOTAL_A=$(cut -f9 ${p} | paste -sd+ | bc)
  TOTAL_D=$(cut -f10 ${p} | paste -sd+ | bc)
  TOTAL_PROP_D=$(echo "scale=5; $TOTAL_D/$((TOTAL_A+TOTAL_D))" | bc)
  INTRONIC_V=$(grep ';Func.refGene=intronic;' ${p} | wc -l)
  INTRONIC_A=$(grep ';Func.refGene=intronic;' ${p} | cut -f9 | paste -sd+ | bc)
  INTRONIC_D=$(grep ';Func.refGene=intronic;' ${p} | cut -f10 | paste -sd+ | bc)
  INTRONIC_PROP_D=$(echo "scale=5; $INTRONIC_D/$((INTRONIC_A+INTRONIC_D))" | bc)
  SYNONYMOUS_V=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${p} | wc -l)
  SYNONYMOUS_A=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${p} | cut -f9 | paste -sd+ | bc)
  SYNONYMOUS_D=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${p} | cut -f10 | paste -sd+ | bc)
  SYNONYMOUS_PROP_D=$(echo "scale=5; $SYNONYMOUS_D/$((SYNONYMOUS_A+SYNONYMOUS_D))" | bc)
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed > ${SAMPLE}_${SUBSET}_synonymous_4fold.txt
  FOURFOLD_V=$(wc -l < ${SAMPLE}_${SUBSET}_synonymous_4fold.txt)
  FOURFOLD_A=$(cut -f10 ${SAMPLE}_${SUBSET}_synonymous_4fold.txt | paste -sd+ | bc)
  FOURFOLD_D=$(cut -f11 ${SAMPLE}_${SUBSET}_synonymous_4fold.txt | paste -sd+ | bc)
  FOURFOLD_PROP_D=$(echo "scale=5; $FOURFOLD_D/$((FOURFOLD_A+FOURFOLD_D))" | bc)
  MISSENSE_V=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${p} | wc -l)
  MISSENSE_A=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${p} | cut -f9 | paste -sd+ | bc)
  MISSENSE_D=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${p} | cut -f10 | paste -sd+ | bc)
  MISSENSE_PROP_D=$(echo "scale=5; $MISSENSE_D/$((MISSENSE_A+MISSENSE_D))" | bc)
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed > ${SAMPLE}_${SUBSET}_missense_tolerated.txt
  MISTOL_V=$(wc -l < ${SAMPLE}_${SUBSET}_missense_tolerated.txt)
  MISTOL_A=$(cut -f10 ${SAMPLE}_${SUBSET}_missense_tolerated.txt | paste -sd+ | bc)
  MISTOL_D=$(cut -f11 ${SAMPLE}_${SUBSET}_missense_tolerated.txt | paste -sd+ | bc)
  MISTOL_PROP_D=$(echo "scale=5; $MISTOL_D/$((MISTOL_A+MISTOL_D))" | bc)
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed > ${SAMPLE}_${SUBSET}_missense_deleterious.txt
  MISDEL_V=$(wc -l < ${SAMPLE}_${SUBSET}_missense_deleterious.txt)
  MISDEL_A=$(cut -f10 ${SAMPLE}_${SUBSET}_missense_deleterious.txt | paste -sd+ | bc)
  MISDEL_D=$(cut -f11 ${SAMPLE}_${SUBSET}_missense_deleterious.txt | paste -sd+ | bc)
  MISDEL_PROP_D=$(echo "scale=5; $MISDEL_D/$((MISDEL_A+MISDEL_D))" | bc)
  STOPGAIN_V=$(grep ';ExonicFunc.refGene=stopgain;' ${p} | wc -l)
  if [ "$STOPGAIN_V" -eq 0 ]
    then 
    STOPGAIN_A=0
    STOPGAIN_D=0
    STOPGAIN_PROP_D=0
    else 
    STOPGAIN_A=$(grep ';ExonicFunc.refGene=stopgain;' ${p} | cut -f9 | paste -sd+ | bc)
    STOPGAIN_D=$(grep ';ExonicFunc.refGene=stopgain;' ${p} | cut -f10 | paste -sd+ | bc)
    STOPGAIN_PROP_D=$(echo "scale=5; $STOPGAIN_D/$((STOPGAIN_A+STOPGAIN_D))" | bc)
  fi
  STOPLOSS_V=$(grep ';ExonicFunc.refGene=stoploss;' ${p} | wc -l)
  if [ "$STOPLOSS_V" -eq 0 ]
    then 
    STOPLOSS_A=0
    STOPLOSS_D=0
    STOPLOSS_PROP_D=0
    else 
    STOPLOSS_A=$(grep ';ExonicFunc.refGene=stoploss;' ${p} | cut -f9 | paste -sd+ | bc)
    STOPLOSS_D=$(grep ';ExonicFunc.refGene=stoploss;' ${p} | cut -f10 | paste -sd+ | bc)
    STOPLOSS_PROP_D=$(echo "scale=5; $STOPLOSS_D/$((STOPLOSS_A+STOPLOSS_D))" | bc)
  fi
  echo -e "$SAMPLE\t$TOTAL_V\t$TOTAL_D\t$TOTAL_PROP_D\t$INTRONIC_V\t$INTRONIC_D\t$INTRONIC_PROP_D\t$SYNONYMOUS_V\t$SYNONYMOUS_D\t$SYNONYMOUS_PROP_D\t$FOURFOLD_V\t$FOURFOLD_D\t$FOURFOLD_PROP_D\t$MISSENSE_V\t$MISSENSE_D\t$MISSENSE_PROP_D\t$MISTOL_V\t$MISTOL_D\t$MISTOL_PROP_D\t$MISDEL_V\t$MISDEL_D\t$MISDEL_PROP_D\t$STOPGAIN_V\t$STOPGAIN_D\t$STOPGAIN_PROP_D\t$STOPLOSS_V\t$STOPLOSS_D\t$STOPLOSS_PROP_D" >> counts_pool_${SUBSET}_all_summary.txt
  done
  
```

######pool_counts_autosomes.sh:
```{R, engine='bash'}

#Launch it as follows:
#cd /share/rdata/ramon.pouso/counts/pools_counts/
#qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/counts/Scripts/pool_counts_autosomes.sh SUBSET
##Where SUBSET should be replaced with either "gen0-40" or "gen140"

SUBSET=$1

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/pools_counts/$SUBSET/
rm counts_pool_${SUBSET}_autosomes_summary.txt
echo -e "sample\ttotal_V\ttotal_D\ttotal_propD\tintronic_V\tintronic_D\tintronic_propD\tsynonymous_V\tsynonymous_D\tsynonymous_propD\tfourfold_V\tfourfold_D\tfourfold_propD\tmissense_V\tmissense_D\tmissense_propD\ttolerated_V\ttolerated_D\ttolerated_propD\tdeleterious_V\tdeleterious_D\tdeleterious_propD\tstopgain_V\tstopgain_D\tstopgain_propD\tstoploss_V\tstoploss_D\tstoploss_propD" > counts_pool_${SUBSET}_autosomes_summary.txt
POOL_LIST=($(ls -rt `find . -name '*_'${SUBSET}'_pool.txt' -print`))
for pool in "${POOL_LIST[@]}"
  do
  grep -v "^X" $pool > ${pool/.txt/_autosomes.txt}
  SAMPLE=$(echo "${pool}" | cut -d'/' -f2 | cut -d'_' -f1)
  echo $SAMPLE
  p=${pool/.txt/_autosomes.txt}
  TOTAL_V=$(wc -l < ${p})
  TOTAL_A=$(cut -f9 ${p} | paste -sd+ | bc)
  TOTAL_D=$(cut -f10 ${p} | paste -sd+ | bc)
  TOTAL_PROP_D=$(echo "scale=5; $TOTAL_D/$((TOTAL_A+TOTAL_D))" | bc)
  INTRONIC_V=$(grep ';Func.refGene=intronic;' ${p} | wc -l)
  INTRONIC_A=$(grep ';Func.refGene=intronic;' ${p} | cut -f9 | paste -sd+ | bc)
  INTRONIC_D=$(grep ';Func.refGene=intronic;' ${p} | cut -f10 | paste -sd+ | bc)
  INTRONIC_PROP_D=$(echo "scale=5; $INTRONIC_D/$((INTRONIC_A+INTRONIC_D))" | bc)
  SYNONYMOUS_V=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${p} | wc -l)
  SYNONYMOUS_A=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${p} | cut -f9 | paste -sd+ | bc)
  SYNONYMOUS_D=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${p} | cut -f10 | paste -sd+ | bc)
  SYNONYMOUS_PROP_D=$(echo "scale=5; $SYNONYMOUS_D/$((SYNONYMOUS_A+SYNONYMOUS_D))" | bc)
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed > ${SAMPLE}_${SUBSET}_synonymous_4fold_autosomes.txt
  FOURFOLD_V=$(wc -l < ${SAMPLE}_${SUBSET}_synonymous_4fold_autosomes.txt)
  FOURFOLD_A=$(cut -f10 ${SAMPLE}_${SUBSET}_synonymous_4fold_autosomes.txt | paste -sd+ | bc)
  FOURFOLD_D=$(cut -f11 ${SAMPLE}_${SUBSET}_synonymous_4fold_autosomes.txt | paste -sd+ | bc)
  FOURFOLD_PROP_D=$(echo "scale=5; $FOURFOLD_D/$((FOURFOLD_A+FOURFOLD_D))" | bc)
  MISSENSE_V=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${p} | wc -l)
  MISSENSE_A=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${p} | cut -f9 | paste -sd+ | bc)
  MISSENSE_D=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${p} | cut -f10 | paste -sd+ | bc)
  MISSENSE_PROP_D=$(echo "scale=5; $MISSENSE_D/$((MISSENSE_A+MISSENSE_D))" | bc)
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed > ${SAMPLE}_${SUBSET}_missense_tolerated_autosomes.txt
  MISTOL_V=$(wc -l < ${SAMPLE}_${SUBSET}_missense_tolerated_autosomes.txt)
  MISTOL_A=$(cut -f10 ${SAMPLE}_${SUBSET}_missense_tolerated_autosomes.txt | paste -sd+ | bc)
  MISTOL_D=$(cut -f11 ${SAMPLE}_${SUBSET}_missense_tolerated_autosomes.txt | paste -sd+ | bc)
  MISTOL_PROP_D=$(echo "scale=5; $MISTOL_D/$((MISTOL_A+MISTOL_D))" | bc)
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed > ${SAMPLE}_${SUBSET}_missense_deleterious_autosomes.txt
  MISDEL_V=$(wc -l < ${SAMPLE}_${SUBSET}_missense_deleterious_autosomes.txt)
  MISDEL_A=$(cut -f10 ${SAMPLE}_${SUBSET}_missense_deleterious_autosomes.txt | paste -sd+ | bc)
  MISDEL_D=$(cut -f11 ${SAMPLE}_${SUBSET}_missense_deleterious_autosomes.txt | paste -sd+ | bc)
  MISDEL_PROP_D=$(echo "scale=5; $MISDEL_D/$((MISDEL_A+MISDEL_D))" | bc)
  STOPGAIN_V=$(grep ';ExonicFunc.refGene=stopgain;' ${p} | wc -l)
  if [ "$STOPGAIN_V" -eq 0 ]
    then 
    STOPGAIN_A=0
    STOPGAIN_D=0
    STOPGAIN_PROP_D=0
    else 
    STOPGAIN_A=$(grep ';ExonicFunc.refGene=stopgain;' ${p} | cut -f9 | paste -sd+ | bc)
    STOPGAIN_D=$(grep ';ExonicFunc.refGene=stopgain;' ${p} | cut -f10 | paste -sd+ | bc)
    STOPGAIN_PROP_D=$(echo "scale=5; $STOPGAIN_D/$((STOPGAIN_A+STOPGAIN_D))" | bc)
  fi
  STOPLOSS_V=$(grep ';ExonicFunc.refGene=stoploss;' ${p} | wc -l)
  if [ "$STOPLOSS_V" -eq 0 ]
    then 
    STOPLOSS_A=0
    STOPLOSS_D=0
    STOPLOSS_PROP_D=0
    else 
    STOPLOSS_A=$(grep ';ExonicFunc.refGene=stoploss;' ${p} | cut -f9 | paste -sd+ | bc)
    STOPLOSS_D=$(grep ';ExonicFunc.refGene=stoploss;' ${p} | cut -f10 | paste -sd+ | bc)
    STOPLOSS_PROP_D=$(echo "scale=5; $STOPLOSS_D/$((STOPLOSS_A+STOPLOSS_D))" | bc)
  fi
  echo -e "$SAMPLE\t$TOTAL_V\t$TOTAL_D\t$TOTAL_PROP_D\t$INTRONIC_V\t$INTRONIC_D\t$INTRONIC_PROP_D\t$SYNONYMOUS_V\t$SYNONYMOUS_D\t$SYNONYMOUS_PROP_D\t$FOURFOLD_V\t$FOURFOLD_D\t$FOURFOLD_PROP_D\t$MISSENSE_V\t$MISSENSE_D\t$MISSENSE_PROP_D\t$MISTOL_V\t$MISTOL_D\t$MISTOL_PROP_D\t$MISDEL_V\t$MISDEL_D\t$MISDEL_PROP_D\t$STOPGAIN_V\t$STOPGAIN_D\t$STOPGAIN_PROP_D\t$STOPLOSS_V\t$STOPLOSS_D\t$STOPLOSS_PROP_D" >> counts_pool_${SUBSET}_autosomes_summary.txt
  done
  
```

######pool_counts_Xchr.sh:
```{R, engine='bash'}

#Launch it as follows:
#cd /share/rdata/ramon.pouso/counts/pools_counts/
#qsub -cwd -l h=compute-0-9 /share/rdata/ramon.pouso/counts/Scripts/pool_counts_Xchr.sh SUBSET
##Where SUBSET should be replaced with either "gen0-40" or "gen140"

SUBSET=$1

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/pools_counts/$SUBSET/
rm counts_pool_${SUBSET}_Xchr_summary.txt
echo -e "sample\ttotal_V\ttotal_D\ttotal_propD\tintronic_V\tintronic_D\tintronic_propD\tsynonymous_V\tsynonymous_D\tsynonymous_propD\tfourfold_V\tfourfold_D\tfourfold_propD\tmissense_V\tmissense_D\tmissense_propD\ttolerated_V\ttolerated_D\ttolerated_propD\tdeleterious_V\tdeleterious_D\tdeleterious_propD\tstopgain_V\tstopgain_D\tstopgain_propD\tstoploss_V\tstoploss_D\tstoploss_propD" > counts_pool_${SUBSET}_Xchr_summary.txt
POOL_LIST=($(ls -rt `find . -name '*_'${SUBSET}'_pool.txt' -print`))
for pool in "${POOL_LIST[@]}"
  do
  grep "^X" $pool > ${pool/.txt/_Xchr.txt}
  SAMPLE=$(echo "${pool}" | cut -d'/' -f2 | cut -d'_' -f1)
  echo $SAMPLE
  p=${pool/.txt/_Xchr.txt}
  TOTAL_V=$(wc -l < ${p})
  TOTAL_A=$(cut -f9 ${p} | paste -sd+ | bc)
  TOTAL_D=$(cut -f10 ${p} | paste -sd+ | bc)
  TOTAL_PROP_D=$(echo "scale=5; $TOTAL_D/$((TOTAL_A+TOTAL_D))" | bc)
  INTRONIC_V=$(grep ';Func.refGene=intronic;' ${p} | wc -l)
  INTRONIC_A=$(grep ';Func.refGene=intronic;' ${p} | cut -f9 | paste -sd+ | bc)
  INTRONIC_D=$(grep ';Func.refGene=intronic;' ${p} | cut -f10 | paste -sd+ | bc)
  INTRONIC_PROP_D=$(echo "scale=5; $INTRONIC_D/$((INTRONIC_A+INTRONIC_D))" | bc)
  SYNONYMOUS_V=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${p} | wc -l)
  SYNONYMOUS_A=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${p} | cut -f9 | paste -sd+ | bc)
  SYNONYMOUS_D=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${p} | cut -f10 | paste -sd+ | bc)
  SYNONYMOUS_PROP_D=$(echo "scale=5; $SYNONYMOUS_D/$((SYNONYMOUS_A+SYNONYMOUS_D))" | bc)
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed > ${SAMPLE}_${SUBSET}_synonymous_4fold_Xchr.txt
  FOURFOLD_V=$(wc -l < ${SAMPLE}_${SUBSET}_synonymous_4fold_Xchr.txt)
  FOURFOLD_A=$(cut -f10 ${SAMPLE}_${SUBSET}_synonymous_4fold_Xchr.txt | paste -sd+ | bc)
  FOURFOLD_D=$(cut -f11 ${SAMPLE}_${SUBSET}_synonymous_4fold_Xchr.txt | paste -sd+ | bc)
  FOURFOLD_PROP_D=$(echo "scale=5; $FOURFOLD_D/$((FOURFOLD_A+FOURFOLD_D))" | bc)
  MISSENSE_V=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${p} | wc -l)
  MISSENSE_A=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${p} | cut -f9 | paste -sd+ | bc)
  MISSENSE_D=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${p} | cut -f10 | paste -sd+ | bc)
  MISSENSE_PROP_D=$(echo "scale=5; $MISSENSE_D/$((MISSENSE_A+MISSENSE_D))" | bc)
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed > ${SAMPLE}_${SUBSET}_missense_tolerated_Xchr.txt
  MISTOL_V=$(wc -l < ${SAMPLE}_${SUBSET}_missense_tolerated_Xchr.txt)
  MISTOL_A=$(cut -f10 ${SAMPLE}_${SUBSET}_missense_tolerated_Xchr.txt | paste -sd+ | bc)
  MISTOL_D=$(cut -f11 ${SAMPLE}_${SUBSET}_missense_tolerated_Xchr.txt | paste -sd+ | bc)
  MISTOL_PROP_D=$(echo "scale=5; $MISTOL_D/$((MISTOL_A+MISTOL_D))" | bc)
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed > ${SAMPLE}_${SUBSET}_missense_deleterious_Xchr.txt
  MISDEL_V=$(wc -l < ${SAMPLE}_${SUBSET}_missense_deleterious_Xchr.txt)
  MISDEL_A=$(cut -f10 ${SAMPLE}_${SUBSET}_missense_deleterious_Xchr.txt | paste -sd+ | bc)
  MISDEL_D=$(cut -f11 ${SAMPLE}_${SUBSET}_missense_deleterious_Xchr.txt | paste -sd+ | bc)
  MISDEL_PROP_D=$(echo "scale=5; $MISDEL_D/$((MISDEL_A+MISDEL_D))" | bc)
  STOPGAIN_V=$(grep ';ExonicFunc.refGene=stopgain;' ${p} | wc -l)
  if [ "$STOPGAIN_V" -eq 0 ]
    then 
    STOPGAIN_A=0
    STOPGAIN_D=0
    STOPGAIN_PROP_D=0
    else 
    STOPGAIN_A=$(grep ';ExonicFunc.refGene=stopgain;' ${p} | cut -f9 | paste -sd+ | bc)
    STOPGAIN_D=$(grep ';ExonicFunc.refGene=stopgain;' ${p} | cut -f10 | paste -sd+ | bc)
    STOPGAIN_PROP_D=$(echo "scale=5; $STOPGAIN_D/$((STOPGAIN_A+STOPGAIN_D))" | bc)
  fi
  STOPLOSS_V=$(grep ';ExonicFunc.refGene=stoploss;' ${p} | wc -l)
  if [ "$STOPLOSS_V" -eq 0 ]
    then 
    STOPLOSS_A=0
    STOPLOSS_D=0
    STOPLOSS_PROP_D=0
    else 
    STOPLOSS_A=$(grep ';ExonicFunc.refGene=stoploss;' ${p} | cut -f9 | paste -sd+ | bc)
    STOPLOSS_D=$(grep ';ExonicFunc.refGene=stoploss;' ${p} | cut -f10 | paste -sd+ | bc)
    STOPLOSS_PROP_D=$(echo "scale=5; $STOPLOSS_D/$((STOPLOSS_A+STOPLOSS_D))" | bc)
  fi
  echo -e "$SAMPLE\t$TOTAL_V\t$TOTAL_D\t$TOTAL_PROP_D\t$INTRONIC_V\t$INTRONIC_D\t$INTRONIC_PROP_D\t$SYNONYMOUS_V\t$SYNONYMOUS_D\t$SYNONYMOUS_PROP_D\t$FOURFOLD_V\t$FOURFOLD_D\t$FOURFOLD_PROP_D\t$MISSENSE_V\t$MISSENSE_D\t$MISSENSE_PROP_D\t$MISTOL_V\t$MISTOL_D\t$MISTOL_PROP_D\t$MISDEL_V\t$MISDEL_D\t$MISDEL_PROP_D\t$STOPGAIN_V\t$STOPGAIN_D\t$STOPGAIN_PROP_D\t$STOPLOSS_V\t$STOPLOSS_D\t$STOPLOSS_PROP_D" >> counts_pool_${SUBSET}_Xchr_summary.txt
  done
  
```

####Keep non-variants VCFs:
#####Sites with 10 or more reads:
######Build SFS per category:
```{R, engine='bash'}

SUBSET="gen0-40" #"gen0-40" or "gen140"
CATEGORY="fourfold" #"intronic", "synonymous", "fourfold", "missense", "tolerated", "deleterious", "LoF"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/pools_counts/$SUBSET/
rm SFS_pool_${SUBSET}_${CATEGORY}_all_summary.txt
echo -e "sample\tcategory\taf_00\taf_00_01\taf_01_02\taf_02_03\taf_03_04\taf_04_05\taf_05_06\taf_06_07\taf_07_08\taf_08_09\taf_09_10\taf_10" > SFS_pool_${SUBSET}_${CATEGORY}_all_summary.txt
  #First, retrieve the number of lost and fixed sites that are missing from the gen140 VCF (if working with gen0-40, these variables will be set to 0), which are common to all individuals:
  if [ $SUBSET == "gen140" ]
    then
    if [ $CATEGORY == "LoF" ]
      then
      LOST_00=$(grep -E "stopgain_V|stoploss_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.all_summary.txt | cut -f2 | paste -sd+ | bc)
      FIXED_10=$(grep -E "stopgain_V|stoploss_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.all_summary.txt | cut -f2 | paste -sd+ | bc)
      else
      LOST_00=$(grep "${CATEGORY}_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.all_summary.txt | cut -f2)
      FIXED_10=$(grep "${CATEGORY}_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.all_summary.txt | cut -f2)
    fi
  elif [ $SUBSET == "gen0-40" ]
    then
    LOST_00=0
    FIXED_10=0
  fi
#Then apply the following loop to retrieve the SFS for each pool:
POOL_LIST=($(ls -rt `find . -name '*_'${SUBSET}'_pool_d10.nv.txt' -print`))
for pool in "${POOL_LIST[@]}"
  do
  SAMPLE=$(echo "${pool}" | cut -d'/' -f2 | cut -d'_' -f1)
  echo $SAMPLE
  #First generate the subset file for the respective category:
  if [ $CATEGORY == "intronic" ]
    then grep ';Func.refGene=intronic;' ${pool} > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "synonymous" ]
    then grep ';ExonicFunc.refGene=synonymous_SNV;' ${pool} > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "fourfold" ]
    then awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${pool} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed | awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$3,$4,$5,$6,$7,$8,$9,$10,$11)}' > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "missense" ]
    then grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${pool} > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "tolerated" ]
    then awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${pool} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed | awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$3,$4,$5,$6,$7,$8,$9,$10,$11)}' > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "deleterious" ]
    then awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${pool} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed | awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$3,$4,$5,$6,$7,$8,$9,$10,$11)}' > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "LoF" ]
    then grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${pool} > ${pool/.txt/.$CATEGORY.txt}
  fi
  #Now calculate the number of sites within each AF category. For categories with AF=0 and AF=1, add the number of sites missing from the gen140 VCF:
  AF_00=$(awk '$10=="0"' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  ALL_00=$(($AF_00 + $LOST_00))
  AF_00_01=$(awk '($10/($9+$10)>0) && ($10/($9+$10)<=0.1)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_01_02=$(awk '($10/($9+$10)>0.1) && ($10/($9+$10)<=0.2)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_02_03=$(awk '($10/($9+$10)>0.2) && ($10/($9+$10)<=0.3)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_03_04=$(awk '($10/($9+$10)>0.3) && ($10/($9+$10)<=0.4)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_04_05=$(awk '($10/($9+$10)>0.4) && ($10/($9+$10)<=0.5)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_05_06=$(awk '($10/($9+$10)>0.5) && ($10/($9+$10)<=0.6)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_06_07=$(awk '($10/($9+$10)>0.6) && ($10/($9+$10)<=0.7)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_07_08=$(awk '($10/($9+$10)>0.7) && ($10/($9+$10)<=0.8)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_08_09=$(awk '($10/($9+$10)>0.8) && ($10/($9+$10)<=0.9)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_09_10=$(awk '($10/($9+$10)>0.9) && ($10/($9+$10)<1.0)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_10=$(awk '$9=="0"' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  ALL_10=$(($AF_10 + $FIXED_10))
  #Finally, write the SFS to the output:
  echo -e "$SAMPLE\t$CATEGORY\t$ALL_00\t$AF_00_01\t$AF_01_02\t$AF_02_03\t$AF_03_04\t$AF_04_05\t$AF_05_06\t$AF_06_07\t$AF_07_08\t$AF_08_09\t$AF_09_10\t$ALL_10" >> SFS_pool_${SUBSET}_${CATEGORY}_all_summary.txt
  done

```

######Combine SFS categories:
```{bash}

#Concatenate all category-files into a single one before downloading it:
SUBSET="gen140" #"gen0-40" or "gen140"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/pools_counts/$SUBSET/
cat SFS_pool_${SUBSET}_fourfold_all_summary.txt <(tail -n+2 SFS_pool_${SUBSET}_tolerated_all_summary.txt) <(tail -n+2 SFS_pool_${SUBSET}_deleterious_all_summary.txt) <(tail -n+2 SFS_pool_${SUBSET}_LoF_all_summary.txt) > SFS_pool_${SUBSET}_combined_all_summary.txt

scp ramon.pouso@rua2.uvigo.es://share/rdata/ramon.pouso/counts/pools_counts/gen0-40/SFS_pool_gen0-40_combined_all_summary.txt /Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/
scp ramon.pouso@rua2.uvigo.es://share/rdata/ramon.pouso/counts/pools_counts/gen140/SFS_pool_gen140_combined_all_summary.txt /Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/

```

##Only sites with variation in Pb-000:
###Retrieve sites missing in gen 140:
####All missing sites:
```{R, engine='bash'}

#Extract list of sites missing in the VCF from gen140 which should be fixed for the derived allele instead of the ancestral one.

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/

#Extract sites missing from gen140:
bedtools subtract -a crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf -header > crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf

```

####Fixed subset:
```{R, engine='bash'}

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/

#Then extract those which were polarized in gens 0-40 (i.e. which should be fixed in gen140), for autosomes:
bedtools intersect -a /share/rdata/ramon.pouso/polarisation/estsfs/output/final_snp_set.orthologs.sum_gen0_gen20_gen30_allele_counts.extrapolated_to_100.ancestral_vs_derived_complete.confident_inconsistent.rate6.pval.bed -b crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf | awk '$1!="X"' | cut -f-3 > /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.autosomes.bed
#And generate the VCF:
bedtools intersect -a crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.autosomes.bed > crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.missing_from_gen140_fixed_in_gen140.annovar.dm6_multianno.autosomes.vcf #6822

#Then extract those which were polarized in gens 0-40 (i.e. which should be fixed in gen140), for the Xchr:
bedtools intersect -a /share/rdata/ramon.pouso/polarisation/estsfs/output/final_snp_set.orthologs.sum_gen0_gen20_gen30_allele_counts.extrapolated_to_100.ancestral_vs_derived_complete.confident_inconsistent.rate6.pval.bed -b crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf | awk '$1=="X"' | cut -f-3 > /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.Xchr.bed
#And generate the VCF:
bedtools intersect -a crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.Xchr.bed > crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.missing_from_gen140_fixed_in_gen140.annovar.dm6_multianno.Xchr.vcf #1367

```

####Lost subset:
```{R, engine='bash'}

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/

#Then extract the rest (i.e. those which should be lost in gen140), for autosomes:
bedtools subtract -a crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.autosomes.bed | awk '$1!="X" {printf("%s\t%s\t%s\n",$1,$2-1,$2)}' > /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.autosomes.bed
#And generate the VCF:
bedtools intersect -a crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.autosomes.bed > crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.missing_from_gen140_lost_in_gen140.annovar.dm6_multianno.autosomes.vcf #51784

#Then extract the rest (i.e. those which should be lost in gen140), for the Xchr:
bedtools subtract -a crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.Xchr.bed | awk '$1=="X" {printf("%s\t%s\t%s\n",$1,$2-1,$2)}' > /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.Xchr.bed
#And generate the VCF:
bedtools intersect -a crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.Xchr.bed > crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.missing_from_gen140_lost_in_gen140.annovar.dm6_multianno.Xchr.vcf #6243

```

####Count sites per category:
#####Fixed subset:
```{R, engine='bash'}

SUBSET="autosomes" #"autosomes" or "Xchr"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/

p=crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.missing_from_gen140_fixed_in_gen140.annovar.dm6_multianno.${SUBSET}.vcf
TOTAL_V=$(wc -l < ${p})
INTRONIC_V=$(grep ';Func.refGene=intronic;' ${p} | wc -l)
SYNONYMOUS_V=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${p} | wc -l)
FOURFOLD_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed | wc -l)
MISSENSE_V=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${p} | wc -l)
MISTOL_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed | wc -l)
MISDEL_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed | wc -l)
STOPGAIN_V=$(grep ';ExonicFunc.refGene=stopgain;' ${p} | wc -l)
STOPLOSS_V=$(grep ';ExonicFunc.refGene=stoploss;' ${p} | wc -l)
echo -e "total_V\t$TOTAL_V\nintronic_V\t$INTRONIC_V\nsynonymous_V\t$SYNONYMOUS_V\nfourfold_V\t$FOURFOLD_V\nmissense_V\t$MISSENSE_V\ntolerated_V\t$MISTOL_V\ndeleterious_V\t$MISDEL_V\nstopgain_V\t$STOPGAIN_V\nstoploss_V\t$STOPLOSS_V" > /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.${SUBSET}_summary.txt

#Combine them:
cd /share/rdata/ramon.pouso/counts/
paste sites_missing_from_gen140_fixed_in_gen140.autosomes_summary.txt sites_missing_from_gen140_fixed_in_gen140.Xchr_summary.txt | awk '{printf ("%s\t%s\n",$1,$2+$4)}' > sites_missing_from_gen140_fixed_in_gen140.all_summary.txt

```

#####Lost subset:
```{R, engine='bash'}

SUBSET="Xchr" #"autosomes" or "Xchr"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/POOLS/gen0-40/gen0-40/5variants/multirealigned/

p=crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.missing_from_gen140_lost_in_gen140.annovar.dm6_multianno.${SUBSET}.vcf
TOTAL_V=$(wc -l < ${p})
INTRONIC_V=$(grep ';Func.refGene=intronic;' ${p} | wc -l)
SYNONYMOUS_V=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${p} | wc -l)
FOURFOLD_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed | wc -l)
MISSENSE_V=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${p} | wc -l)
MISTOL_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed | wc -l)
MISDEL_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed | wc -l)
STOPGAIN_V=$(grep ';ExonicFunc.refGene=stopgain;' ${p} | wc -l)
STOPLOSS_V=$(grep ';ExonicFunc.refGene=stoploss;' ${p} | wc -l)
echo -e "total_V\t$TOTAL_V\nintronic_V\t$INTRONIC_V\nsynonymous_V\t$SYNONYMOUS_V\nfourfold_V\t$FOURFOLD_V\nmissense_V\t$MISSENSE_V\ntolerated_V\t$MISTOL_V\ndeleterious_V\t$MISDEL_V\nstopgain_V\t$STOPGAIN_V\nstoploss_V\t$STOPLOSS_V" > /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.${SUBSET}_summary.txt

#Combine them:
cd /share/rdata/ramon.pouso/counts/
paste sites_missing_from_gen140_lost_in_gen140.autosomes_summary.txt sites_missing_from_gen140_lost_in_gen140.Xchr_summary.txt | awk '{printf ("%s\t%s\n",$1,$2+$4)}' > sites_missing_from_gen140_lost_in_gen140.all_summary.txt

```

###Pools:
####Keep non-variants VCFs:
#####Sites with 10 or more reads:
######Calculate proportion of derived reads:
```{R, engine='bash'}

SUBSET="gen140" #"gen0-40" or "gen140"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/pools_counts/$SUBSET/
rm propD_pool_${SUBSET}_all_d10_summary.txt
echo -e "sample\tfourfold_V\tfourfold_propD\ttolerated_V\ttolerated_propD\tdeleterious_V\tdeleterious_propD\tLoF_V\tLoF_propD" > propD_pool_${SUBSET}_all_d10_summary.txt
POOL_LIST=($(ls -rt `find . -name '*_'${SUBSET}'_pool_d10.nv.txt' -print`))
for p in "${POOL_LIST[@]}"
  do
  #echo "${p}"
  SAMPLE=$(echo "${p}" | cut -d'/' -f2 | cut -d'_' -f1)
  echo $SAMPLE
  
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed > ${SAMPLE}_${SUBSET}_synonymous_4fold_d10.txt
  FOURFOLD_V=$(wc -l < ${SAMPLE}_${SUBSET}_synonymous_4fold_d10.txt)
  FOURFOLD_SUM=$(awk '{print $11/($10+$11)}' ${SAMPLE}_${SUBSET}_synonymous_4fold_d10.txt | paste -sd+ | bc)
  if [ $SUBSET == "gen140" ]
    then
    FOURFOLD_FIXED=$(grep "fourfold_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.all_summary.txt | cut -f2)
    FOURFOLD_LOST=$(grep "fourfold_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.all_summary.txt | cut -f2)
    FOURFOLD_V=$(cat <(echo $FOURFOLD_V) <(echo $FOURFOLD_FIXED) <(echo $FOURFOLD_LOST) | paste -sd+ | bc)
    FOURFOLD_SUM=$(cat <(echo $FOURFOLD_SUM) <(yes "1.0" | head -n $FOURFOLD_FIXED) | paste -sd+ | bc)
  fi
  FOURFOLD_PROP_D=$(echo "scale=5; $FOURFOLD_SUM/$FOURFOLD_V" | bc)

  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed > ${SAMPLE}_${SUBSET}_missense_tolerated_d10.txt
  MISTOL_V=$(wc -l < ${SAMPLE}_${SUBSET}_missense_tolerated_d10.txt)
  MISTOL_SUM=$(awk '{print $11/($10+$11)}' ${SAMPLE}_${SUBSET}_missense_tolerated_d10.txt | paste -sd+ | bc)
  if [ $SUBSET == "gen140" ]
    then
    MISTOL_FIXED=$(grep "tolerated_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.all_summary.txt | cut -f2)
    MISTOL_LOST=$(grep "tolerated_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.all_summary.txt | cut -f2)
    MISTOL_V=$(cat <(echo $MISTOL_V) <(echo $MISTOL_FIXED) <(echo $MISTOL_LOST) | paste -sd+ | bc)
    MISTOL_SUM=$(cat <(echo $MISTOL_SUM) <(yes "1.0" | head -n $MISTOL_FIXED) | paste -sd+ | bc)
  fi
  MISTOL_PROP_D=$(echo "scale=5; $MISTOL_SUM/$MISTOL_V" | bc)
  
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed > ${SAMPLE}_${SUBSET}_missense_deleterious_d10.txt
  MISDEL_V=$(wc -l < ${SAMPLE}_${SUBSET}_missense_deleterious_d10.txt)
  MISDEL_SUM=$(awk '{print $11/($10+$11)}' ${SAMPLE}_${SUBSET}_missense_deleterious_d10.txt | paste -sd+ | bc)
  if [ $SUBSET == "gen140" ]
    then
    MISDEL_FIXED=$(grep "deleterious_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.all_summary.txt | cut -f2)
    MISDEL_LOST=$(grep "deleterious_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.all_summary.txt | cut -f2)
    MISDEL_V=$(cat <(echo $MISDEL_V) <(echo $MISDEL_FIXED) <(echo $MISDEL_LOST) | paste -sd+ | bc)
    MISDEL_SUM=$(cat <(echo $MISDEL_SUM) <(yes "1.0" | head -n $MISDEL_FIXED) | paste -sd+ | bc)
  fi
  MISDEL_PROP_D=$(echo "scale=5; $MISDEL_SUM/$MISDEL_V" | bc)
  
  LOF_V=$(grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${p} | wc -l)
  LOF_SUM=$(grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${p} | awk '{print $10/($9+$10)}' | paste -sd+ | bc)
  if [ $SUBSET == "gen140" ]
    then
    LOF_FIXED=$(grep -E "stopgain_V|stoploss_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.all_summary.txt | cut -f2 | paste -sd+ | bc)
    LOF_LOST=$(grep -E "stopgain_V|stoploss_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.all_summary.txt | cut -f2 | paste -sd+ | bc)
    LOF_V=$(cat <(echo $LOF_V) <(echo $LOF_FIXED) <(echo $LOF_LOST) | paste -sd+ | bc)
    LOF_SUM=$(cat <(echo $LOF_SUM) <(yes "1.0" | head -n $LOF_FIXED) | paste -sd+ | bc)
  fi
  LOF_PROP_D=$(echo "scale=5; $LOF_SUM/$LOF_V" | bc)
  
  echo -e "$SAMPLE\t$FOURFOLD_V\t$FOURFOLD_PROP_D\t$MISTOL_V\t$MISTOL_PROP_D\t$MISDEL_V\t$MISDEL_PROP_D\t$LOF_V\t$LOF_PROP_D" >> propD_pool_${SUBSET}_all_d10_summary.txt
  done

scp ramon.pouso@rua2.uvigo.es://share/rdata/ramon.pouso/counts/pools_counts/gen0-40/propD_pool_gen0-40_all_d10_summary.txt /Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/
scp ramon.pouso@rua2.uvigo.es://share/rdata/ramon.pouso/counts/pools_counts/gen140/propD_pool_gen140_all_d10_summary.txt /Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/


```

######Build SFS per category:
```{R, engine='bash'}

SUBSET="gen140" #"gen0-40" or "gen140"
CATEGORY="LoF" #"intronic", "synonymous", "fourfold", "missense", "tolerated", "deleterious", "LoF"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/pools_counts/$SUBSET/
rm SFS_pool_${SUBSET}_${CATEGORY}_all_summary.txt
echo -e "sample\tcategory\taf_000\taf_000_005\taf_005_010\taf_010_015\taf_015_020\taf_020_025\taf_025_030\taf_030_035\taf_035_040\taf_040_045\taf_045_050\taf_050_055\taf_055_060\taf_060_065\taf_065_070\taf_070_075\taf_075_080\taf_080_085\taf_085_090\taf_090_095\taf_095_100\taf_100" > SFS_pool_${SUBSET}_${CATEGORY}_all_summary.txt
  #First, retrieve the number of lost and fixed sites that are missing from the gen140 VCF (if working with gen0-40, these variables will be set to 0), which are common to all individuals:
  if [ $SUBSET == "gen140" ]
    then
    if [ $CATEGORY == "LoF" ]
      then
      LOST_000=$(grep -E "stopgain_V|stoploss_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.all_summary.txt | cut -f2 | paste -sd+ | bc)
      FIXED_100=$(grep -E "stopgain_V|stoploss_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.all_summary.txt | cut -f2 | paste -sd+ | bc)
      else
      LOST_000=$(grep "${CATEGORY}_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_lost_in_gen140.all_summary.txt | cut -f2)
      FIXED_100=$(grep "${CATEGORY}_V" /share/rdata/ramon.pouso/counts/sites_missing_from_gen140_fixed_in_gen140.all_summary.txt | cut -f2)
    fi
  elif [ $SUBSET == "gen0-40" ]
    then
    LOST_000=0
    FIXED_100=0
  fi
#Then apply the following loop to retrieve the SFS for each pool:
POOL_LIST=($(ls -rt `find . -name '*_'${SUBSET}'_pool_d10.nv.txt' -print`))
for pool in "${POOL_LIST[@]}"
  do
  SAMPLE=$(echo "${pool}" | cut -d'/' -f2 | cut -d'_' -f1)
  echo $SAMPLE
  #First generate the subset file for the respective category:
  if [ $CATEGORY == "intronic" ]
    then grep ';Func.refGene=intronic;' ${pool} > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "synonymous" ]
    then grep ';ExonicFunc.refGene=synonymous_SNV;' ${pool} > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "fourfold" ]
    then awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${pool} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed | awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$3,$4,$5,$6,$7,$8,$9,$10,$11)}' > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "missense" ]
    then grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${pool} > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "tolerated" ]
    then awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${pool} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed | awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$3,$4,$5,$6,$7,$8,$9,$10,$11)}' > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "deleterious" ]
    then awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${pool} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed | awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$3,$4,$5,$6,$7,$8,$9,$10,$11)}' > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "LoF" ]
    then grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${pool} > ${pool/.txt/.$CATEGORY.txt}
  fi
  #Now calculate the number of sites within each AF category. For categories with AF=0 and AF=1, add the number of sites missing from the gen140 VCF:
  AF_000=$(awk '$10=="0"' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  ALL_000=$(($AF_000 + $LOST_000))
  AF_000_005=$(awk '($10/($9+$10)>0.0) && ($10/($9+$10)<=0.05)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_005_010=$(awk '($10/($9+$10)>0.05) && ($10/($9+$10)<=0.1)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_010_015=$(awk '($10/($9+$10)>0.1) && ($10/($9+$10)<=0.15)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_015_020=$(awk '($10/($9+$10)>0.15) && ($10/($9+$10)<=0.2)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_020_025=$(awk '($10/($9+$10)>0.2) && ($10/($9+$10)<=0.25)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_025_030=$(awk '($10/($9+$10)>0.25) && ($10/($9+$10)<=0.3)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_030_035=$(awk '($10/($9+$10)>0.3) && ($10/($9+$10)<=0.35)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_035_040=$(awk '($10/($9+$10)>0.35) && ($10/($9+$10)<=0.4)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_040_045=$(awk '($10/($9+$10)>0.4) && ($10/($9+$10)<=0.45)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_045_050=$(awk '($10/($9+$10)>0.45) && ($10/($9+$10)<=0.5)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_050_055=$(awk '($10/($9+$10)>0.5) && ($10/($9+$10)<=0.55)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_055_060=$(awk '($10/($9+$10)>0.55) && ($10/($9+$10)<=0.6)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_060_065=$(awk '($10/($9+$10)>0.6) && ($10/($9+$10)<=0.65)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_065_070=$(awk '($10/($9+$10)>0.65) && ($10/($9+$10)<=0.7)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_070_075=$(awk '($10/($9+$10)>0.7) && ($10/($9+$10)<=0.75)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_075_080=$(awk '($10/($9+$10)>0.75) && ($10/($9+$10)<=0.8)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_080_085=$(awk '($10/($9+$10)>0.8) && ($10/($9+$10)<=0.85)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_085_090=$(awk '($10/($9+$10)>0.85) && ($10/($9+$10)<=0.9)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_090_095=$(awk '($10/($9+$10)>0.9) && ($10/($9+$10)<=0.95)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_095_100=$(awk '($10/($9+$10)>0.95) && ($10/($9+$10)<1.0)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_100=$(awk '$9=="0"' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  ALL_100=$(($AF_100 + $FIXED_100))
  #Finally, write the SFS to the output:
  echo -e "$SAMPLE\t$CATEGORY\t$ALL_000\t$AF_000_005\t$AF_005_010\t$AF_010_015\t$AF_015_020\t$AF_020_025\t$AF_025_030\t$AF_030_035\t$AF_035_040\t$AF_040_045\t$AF_045_050\t$AF_050_055\t$AF_055_060\t$AF_060_065\t$AF_065_070\t$AF_070_075\t$AF_075_080\t$AF_080_085\t$AF_085_090\t$AF_090_095\t$AF_095_100\t$ALL_100" >> SFS_pool_${SUBSET}_${CATEGORY}_all_summary.txt
  done

```

######Combine SFS categories:
```{bash}

#Concatenate all category-files into a single one before downloading it:
SUBSET="gen0-40" #"gen0-40" or "gen140"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/pools_counts/$SUBSET/
cat SFS_pool_${SUBSET}_fourfold_all_summary.txt <(tail -n+2 SFS_pool_${SUBSET}_tolerated_all_summary.txt) <(tail -n+2 SFS_pool_${SUBSET}_deleterious_all_summary.txt) <(tail -n+2 SFS_pool_${SUBSET}_LoF_all_summary.txt) > SFS_pool_${SUBSET}_combined_all_summary.txt

scp ramon.pouso@rua2.uvigo.es://share/rdata/ramon.pouso/counts/pools_counts/gen0-40/SFS_pool_gen0-40_combined_all_summary.txt /Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/
scp ramon.pouso@rua2.uvigo.es://share/rdata/ramon.pouso/counts/pools_counts/gen140/SFS_pool_gen140_combined_all_summary.txt /Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/

```

##Only sites in gen 0-140 VCF with variation in Pb-000:
###Retrieve sites missing in gen 140:
####All missing sites:
```{R, engine='bash'}

#Extract list of sites missing in the VCF from gen140 which should be fixed for the derived allele instead of the ancestral one.

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/POOLS/all_gens_0_140/variants

#Extract sites missing from gen140:
bedtools subtract -a crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/POOLS/gen140/gen140/6variants/masked/crisp_multi_all_masked.recode_snps.Pb_0_variable_set.orthologs.confident.polarized.annovar.dm6_multianno.vcf -header > crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.gen0-140_missing_from_gen140.annovar.dm6_multianno.vcf

```

####Fixed subset:
```{R, engine='bash'}

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/POOLS/all_gens_0_140/variants

#Then extract those which were polarized in gens 0-40 (i.e. which should be fixed in gen140), for autosomes:
bedtools intersect -a /share/rdata/ramon.pouso/polarisation/estsfs/output/final_snp_set.orthologs.sum_gen0_gen20_gen30_allele_counts.extrapolated_to_100.ancestral_vs_derived_complete.confident_inconsistent.rate6.pval.bed -b crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.gen0-140_missing_from_gen140.annovar.dm6_multianno.vcf | awk '$1!="X"' | cut -f-3 > /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_fixed_in_gen140.autosomes.bed
#And generate the VCF:
bedtools intersect -a crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.gen0-140_missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_fixed_in_gen140.autosomes.bed > crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.gen0-140_missing_from_gen140_fixed_in_gen140.annovar.dm6_multianno.autosomes.vcf #6182

#Then extract those which were polarized in gens 0-40 (i.e. which should be fixed in gen140), for the Xchr:
bedtools intersect -a /share/rdata/ramon.pouso/polarisation/estsfs/output/final_snp_set.orthologs.sum_gen0_gen20_gen30_allele_counts.extrapolated_to_100.ancestral_vs_derived_complete.confident_inconsistent.rate6.pval.bed -b crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.gen0-140_missing_from_gen140.annovar.dm6_multianno.vcf | awk '$1=="X"' | cut -f-3 > /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_fixed_in_gen140.Xchr.bed
#And generate the VCF:
bedtools intersect -a crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.gen0-140_missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_fixed_in_gen140.Xchr.bed > crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.gen0-140_missing_from_gen140_fixed_in_gen140.annovar.dm6_multianno.Xchr.vcf #1323

#Generate the combined bed:
cat /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_fixed_in_gen140.autosomes.bed /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_fixed_in_gen140.Xchr.bed > /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_fixed_in_gen140.all.bed

```

####Lost subset:
```{R, engine='bash'}

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/POOLS/all_gens_0_140/variants

#Then extract the rest (i.e. those which should be lost in gen140), for autosomes:
bedtools subtract -a crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.gen0-140_missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_fixed_in_gen140.autosomes.bed | awk '$1!="X" {printf("%s\t%s\t%s\n",$1,$2-1,$2)}' > /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_lost_in_gen140.autosomes.bed
#And generate the VCF:
bedtools intersect -a crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.gen0-140_missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_lost_in_gen140.autosomes.bed > crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.gen0-140_missing_from_gen140_lost_in_gen140.annovar.dm6_multianno.autosomes.vcf #48648

#Then extract the rest (i.e. those which should be lost in gen140), for the Xchr:
bedtools subtract -a crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.gen0-140_missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_fixed_in_gen140.Xchr.bed | awk '$1=="X" {printf("%s\t%s\t%s\n",$1,$2-1,$2)}' > /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_lost_in_gen140.Xchr.bed
#And generate the VCF:
bedtools intersect -a crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.gen0-140_missing_from_gen140.annovar.dm6_multianno.vcf -b /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_lost_in_gen140.Xchr.bed > crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.gen0-140_missing_from_gen140_lost_in_gen140.annovar.dm6_multianno.Xchr.vcf #6067

#Generate the combined bed:
cat /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_lost_in_gen140.autosomes.bed /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_lost_in_gen140.Xchr.bed > /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_lost_in_gen140.all.bed

```

####Count sites per category:
#####Fixed subset:
```{R, engine='bash'}

SUBSET="Xchr" #"autosomes" or "Xchr"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/POOLS/all_gens_0_140/variants

p=crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.gen0-140_missing_from_gen140_fixed_in_gen140.annovar.dm6_multianno.${SUBSET}.vcf
TOTAL_V=$(wc -l < ${p})
INTRONIC_V=$(grep ';Func.refGene=intronic;' ${p} | wc -l)
SYNONYMOUS_V=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${p} | wc -l)
FOURFOLD_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed | wc -l)
MISSENSE_V=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${p} | wc -l)
MISTOL_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed | wc -l)
MISDEL_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed | wc -l)
STOPGAIN_V=$(grep ';ExonicFunc.refGene=stopgain;' ${p} | wc -l)
STOPLOSS_V=$(grep ';ExonicFunc.refGene=stoploss;' ${p} | wc -l)
echo -e "total_V\t$TOTAL_V\nintronic_V\t$INTRONIC_V\nsynonymous_V\t$SYNONYMOUS_V\nfourfold_V\t$FOURFOLD_V\nmissense_V\t$MISSENSE_V\ntolerated_V\t$MISTOL_V\ndeleterious_V\t$MISDEL_V\nstopgain_V\t$STOPGAIN_V\nstoploss_V\t$STOPLOSS_V" > /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_fixed_in_gen140.${SUBSET}_summary.txt

#Combine them:
cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/
paste sites_missing_from_gen140_fixed_in_gen140.autosomes_summary.txt sites_missing_from_gen140_fixed_in_gen140.Xchr_summary.txt | awk '{printf ("%s\t%s\n",$1,$2+$4)}' > sites_missing_from_gen140_fixed_in_gen140.all_summary.txt

```

#####Lost subset:
```{R, engine='bash'}

SUBSET="Xchr" #"autosomes" or "Xchr"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/POOLS/all_gens_0_140/variants

p=crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.gen0-140_missing_from_gen140_lost_in_gen140.annovar.dm6_multianno.${SUBSET}.vcf
TOTAL_V=$(wc -l < ${p})
INTRONIC_V=$(grep ';Func.refGene=intronic;' ${p} | wc -l)
SYNONYMOUS_V=$(grep ';ExonicFunc.refGene=synonymous_SNV;' ${p} | wc -l)
FOURFOLD_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed | wc -l)
MISSENSE_V=$(grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${p} | wc -l)
MISTOL_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed | wc -l)
MISDEL_V=$(awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed | wc -l)
STOPGAIN_V=$(grep ';ExonicFunc.refGene=stopgain;' ${p} | wc -l)
STOPLOSS_V=$(grep ';ExonicFunc.refGene=stoploss;' ${p} | wc -l)
echo -e "total_V\t$TOTAL_V\nintronic_V\t$INTRONIC_V\nsynonymous_V\t$SYNONYMOUS_V\nfourfold_V\t$FOURFOLD_V\nmissense_V\t$MISSENSE_V\ntolerated_V\t$MISTOL_V\ndeleterious_V\t$MISDEL_V\nstopgain_V\t$STOPGAIN_V\nstoploss_V\t$STOPLOSS_V" > /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/sites_missing_from_gen140_lost_in_gen140.${SUBSET}_summary.txt

#Combine them:
cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/
paste sites_missing_from_gen140_lost_in_gen140.autosomes_summary.txt sites_missing_from_gen140_lost_in_gen140.Xchr_summary.txt | awk '{printf ("%s\t%s\n",$1,$2+$4)}' > sites_missing_from_gen140_lost_in_gen140.all_summary.txt

```

###Pools:
####Exclude non-variants VCFs:
#####Count variants and derived copies, gen 0-140 VCF:
######All sites in the dataset:
```{R, engine='bash'}

SUBSET="gen0-140"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/$SUBSET/
rm counts_pool_${SUBSET}_all_summary.txt
echo -e "sample\tgeneration\tfourfold_V\tfourfold_D\ttolerated_V\ttolerated_D\tdeleterious_V\tdeleterious_D\tLoF_V\tLoF_D" > counts_pool_${SUBSET}_all_summary.txt
POOL_LIST=($(ls -rt `find . -name '*_'${SUBSET}'_pool.txt' -print`))
for p in "${POOL_LIST[@]}"
  do
  echo "${p}"
  SAMPLE=$(echo "${p}" | cut -d'/' -f2 | cut -d'_' -f1)
  GEN=$(echo "${p}" | cut -d'/' -f2 | cut -d'_' -f2)
  #q=/share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/${SUBSET}/${SAMPLE}_${GEN}_${SUBSET}_pool.nv.txt
  
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed > ${SAMPLE}_${GEN}_${SUBSET}_synonymous_4fold.txt
  FOURFOLD_V=$(wc -l < ${SAMPLE}_${GEN}_${SUBSET}_synonymous_4fold.txt)
  FOURFOLD_D=$(awk '{print $11}' ${SAMPLE}_${GEN}_${SUBSET}_synonymous_4fold.txt | paste -sd+ | bc)

  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed > ${SAMPLE}_${GEN}_${SUBSET}_missense_tolerated.txt
  MISTOL_V=$(wc -l < ${SAMPLE}_${GEN}_${SUBSET}_missense_tolerated.txt)
  MISTOL_D=$(awk '{print $11}' ${SAMPLE}_${GEN}_${SUBSET}_missense_tolerated.txt | paste -sd+ | bc)

  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed > ${SAMPLE}_${GEN}_${SUBSET}_missense_deleterious.txt
  MISDEL_V=$(wc -l < ${SAMPLE}_${GEN}_${SUBSET}_missense_deleterious.txt)
  MISDEL_D=$(awk '{print $11}' ${SAMPLE}_${GEN}_${SUBSET}_missense_deleterious.txt | paste -sd+ | bc)

  LOF_V=$(grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${p} | wc -l)
  LOF_D=$(grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${p} | awk '{print $10}' | paste -sd+ | bc)

  echo -e "$SAMPLE\t$GEN\t$FOURFOLD_V\t$FOURFOLD_D\t$MISTOL_V\t$MISTOL_D\t$MISDEL_V\t$MISDEL_D\t$LOF_V\t$LOF_D" >> counts_pool_${SUBSET}_all_summary.txt
  done

scp ramon.pouso@rua2.uvigo.es://share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/gen0-140/counts_pool_gen0-140_all_summary.txt /Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/

```

######Autosomes or Xchr:
```{R, engine='bash'}

SUBSET="gen0-140"
REGION="Xchr" #autosomes #Xchr

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/$SUBSET/
rm counts_pool_${SUBSET}_${REGION}_summary.txt
echo -e "sample\tgeneration\tfourfold_V\tfourfold_D\ttolerated_V\ttolerated_D\tdeleterious_V\tdeleterious_D\tLoF_V\tLoF_D" > counts_pool_${SUBSET}_${REGION}_summary.txt
POOL_LIST=($(ls -rt `find . -name '*_'${SUBSET}'_pool.txt' -print`))
for pool in "${POOL_LIST[@]}"
  do
  SAMPLE=$(echo "${pool}" | cut -d'/' -f2 | cut -d'_' -f1)
  echo $SAMPLE
  GEN=$(echo "${pool}" | cut -d'/' -f2 | cut -d'_' -f2)
  if [ $REGION == "autosomes" ]
    then grep -v "^X" $pool > ${pool/.txt/_${REGION}.txt}
  elif [ $REGION == "Xchr" ]
    then grep "^X" $pool > ${pool/.txt/_${REGION}.txt}
  fi
  p=${pool/.txt/_${REGION}.txt}

  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed > ${SAMPLE}_${GEN}_${SUBSET}_synonymous_4fold.${REGION}.txt
  FOURFOLD_V=$(wc -l < ${SAMPLE}_${GEN}_${SUBSET}_synonymous_4fold.${REGION}.txt)
  FOURFOLD_D=$(awk '{print $11}' ${SAMPLE}_${GEN}_${SUBSET}_synonymous_4fold.${REGION}.txt | paste -sd+ | bc)

  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed > ${SAMPLE}_${GEN}_${SUBSET}_missense_tolerated.${REGION}.txt
  MISTOL_V=$(wc -l < ${SAMPLE}_${GEN}_${SUBSET}_missense_tolerated.${REGION}.txt)
  MISTOL_D=$(awk '{print $11}' ${SAMPLE}_${GEN}_${SUBSET}_missense_tolerated.${REGION}.txt | paste -sd+ | bc)

  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed > ${SAMPLE}_${GEN}_${SUBSET}_missense_deleterious.${REGION}.txt
  MISDEL_V=$(wc -l < ${SAMPLE}_${GEN}_${SUBSET}_missense_deleterious.${REGION}.txt)
  MISDEL_D=$(awk '{print $11}' ${SAMPLE}_${GEN}_${SUBSET}_missense_deleterious.${REGION}.txt | paste -sd+ | bc)

  LOF_V=$(grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${p} | wc -l)
  LOF_D=$(grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${p} | awk '{print $10}' | paste -sd+ | bc)

  echo -e "$SAMPLE\t$GEN\t$FOURFOLD_V\t$FOURFOLD_D\t$MISTOL_V\t$MISTOL_D\t$MISDEL_V\t$MISDEL_D\t$LOF_V\t$LOF_D" >> counts_pool_${SUBSET}_${REGION}_summary.txt
  done

scp ramon.pouso@rua2.uvigo.es://share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/gen0-140/counts_pool_gen0-140_all_summary.txt /Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/

```

######Low recombination subset:
```{R, engine='bash'}

SUBSET="gen0-140"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/$SUBSET/
rm counts_pool_${SUBSET}_all_summary.low_recombination.txt
echo -e "sample\tgeneration\tfourfold_V\tfourfold_D\ttolerated_V\ttolerated_D\tdeleterious_V\tdeleterious_D\tLoF_V\tLoF_D" > counts_pool_${SUBSET}_all_summary.low_recombination.txt
POOL_LIST=($(ls -v `find . -name '*_'${SUBSET}'_pool.low_recombination.txt' -print`))
for p in "${POOL_LIST[@]}"
  do
  echo "${p}"
  SAMPLE=$(echo "${p}" | cut -d'/' -f2 | cut -d'_' -f1)
  GEN=$(echo "${p}" | cut -d'/' -f2 | cut -d'_' -f2)
  #q=/share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/${SUBSET}/${SAMPLE}_${GEN}_${SUBSET}_pool.nv.txt
  
  bedtools intersect -a ${p} -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed > ${SAMPLE}_${GEN}_${SUBSET}_synonymous_4fold.low_recombination.txt
  FOURFOLD_V=$(wc -l < ${SAMPLE}_${GEN}_${SUBSET}_synonymous_4fold.low_recombination.txt)
  FOURFOLD_D=$(awk '{print $11}' ${SAMPLE}_${GEN}_${SUBSET}_synonymous_4fold.low_recombination.txt | paste -sd+ | bc)

  bedtools intersect -a ${p} -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed > ${SAMPLE}_${GEN}_${SUBSET}_missense_tolerated.low_recombination.txt
  MISTOL_V=$(wc -l < ${SAMPLE}_${GEN}_${SUBSET}_missense_tolerated.low_recombination.txt)
  MISTOL_D=$(awk '{print $11}' ${SAMPLE}_${GEN}_${SUBSET}_missense_tolerated.low_recombination.txt | paste -sd+ | bc)

  bedtools intersect -a ${p} -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed > ${SAMPLE}_${GEN}_${SUBSET}_missense_deleterious.low_recombination.txt
  MISDEL_V=$(wc -l < ${SAMPLE}_${GEN}_${SUBSET}_missense_deleterious.low_recombination.txt)
  MISDEL_D=$(awk '{print $11}' ${SAMPLE}_${GEN}_${SUBSET}_missense_deleterious.low_recombination.txt | paste -sd+ | bc)

  LOF_V=$(grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${p} | wc -l)
  LOF_D=$(grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${p} | awk '{print $11}' | paste -sd+ | bc)

  echo -e "$SAMPLE\t$GEN\t$FOURFOLD_V\t$FOURFOLD_D\t$MISTOL_V\t$MISTOL_D\t$MISDEL_V\t$MISDEL_D\t$LOF_V\t$LOF_D" >> counts_pool_${SUBSET}_all_summary.low_recombination.txt
  done

scp ramon.pouso@rua2.uvigo.es://share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/gen0-140/counts_pool_gen0-140_all_summary.low_recombination.txt /Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/

```

######High recombination subset:
```{R, engine='bash'}

SUBSET="gen0-140"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/$SUBSET/
rm counts_pool_${SUBSET}_all_summary.high_recombination.txt
echo -e "sample\tgeneration\tfourfold_V\tfourfold_D\ttolerated_V\ttolerated_D\tdeleterious_V\tdeleterious_D\tLoF_V\tLoF_D" > counts_pool_${SUBSET}_all_summary.high_recombination.txt
POOL_LIST=($(ls -v `find . -name '*_'${SUBSET}'_pool.high_recombination.txt' -print`))
for p in "${POOL_LIST[@]}"
  do
  echo "${p}"
  SAMPLE=$(echo "${p}" | cut -d'/' -f2 | cut -d'_' -f1)
  GEN=$(echo "${p}" | cut -d'/' -f2 | cut -d'_' -f2)
  #q=/share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/${SUBSET}/${SAMPLE}_${GEN}_${SUBSET}_pool.nv.txt
  
  bedtools intersect -a ${p} -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed > ${SAMPLE}_${GEN}_${SUBSET}_synonymous_4fold.high_recombination.txt
  FOURFOLD_V=$(wc -l < ${SAMPLE}_${GEN}_${SUBSET}_synonymous_4fold.high_recombination.txt)
  FOURFOLD_D=$(awk '{print $11}' ${SAMPLE}_${GEN}_${SUBSET}_synonymous_4fold.high_recombination.txt | paste -sd+ | bc)

  bedtools intersect -a ${p} -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed > ${SAMPLE}_${GEN}_${SUBSET}_missense_tolerated.high_recombination.txt
  MISTOL_V=$(wc -l < ${SAMPLE}_${GEN}_${SUBSET}_missense_tolerated.high_recombination.txt)
  MISTOL_D=$(awk '{print $11}' ${SAMPLE}_${GEN}_${SUBSET}_missense_tolerated.high_recombination.txt | paste -sd+ | bc)

  bedtools intersect -a ${p} -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed > ${SAMPLE}_${GEN}_${SUBSET}_missense_deleterious.high_recombination.txt
  MISDEL_V=$(wc -l < ${SAMPLE}_${GEN}_${SUBSET}_missense_deleterious.high_recombination.txt)
  MISDEL_D=$(awk '{print $11}' ${SAMPLE}_${GEN}_${SUBSET}_missense_deleterious.high_recombination.txt | paste -sd+ | bc)

  LOF_V=$(grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${p} | wc -l)
  LOF_D=$(grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${p} | awk '{print $11}' | paste -sd+ | bc)

  echo -e "$SAMPLE\t$GEN\t$FOURFOLD_V\t$FOURFOLD_D\t$MISTOL_V\t$MISTOL_D\t$MISDEL_V\t$MISDEL_D\t$LOF_V\t$LOF_D" >> counts_pool_${SUBSET}_all_summary.high_recombination.txt
  done

scp ramon.pouso@rua2.uvigo.es://share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/gen0-140/counts_pool_gen0-140_all_summary.high_recombination.txt /Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/

```

#####Key categories, fast derived counts:
######All, autosomes or Xchr sites:
```{R, engine='bash'}

SUBSET="gen0-140"
REGION="Xchr" #all #autosomes #Xchr

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/$SUBSET/key_categories
rm counts_pool_${SUBSET}_${REGION}_summary.txt
echo -e "sample\tgeneration\tfourfold_V\tfourfold_D\ttolerated_V\ttolerated_D\tdeleterious_V\tdeleterious_D\tLoF_V\tLoF_D" > counts_pool_${SUBSET}_${REGION}_summary.txt
POOL_LIST=($(ls -v `find . -name '*_'${SUBSET}'_pool.txt' -print`))
for pool in "${POOL_LIST[@]}"
  do
  echo "${pool}"
  SAMPLE=$(echo "${pool}" | cut -d'/' -f2 | cut -d'_' -f1)
  GEN=$(echo "${pool}" | cut -d'/' -f2 | cut -d'_' -f2)
  if [ $REGION == "autosomes" ]
    then
    grep -v "^X" $pool > ${pool/.txt/_${REGION}.txt}
    p=${pool/.txt/_${REGION}.txt}
  elif [ $REGION == "Xchr" ]
    then
    grep "^X" $pool > ${pool/.txt/_${REGION}.txt}
    p=${pool/.txt/_${REGION}.txt}
  elif [ $REGION == "all" ]
    then
    p=${pool}
  fi
  FOURFOLD_V=$(grep "CUSTOM=fourfold;" $p | wc -l)
  FOURFOLD_D=$(grep "CUSTOM=fourfold;" $p | awk '{print $11}' | paste -sd+ | bc)
  MISTOL_V=$(grep "CUSTOM=tolerated;" $p | wc -l)
  MISTOL_D=$(grep "CUSTOM=tolerated;" $p | awk '{print $11}' | paste -sd+ | bc)
  MISDEL_V=$(grep "CUSTOM=deleterious;" $p | wc -l)
  MISDEL_D=$(grep "CUSTOM=deleterious;" $p | awk '{print $11}' | paste -sd+ | bc)
  LOF_V=$(grep "CUSTOM=LoF;" $p | wc -l)
  LOF_D=$(grep "CUSTOM=LoF;" $p | awk '{print $11}' | paste -sd+ | bc)
  echo -e "$SAMPLE\t$GEN\t$FOURFOLD_V\t$FOURFOLD_D\t$MISTOL_V\t$MISTOL_D\t$MISDEL_V\t$MISDEL_D\t$LOF_V\t$LOF_D" >> counts_pool_${SUBSET}_${REGION}_summary.txt
  done

scp ramon.pouso@rua2.uvigo.es://share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/gen0-140/counts_pool_gen0-140_all_summary.txt /Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/

```

######Low/High recombination subset:
```{R, engine='bash'}

SUBSET="gen0-140"
REGION="Xchr" #all #autosomes #Xchr
RECOMBINATION="high" #low #high

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/$SUBSET/key_categories
rm counts_pool_${SUBSET}_${REGION}_summary.${RECOMBINATION}_recombination.txt
echo -e "sample\tgeneration\tfourfold_V\tfourfold_D\ttolerated_V\ttolerated_D\tdeleterious_V\tdeleterious_D\tLoF_V\tLoF_D" > counts_pool_${SUBSET}_${REGION}_summary.${RECOMBINATION}_recombination.txt
POOL_LIST=($(ls -v `find . -name '*_'${SUBSET}'_pool_'${REGION}'.txt' -print`))
for pool in "${POOL_LIST[@]}"
  do
  echo "${pool}"
  SAMPLE=$(echo "${pool}" | cut -d'/' -f2 | cut -d'_' -f1)
  GEN=$(echo "${pool}" | cut -d'/' -f2 | cut -d'_' -f2)
  bedtools intersect -a ${pool} -b /share/rdata/ramon.pouso/POOLS/all_gens_0_140/variants/crisp_all_pools_gen0-140.recode_snps.gen0-140_Pb_0_variable_set.orthologs.confident.polarized.nr_annovar.dm6nr_multianno.${RECOMBINATION}_recombination.bed > ${pool/.txt/.${RECOMBINATION}_recombination.txt}
  p=${pool/.txt/.${RECOMBINATION}_recombination.txt}
  
  FOURFOLD_V=$(grep "CUSTOM=fourfold;" $p | wc -l)
  FOURFOLD_D=$(grep "CUSTOM=fourfold;" $p | awk '{print $11}' | paste -sd+ | bc)
  MISTOL_V=$(grep "CUSTOM=tolerated;" $p | wc -l)
  MISTOL_D=$(grep "CUSTOM=tolerated;" $p | awk '{print $11}' | paste -sd+ | bc)
  MISDEL_V=$(grep "CUSTOM=deleterious;" $p | wc -l)
  MISDEL_D=$(grep "CUSTOM=deleterious;" $p | awk '{print $11}' | paste -sd+ | bc)
  LOF_V=$(grep "CUSTOM=LoF;" $p | wc -l)
  LOF_D=$(grep "CUSTOM=LoF;" $p | awk '{print $11}' | paste -sd+ | bc)
  echo -e "$SAMPLE\t$GEN\t$FOURFOLD_V\t$FOURFOLD_D\t$MISTOL_V\t$MISTOL_D\t$MISDEL_V\t$MISDEL_D\t$LOF_V\t$LOF_D" >> counts_pool_${SUBSET}_${REGION}_summary.${RECOMBINATION}_recombination.txt
  done

```

####Keep non-variants VCFs:
#####Sites with 10 or more reads:
######Calculate proportion of derived reads:
```{R, engine='bash'}

SUBSET="gen0-140"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/$SUBSET/
rm propD_pool_${SUBSET}_all_d10_summary.txt
echo -e "sample\tgeneration\tfourfold_V\tfourfold_propD\ttolerated_V\ttolerated_propD\tdeleterious_V\tdeleterious_propD\tLoF_V\tLoF_propD" > propD_pool_${SUBSET}_all_d10_summary.txt
POOL_LIST=($(ls -rt `find . -name '*_'${SUBSET}'_pool_d10.nv.txt' -print`))
for p in "${POOL_LIST[@]}"
  do
  #echo "${p}"
  SAMPLE=$(echo "${p}" | cut -d'/' -f2 | cut -d'_' -f1)
  GEN=$(echo "${p}" | cut -d'/' -f2 | cut -d'_' -f2)
  echo $SAMPLE
  
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed > ${SAMPLE}_${GEN}_${SUBSET}_synonymous_4fold_d10.txt
  FOURFOLD_V=$(wc -l < ${SAMPLE}_${GEN}_${SUBSET}_synonymous_4fold_d10.txt)
  FOURFOLD_SUM=$(awk '{print $11/($10+$11)}' ${SAMPLE}_${GEN}_${SUBSET}_synonymous_4fold_d10.txt | paste -sd+ | bc)
  FOURFOLD_PROP_D=$(echo "scale=5; $FOURFOLD_SUM/$FOURFOLD_V" | bc)

  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed > ${SAMPLE}_${GEN}_${SUBSET}_missense_tolerated_d10.txt
  MISTOL_V=$(wc -l < ${SAMPLE}_${GEN}_${SUBSET}_missense_tolerated_d10.txt)
  MISTOL_SUM=$(awk '{print $11/($10+$11)}' ${SAMPLE}_${GEN}_${SUBSET}_missense_tolerated_d10.txt | paste -sd+ | bc)
  MISTOL_PROP_D=$(echo "scale=5; $MISTOL_SUM/$MISTOL_V" | bc)
  
  awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${p} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed > ${SAMPLE}_${GEN}_${SUBSET}_missense_deleterious_d10.txt
  MISDEL_V=$(wc -l < ${SAMPLE}_${GEN}_${SUBSET}_missense_deleterious_d10.txt)
  MISDEL_SUM=$(awk '{print $11/($10+$11)}' ${SAMPLE}_${GEN}_${SUBSET}_missense_deleterious_d10.txt | paste -sd+ | bc)
  MISDEL_PROP_D=$(echo "scale=5; $MISDEL_SUM/$MISDEL_V" | bc)
  
  LOF_V=$(grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${p} | wc -l)
  LOF_SUM=$(grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${p} | awk '{print $10/($9+$10)}' | paste -sd+ | bc)
  LOF_PROP_D=$(echo "scale=5; $LOF_SUM/$LOF_V" | bc)
  
  echo -e "$SAMPLE\t$GEN\t$FOURFOLD_V\t$FOURFOLD_PROP_D\t$MISTOL_V\t$MISTOL_PROP_D\t$MISDEL_V\t$MISDEL_PROP_D\t$LOF_V\t$LOF_PROP_D" >> propD_pool_${SUBSET}_all_d10_summary.txt
  done

scp ramon.pouso@rua2.uvigo.es://share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/gen0-140/propD_pool_gen0-140_all_d10_summary.txt /Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140

```

######Build SFS per category:
```{R, engine='bash'}

SUBSET="gen0-140"
CATEGORY="LoF" #"intronic", "synonymous", "fourfold", "missense", "tolerated", "deleterious", "LoF"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/$SUBSET/
rm SFS_pool_${SUBSET}_${CATEGORY}_all_summary.txt
echo -e "sample\tgeneration\tcategory\taf_000\taf_000_005\taf_005_010\taf_010_015\taf_015_020\taf_020_025\taf_025_030\taf_030_035\taf_035_040\taf_040_045\taf_045_050\taf_050_055\taf_055_060\taf_060_065\taf_065_070\taf_070_075\taf_075_080\taf_080_085\taf_085_090\taf_090_095\taf_095_100\taf_100" > SFS_pool_${SUBSET}_${CATEGORY}_all_summary.txt
#Then apply the following loop to retrieve the SFS for each pool:
POOL_LIST=($(ls -rt `find . -name '*_'${SUBSET}'_pool_d10.nv.txt' -print`))
for pool in "${POOL_LIST[@]}"
  do
  SAMPLE=$(echo "${pool}" | cut -d'/' -f2 | cut -d'_' -f1)
  GEN=$(echo "${pool}" | cut -d'/' -f2 | cut -d'_' -f2)
  echo $pool
  #First generate the subset file for the respective category:
  if [ $CATEGORY == "intronic" ]
    then grep ';Func.refGene=intronic;' ${pool} > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "synonymous" ]
    then grep ';ExonicFunc.refGene=synonymous_SNV;' ${pool} > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "fourfold" ]
    then awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${pool} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed | awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$3,$4,$5,$6,$7,$8,$9,$10,$11)}' > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "missense" ]
    then grep ';ExonicFunc.refGene=nonsynonymous_SNV;' ${pool} > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "tolerated" ]
    then awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${pool} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed | awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$3,$4,$5,$6,$7,$8,$9,$10,$11)}' > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "deleterious" ]
    then awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$2-1,$2,$3,$4,$5,$6,$7,$8,$9,$10)}' ${pool} | bedtools intersect -a stdin -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed | awk '{printf ("%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", $1,$3,$4,$5,$6,$7,$8,$9,$10,$11)}' > ${pool/.txt/.$CATEGORY.txt}
  elif [ $CATEGORY == "LoF" ]
    then grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${pool} > ${pool/.txt/.$CATEGORY.txt}
  fi
  #Now calculate the number of sites within each AF category. For categories with AF=0 and AF=1, add the number of sites missing from the gen140 VCF:
  AF_000=$(awk '$10=="0"' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_000_005=$(awk '($10/($9+$10)>0.0) && ($10/($9+$10)<=0.05)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_005_010=$(awk '($10/($9+$10)>0.05) && ($10/($9+$10)<=0.1)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_010_015=$(awk '($10/($9+$10)>0.1) && ($10/($9+$10)<=0.15)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_015_020=$(awk '($10/($9+$10)>0.15) && ($10/($9+$10)<=0.2)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_020_025=$(awk '($10/($9+$10)>0.2) && ($10/($9+$10)<=0.25)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_025_030=$(awk '($10/($9+$10)>0.25) && ($10/($9+$10)<=0.3)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_030_035=$(awk '($10/($9+$10)>0.3) && ($10/($9+$10)<=0.35)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_035_040=$(awk '($10/($9+$10)>0.35) && ($10/($9+$10)<=0.4)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_040_045=$(awk '($10/($9+$10)>0.4) && ($10/($9+$10)<=0.45)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_045_050=$(awk '($10/($9+$10)>0.45) && ($10/($9+$10)<=0.5)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_050_055=$(awk '($10/($9+$10)>0.5) && ($10/($9+$10)<=0.55)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_055_060=$(awk '($10/($9+$10)>0.55) && ($10/($9+$10)<=0.6)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_060_065=$(awk '($10/($9+$10)>0.6) && ($10/($9+$10)<=0.65)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_065_070=$(awk '($10/($9+$10)>0.65) && ($10/($9+$10)<=0.7)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_070_075=$(awk '($10/($9+$10)>0.7) && ($10/($9+$10)<=0.75)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_075_080=$(awk '($10/($9+$10)>0.75) && ($10/($9+$10)<=0.8)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_080_085=$(awk '($10/($9+$10)>0.8) && ($10/($9+$10)<=0.85)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_085_090=$(awk '($10/($9+$10)>0.85) && ($10/($9+$10)<=0.9)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_090_095=$(awk '($10/($9+$10)>0.9) && ($10/($9+$10)<=0.95)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_095_100=$(awk '($10/($9+$10)>0.95) && ($10/($9+$10)<1.0)' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  AF_100=$(awk '$9=="0"' ${pool/.txt/.$CATEGORY.txt} | wc -l)
  #Finally, write the SFS to the output:
  echo -e "$SAMPLE\t$GEN\t$CATEGORY\t$AF_000\t$AF_000_005\t$AF_005_010\t$AF_010_015\t$AF_015_020\t$AF_020_025\t$AF_025_030\t$AF_030_035\t$AF_035_040\t$AF_040_045\t$AF_045_050\t$AF_050_055\t$AF_055_060\t$AF_060_065\t$AF_065_070\t$AF_070_075\t$AF_075_080\t$AF_080_085\t$AF_085_090\t$AF_090_095\t$AF_095_100\t$AF_100" >> SFS_pool_${SUBSET}_${CATEGORY}_all_summary.txt
  done

```

######Combine SFS categories:
```{bash}

#Concatenate all category-files into a single one before downloading it:
SUBSET="gen0-140"

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/pools_counts/$SUBSET/
cat SFS_pool_${SUBSET}_fourfold_all_summary.txt <(tail -n+2 SFS_pool_${SUBSET}_tolerated_all_summary.txt) <(tail -n+2 SFS_pool_${SUBSET}_deleterious_all_summary.txt) <(tail -n+2 SFS_pool_${SUBSET}_LoF_all_summary.txt) > SFS_pool_${SUBSET}_combined_all_summary.txt

scp ramon.pouso@rua2.uvigo.es://share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/pools_counts/gen0-140/SFS_pool_gen0-140_combined_all_summary.txt /Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/

```

###Individuals:
```{R, engine='bash'}

#This code doesn't work as a .sh in the cluster because bedtools isn't properly loaded when using the queue system.

SUBSET="autosomes" #autosomes #Xchr

module load gcc/7.2.0
module add gcc/7.2.0
cd /share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/individuals_counts/$SUBSET/
rm counts_individual_${SUBSET}_summary.txt
echo -e "sample\tgeneration\tfourfold_V\tfourfold_D\ttolerated_V\ttolerated_D\tdeleterious_V\tdeleterious_D\tLoF_V\tLoF_D" > counts_individual_${SUBSET}_summary.txt
INDLIST=($(ls `find . -name '*_gen140_gen0-140_individual_'${SUBSET}'.vcf' -print`))
for i in "${INDLIST[@]}"
  do
  #echo "${i}"
  SAMPLE=$(echo "${i}" | cut -d'/' -f2 | cut -d'_' -f1)
  echo $SAMPLE
  bedtools intersect -a ${i} -b /share/rdata/ramon.pouso/counts/synonymous_variants_4fold.bed > ${SAMPLE}_140_${SUBSET}_synonymous_4fold.txt
  FOURFOLD_V=$(wc -l < ${SAMPLE}_140_${SUBSET}_synonymous_4fold.txt)
  FOURFOLD_D=$(cut -f8 ${SAMPLE}_140_${SUBSET}_synonymous_4fold.txt | awk -F";AC=|;AF=" '{printf ("%s\n"),$2}' | paste -sd+ | bc)
  bedtools intersect -a ${i} -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_tolerated.bed > ${SAMPLE}_140_${SUBSET}_missense_tolerated.txt
  MISTOL_V=$(wc -l < ${SAMPLE}_140_${SUBSET}_missense_tolerated.txt)
  MISTOL_D=$(cut -f8 ${SAMPLE}_140_${SUBSET}_missense_tolerated.txt | awk -F";AC=|;AF=" '{printf ("%s\n"),$2}' | paste -sd+ | bc)
  bedtools intersect -a ${i} -b /share/rdata/ramon.pouso/counts/missense_variants_provean_SIFT_deleterious.bed > ${SAMPLE}_140_${SUBSET}_missense_deleterious.txt
  MISDEL_V=$(wc -l < ${SAMPLE}_140_${SUBSET}_missense_deleterious.txt)
  MISDEL_D=$(cut -f8 ${SAMPLE}_140_${SUBSET}_missense_deleterious.txt | awk -F";AC=|;AF=" '{printf ("%s\n"),$2}' | paste -sd+ | bc)
  LOF_V=$(grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${i} | wc -l)
  if [ "$LOF_V" -eq 0 ]; then LOF_D=0; else LOF_D=$(grep -E ';ExonicFunc.refGene=stopgain;|;ExonicFunc.refGene=stoploss;' ${i} | cut -f8 | awk -F";AC=|;AF=" '{printf ("%s\n"),$2}' | paste -sd+ | bc); fi
  echo -e "$SAMPLE\t140\t$FOURFOLD_V\t$FOURFOLD_D\t$MISTOL_V\t$MISTOL_D\t$MISDEL_V\t$MISDEL_D\t$LOF_V\t$LOF_D" >> counts_individual_${SUBSET}_summary.txt
  done


scp ramon.pouso@rua2.uvigo.es://share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/individuals_counts/autosomes/counts_individual_autosomes_summary.txt /Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/
scp ramon.pouso@rua2.uvigo.es://share/rdata/ramon.pouso/counts/sites_in_0-140_VCF/individuals_counts/Xchr/counts_individual_autosomes_summary.txt /Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/

```

#3. Plot counts.
##Derived count ratios:
###No category relativisation:
####Relative to Pb:
#####Gen0-140 pools Pb and line averages (combined version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
pool_counts <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.txt")) 

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$generation <- as.numeric(gsub("gen","",pool_counts$generation))
pool_counts <- pool_counts %>% arrange(generation)
pool_counts$generation = factor(pool_counts$generation)
pool_counts <- pool_counts %>% arrange(population,generation)

pool_counts_copies <- pool_counts %>% mutate(fourfold=fourfold_D,tolerated=tolerated_D,deleterious=deleterious_D,LoF=LoF_D) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_copies

r_average_vector <- c()
for (r in unique(pool_counts_copies$ratio)) {
  print(r)
  r_average <- filter(pool_counts_copies,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_copies,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_copies <- mutate(pool_counts_copies, Pb_relative_value=value/r_average_vector, group=ifelse(population=="Pb","Pb","lines"))
relativised_pool_counts_copies$group = factor(relativised_pool_counts_copies$group,levels=c("Pb","lines"))

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_copies <- data_frame("population"=character(0),"generation"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_copies$group)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_copies,group==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_copies,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_copies$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_copies,ratio==r & generation==g & group==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_copies,ratio==r & generation==g & group==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_copies <- rbind(average_relativised_pool_counts_copies,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_copies$population = factor(average_relativised_pool_counts_copies$population,levels=c("Pb","lines"))
average_relativised_pool_counts_copies$ratio = factor(average_relativised_pool_counts_copies$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_copies$generation <- as.factor(as.numeric(average_relativised_pool_counts_copies$generation))
average_relativised_pool_counts_copies$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies$avg_Pb_relative_value)
average_relativised_pool_counts_copies$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies$se_Pb_relative_value)
average_relativised_pool_counts_copies
#write_tsv(average_relativised_pool_counts_copies,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_",type,".txt"))


#Combined version (Pb):
Pb_lines_relativised_pool_counts_copies_ggplot <- ggplot(data=filter(average_relativised_pool_counts_copies,ratio!="fourfold", generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_errorbar(aes(ymin=avg_Pb_relative_value-2*se_Pb_relative_value, ymax=avg_Pb_relative_value+2*se_Pb_relative_value, colour=generation, group=population), position=position_dodge(0.8),size=0.5,width=0.5) +
  geom_point(aes(colour=generation,shape=population),size=2,position=position_dodge(0.8)) + 
  #ggtitle("Proportion of reads at different NM") +
  ylab("Derived count relative to Pb-000") +
  scale_y_continuous(breaks = seq(0.5, 1.0, by = 0.1)) +
  scale_shape_manual(values=c(1,16)) +
  ylim(0.4,1.1) +
  ggtitle("Pools") +
  guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
Pb_lines_relativised_pool_counts_copies_ggplot
ggsave(paste0("Pb_lines_relativised_pool_counts_copies_",type,".pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

###Relative to synonymous:
####Relative to Pb:
#####Individuals (relative to Pb-000):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "Xchr" #autosomes #Xchr

#Import the individual counts:
wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
individual_counts <- read_tsv(paste0(wd_path,"counts_individual_",type,"_summary.txt"))
individual_counts <- mutate(individual_counts, population=gsub("-","",substr(sample,1,3)))
individual_counts$population = factor(individual_counts$population)
individual_counts$population <- gsub('Pb', 'Pb-140', individual_counts$population)

#Relativise them by the synonymous category:
individual_counts_copies_synR <- individual_counts %>% mutate(synonymous=synonymous_D/synonymous_D,missense=missense_D/synonymous_D,tolerated=tolerated_D/synonymous_D,deleterious=deleterious_D/synonymous_D,LoF=(stopgain_D+stoploss_D)/synonymous_D) %>% select(c(1,18:23)) %>% gather(ratio,value,-sample,-population,factor_key=T)
individual_counts_copies_synR

#Import the pool counts and relativise them by the synonymous category in order to retrieve the Pb-000 relativised values (which we'll use as reference here).
pool_counts <- read_tsv(paste0(wd_path,"counts_pool_gen0-40_",type,"_summary.txt")) %>% mutate(gen=ifelse(sample=="LBT0",000,ifelse(sample=="c1",005,ifelse(sample=="BT20",020,ifelse(sample=="BT30",030,040)))))
codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$gen = factor(pool_counts$gen)
pool_counts <- pool_counts %>% arrange(population,gen)
pool_counts_copies_synR <- pool_counts %>% mutate(synonymous=synonymous_D/synonymous_D,missense=missense_D/synonymous_D,tolerated=tolerated_D/synonymous_D,deleterious=deleterious_D/synonymous_D,LoF=(stopgain_D+stoploss_D)/synonymous_D) %>% select(c(1,26:32)) %>% gather(ratio,value,-sample,-gen,-population,factor_key=T)
pool_counts_copies_synR

#Retrieve the Pb-000 syn-relativised values with the structure of the individual counts database, and relativise the individual values by the Pb-000 ones:
r_average_vector <- c()
for (r in unique(individual_counts_copies_synR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_copies_synR,r==ratio & population=="Pb" & gen==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(individual_counts_copies_synR,r==ratio))))
}
print(r_average_vector)

relativised_individual_counts_copies_synR <- mutate(individual_counts_copies_synR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_individual_counts_copies_synR <- data_frame("population"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_individual_counts_copies_synR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  #species <- filter(relativised_individual_counts_copies_synR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
  for (r in unique(relativised_individual_counts_copies_synR$ratio)) {
    print(r)
    pop_mean <- filter(relativised_individual_counts_copies_synR,ratio==r & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
    #print(paste0(pop," feature ",r," average is ",pop_mean))
    pop_se <- filter(relativised_individual_counts_copies_synR,ratio==r & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
    #print(paste0(pop," feature ",r," std error is ",pop_se))
    row_data <- cbind(pop,r,pop_mean,pop_se)
    colnames(row_data) <- c("population","ratio","avg_Pb_relative_value","se_Pb_relative_value")
    average_relativised_individual_counts_copies_synR <- rbind(average_relativised_individual_counts_copies_synR,row_data,stringsAsFactors=F)
  }
}
average_relativised_individual_counts_copies_synR$population = factor(average_relativised_individual_counts_copies_synR$population,levels=c("Pb-140","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
average_relativised_individual_counts_copies_synR$ratio = factor(average_relativised_individual_counts_copies_synR$ratio,levels=c("synonymous","missense","tolerated","deleterious","LoF"))

average_relativised_individual_counts_copies_synR$avg_Pb_relative_value <- as.numeric(average_relativised_individual_counts_copies_synR$avg_Pb_relative_value)
average_relativised_individual_counts_copies_synR$se_Pb_relative_value <- as.numeric(average_relativised_individual_counts_copies_synR$se_Pb_relative_value)
average_relativised_individual_counts_copies_synR
write_tsv(average_relativised_individual_counts_copies_synR,paste0(wd_path,"/pop_average_relativised_individual_counts_copies_synR_",type,".txt"))

#Plot these means and standard errors:
average_relativised_individual_counts_copies_synR_ggplot <- ggplot(data=filter(average_relativised_individual_counts_copies_synR,ratio!="synonymous"), aes(population,avg_Pb_relative_value,colour=population)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_point() +
  geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  geom_hline(yintercept=1,linetype="dashed",alpha=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count \n relative to synonymous and Pb-000") +
  scale_y_continuous(breaks = seq(0.6, 1.6, by = 0.2)) +
  ylim(0.6,1.6) +
  ggtitle(paste0("Individuals ",type)) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
average_relativised_individual_counts_copies_synR_ggplot
ggsave(paste0("pop_average_relativised_individual_counts_copies_synR_",type,".pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

#####Individuals (relative to Pb-140):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "autosomes" #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
individual_counts <- read_tsv(paste0(wd_path,"counts_individual_",type,"_summary.txt"))
individual_counts <- mutate(individual_counts, population=gsub("-","",substr(sample,1,3)))
individual_counts$population = factor(individual_counts$population)
individual_counts$population <- gsub('Pb', 'Pb-140', individual_counts$population)


individual_counts_copies_synR <- individual_counts %>% mutate(synonymous=synonymous_D/synonymous_D,missense=missense_D/synonymous_D,LoF=(stopgain_D+stoploss_D)/synonymous_D) %>% select(c(1,14:17)) %>% gather(ratio,value,-sample,-population,factor_key=T)
individual_counts_copies_synR

r_average_vector <- c()
for (r in unique(individual_counts_copies_synR$ratio)) {
  print(r)
  r_average <- filter(individual_counts_copies_synR,r==ratio & population=="Pb-140") %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(individual_counts_copies_synR,r==ratio))))
}
print(r_average_vector)

relativised_individual_counts_copies_synR <- mutate(individual_counts_copies_synR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_individual_counts_copies_synR <- data_frame("population"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_individual_counts_copies_synR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  #species <- filter(relativised_individual_counts_copies_synR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
  for (r in unique(relativised_individual_counts_copies_synR$ratio)) {
    print(r)
    pop_mean <- filter(relativised_individual_counts_copies_synR,ratio==r & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
    #print(paste0(pop," feature ",r," average is ",pop_mean))
    pop_se <- filter(relativised_individual_counts_copies_synR,ratio==r & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
    #print(paste0(pop," feature ",r," std error is ",pop_se))
    row_data <- cbind(pop,r,pop_mean,pop_se)
    colnames(row_data) <- c("population","ratio","avg_Pb_relative_value","se_Pb_relative_value")
    average_relativised_individual_counts_copies_synR <- rbind(average_relativised_individual_counts_copies_synR,row_data,stringsAsFactors=F)
  }
}
average_relativised_individual_counts_copies_synR$population = factor(average_relativised_individual_counts_copies_synR$population)
average_relativised_individual_counts_copies_synR$ratio = factor(average_relativised_individual_counts_copies_synR$ratio,levels=c("synonymous","missense","LoF"))

average_relativised_individual_counts_copies_synR$avg_Pb_relative_value <- as.numeric(average_relativised_individual_counts_copies_synR$avg_Pb_relative_value)
average_relativised_individual_counts_copies_synR$se_Pb_relative_value <- as.numeric(average_relativised_individual_counts_copies_synR$se_Pb_relative_value)
average_relativised_individual_counts_copies_synR
write_tsv(average_relativised_individual_counts_copies_synR,paste0(wd_path,"/pop_average_relativised_individual_counts_copies_synR_",type,".txt"))

#Plot these means and standard errors:
average_relativised_individual_counts_copies_synR_ggplot <- ggplot(data=filter(average_relativised_individual_counts_copies_synR,ratio!="synonymous"), aes(population,avg_Pb_relative_value,colour=population)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_point() +
  geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count \n relative to synonymous and Pb-140") +
  scale_y_continuous(breaks = seq(0.8, 1.6, by = 0.2)) +
  ylim(0.8,1.6) +
  ggtitle(paste0("Individuals ",type)) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
average_relativised_individual_counts_copies_synR_ggplot
ggsave(paste0("pop_average_relativised_individual_counts_copies_synR_",type,".pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

#####Pools (combined):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "Xchr" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
pool_counts_gen0_40 <- read_tsv(paste0(wd_path,"counts_pool_gen0-40_",type,"_summary.txt")) %>% mutate(gen=ifelse(sample=="LBT0",000,ifelse(sample=="c1",005,ifelse(sample=="BT20",020,ifelse(sample=="BT30",030,040)))))
pool_counts_gen140 <- read_tsv(paste0(wd_path,"counts_pool_gen140_",type,"_summary.txt")) %>% mutate(gen="140" %>% as.double)

pool_counts <- rbind(pool_counts_gen0_40,pool_counts_gen140)

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$gen = factor(pool_counts$gen)
pool_counts <- pool_counts %>% arrange(population,gen)

pool_counts_copies_synR <- pool_counts %>% mutate(synonymous=synonymous_D/synonymous_D,missense=missense_D/synonymous_D,tolerated=tolerated_D/synonymous_D,deleterious=deleterious_D/synonymous_D,LoF=(stopgain_D+stoploss_D)/synonymous_D) %>% select(c(1,26:32)) %>% gather(ratio,value,-sample,-gen,-population,factor_key=T)
pool_counts_copies_synR

r_average_vector <- c()
for (r in unique(pool_counts_copies_synR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_copies_synR,r==ratio & population=="Pb" & gen==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_copies_synR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_copies_synR <- mutate(pool_counts_copies_synR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_copies_synR <- data_frame("population"=character(0),"generation"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_copies_synR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_copies_synR,population==pop),gen)),use.names=F)) {
  #species <- filter(relativised_pool_counts_copies_synR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_copies_synR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_copies_synR,ratio==r & gen==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_copies_synR,ratio==r & gen==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_copies_synR <- rbind(average_relativised_pool_counts_copies_synR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_copies_synR$population = factor(average_relativised_pool_counts_copies_synR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
average_relativised_pool_counts_copies_synR$ratio = factor(average_relativised_pool_counts_copies_synR$ratio,levels=c("synonymous","missense","tolerated","deleterious","LoF"))
average_relativised_pool_counts_copies_synR$generation <- as.factor(as.numeric(average_relativised_pool_counts_copies_synR$generation))
average_relativised_pool_counts_copies_synR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_synR$avg_Pb_relative_value)
average_relativised_pool_counts_copies_synR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_synR$se_Pb_relative_value)
average_relativised_pool_counts_copies_synR
write_tsv(average_relativised_pool_counts_copies_synR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_synR_",type,".txt"))

#Plot these means and standard errors:
average_relativised_pool_counts_copies_synR_ggplot <- ggplot(data=filter(average_relativised_pool_counts_copies_synR,ratio!="synonymous"), aes(population,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_point(aes(colour=generation),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to synonymous and Pb-000") +
  scale_y_continuous(breaks = seq(0.6, 1.2, by = 0.2)) +
  ylim(0.5,1.4) +
  ggtitle(paste0("Pools ",type)) +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
average_relativised_pool_counts_copies_synR_ggplot
ggsave(paste0("pop_average_relativised_pool_counts_copies_synR_",type,".pdf"), width=40, height=12, units="cm", device="pdf", path=wd_path)

```

#####Pools (separate):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "gen0-40" #gen0-40 #gen140
subtype <- "Xchr" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
pool_counts <- read_tsv(paste0(wd_path,"counts_pool_",type,"_",subtype,"_summary.txt"))

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old")) %>% arrange(population)
pool_counts$population = factor(pool_counts$population)

pool_counts_copies_synR <- pool_counts %>% mutate(synonymous=synonymous_D/synonymous_D,missense=missense_D/synonymous_D,LoF=(stopgain_D+stoploss_D)/synonymous_D) %>% select(c(1,20:23)) %>% gather(ratio,value,-sample,-population,factor_key=T)
pool_counts_copies_synR

r_average_vector <- c()
for (r in unique(pool_counts_copies_synR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_copies_synR,r==ratio & population=="Pb-000") %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_copies_synR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_copies_synR <- mutate(pool_counts_copies_synR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_copies_synR <- data_frame("population"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_copies_synR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  #species <- filter(relativised_pool_counts_copies_synR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
  for (r in unique(relativised_pool_counts_copies_synR$ratio)) {
    print(r)
    pop_mean <- filter(relativised_pool_counts_copies_synR,ratio==r & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
    #print(paste0(pop," feature ",r," average is ",pop_mean))
    pop_se <- filter(relativised_pool_counts_copies_synR,ratio==r & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
    #print(paste0(pop," feature ",r," std error is ",pop_se))
    row_data <- cbind(pop,r,pop_mean,pop_se)
    colnames(row_data) <- c("population","ratio","avg_Pb_relative_value","se_Pb_relative_value")
    average_relativised_pool_counts_copies_synR <- rbind(average_relativised_pool_counts_copies_synR,row_data,stringsAsFactors=F)
  }
}
average_relativised_pool_counts_copies_synR$population = factor(average_relativised_pool_counts_copies_synR$population)
average_relativised_pool_counts_copies_synR$ratio = factor(average_relativised_pool_counts_copies_synR$ratio,levels=c("synonymous","missense","LoF"))

average_relativised_pool_counts_copies_synR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_synR$avg_Pb_relative_value)
average_relativised_pool_counts_copies_synR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_synR$se_Pb_relative_value)
average_relativised_pool_counts_copies_synR
write_tsv(average_relativised_pool_counts_copies_synR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_synR_",type,"_",subtype,".txt"))

#Plot these means and standard errors:
average_relativised_pool_counts_copies_synR_ggplot <- ggplot(data=filter(average_relativised_pool_counts_copies_synR,ratio!="synonymous"), aes(population,avg_Pb_relative_value,colour=population)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_point() +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to synonymous and Pb-000") +
  scale_y_continuous(breaks = seq(0.6, 1.2, by = 0.2)) +
  ylim(0.5,1.2) +
  ggtitle(paste0("Pools ",type," ",subtype)) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=8),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
average_relativised_pool_counts_copies_synR_ggplot
ggsave(paste0("pop_average_relativised_pool_counts_copies_synR_",type,"_",subtype,".pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

#####Pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
pool_counts_gen0_40 <- read_tsv(paste0(wd_path,"counts_pool_gen0-40_",type,"_summary.txt")) %>% mutate(gen=ifelse(sample=="LBT0",000,ifelse(sample=="c1",005,ifelse(sample=="BT20",020,ifelse(sample=="BT30",030,040)))))
pool_counts_gen140 <- read_tsv(paste0(wd_path,"counts_pool_gen140_",type,"_summary.txt")) %>% mutate(gen="140" %>% as.double)

pool_counts <- rbind(pool_counts_gen0_40,pool_counts_gen140)

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$gen = factor(pool_counts$gen)
pool_counts <- pool_counts %>% arrange(population,gen)

pool_counts_copies_synR <- pool_counts %>% mutate(synonymous=synonymous_D/synonymous_D,missense=missense_D/synonymous_D,tolerated=tolerated_D/synonymous_D,deleterious=deleterious_D/synonymous_D,LoF=(stopgain_D+stoploss_D)/synonymous_D) %>% select(c(1,26:32)) %>% gather(ratio,value,-sample,-gen,-population,factor_key=T)
pool_counts_copies_synR

#pool_line_average_counts <- pool_counts %>% filter(population!="Pb") %>% group_by(gen) %>% summarise_if(is.numeric, mean, na.rm = TRUE) %>% mutate(population="lines")

#pool_line_average_counts_copies_synR <- pool_line_average_counts %>% mutate(synonymous=synonymous_D/synonymous_D,missense=missense_D/synonymous_D,tolerated=tolerated_D/synonymous_D,deleterious=deleterious_D/synonymous_D,LoF=(stopgain_D+stoploss_D)/synonymous_D) %>% select(c(1,26:31)) %>% gather(ratio,value,-gen,-population,factor_key=T)
#pool_line_average_counts_copies_synR

r_average_vector <- c()
for (r in unique(pool_counts_copies_synR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_copies_synR,r==ratio & population=="Pb" & gen==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_copies_synR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_copies_synR <- mutate(pool_counts_copies_synR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_copies_synR <- data_frame("population"=character(0),"generation"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_copies_synR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_copies_synR,population==pop),gen)),use.names=F)) {
  #species <- filter(relativised_pool_counts_copies_synR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_copies_synR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_copies_synR,ratio==r & gen==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_copies_synR,ratio==r & gen==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_copies_synR <- rbind(average_relativised_pool_counts_copies_synR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_copies_synR <- average_relativised_pool_counts_copies_synR %>% mutate(group=ifelse(population=="Pb","Pb","lines"))
average_relativised_pool_counts_copies_synR$population = factor(average_relativised_pool_counts_copies_synR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
average_relativised_pool_counts_copies_synR$ratio = factor(average_relativised_pool_counts_copies_synR$ratio,levels=c("synonymous","missense","tolerated","deleterious","LoF"))
average_relativised_pool_counts_copies_synR$generation <- as.factor(as.numeric(average_relativised_pool_counts_copies_synR$generation))
average_relativised_pool_counts_copies_synR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_synR$avg_Pb_relative_value)
average_relativised_pool_counts_copies_synR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_synR$se_Pb_relative_value)
average_relativised_pool_counts_copies_synR
#write_tsv(average_relativised_pool_counts_copies_synR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_synR_",type,".txt"))


#Humberto version (Pb):
average_relativised_pool_counts_copies_synR_ggplot <- ggplot(data=filter(average_relativised_pool_counts_copies_synR,ratio!="synonymous" & population=="Pb" & generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_point(aes(colour=generation),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to synonymous and Pb-000") +
  scale_y_continuous(breaks = seq(0.6, 1.2, by = 0.2)) +
  ylim(0.5,1.4) +
  ggtitle(paste0("Pools ",type)) +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
average_relativised_pool_counts_copies_synR_ggplot
ggsave(paste0("Pb_relativised_pool_counts_copies_synR_",type,"_Humberto_version.pdf"), width=20, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
average_relativised_pool_counts_copies_synR_ggplot <- ggplot(data=filter(average_relativised_pool_counts_copies_synR,ratio!="synonymous" & group=="lines"), aes(ratio,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=generation)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to synonymous and Pb-000") +
  scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  ylim(0.7,1.05) +
  ggtitle(paste0("Pools ",type)) +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
average_relativised_pool_counts_copies_synR_ggplot
ggsave(paste0("lines_relativised_pool_counts_copies_synR_",type,"_Humberto_version.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)


```

####Raw:
#####Individuals:
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "autosomes" #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
individual_counts <- read_tsv(paste0(wd_path,"counts_individual_",type,"_summary.txt"))
individual_counts <- mutate(individual_counts, population=gsub("-","",substr(sample,1,3)))
individual_counts$population = factor(individual_counts$population)
individual_counts$population <- gsub('Pb', 'Pb-140', individual_counts$population)


individual_counts_copies_synR <- individual_counts %>% mutate(synonymous=synonymous_D/synonymous_D,missense=missense_D/synonymous_D,LoF=(stopgain_D+stoploss_D)/synonymous_D) %>% select(c(1,14:17)) %>% gather(ratio,value,-sample,-population,factor_key=T)
individual_counts_copies_synR

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_individual_counts_copies_synR <- data_frame("population"=character(0),"ratio"=character(0),"avg_value"=character(0),"se_value"=character(0)) #next, create the empty dataframe

for (pop in unique(individual_counts_copies_synR$population)) { #then loop over each population and feature to get the mean and standard error, and feed the dataframe
  print(pop)
  #species <- filter(relativised_individual_counts_copies_synR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
  for (r in unique(individual_counts_copies_synR$ratio)) {
    print(r)
    pop_mean <- filter(individual_counts_copies_synR,ratio==r & population==pop) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
    #print(paste0(pop," feature ",r," average is ",pop_mean))
    pop_se <- filter(individual_counts_copies_synR,ratio==r & population==pop) %>% select(value) %>% unlist(.,use.names=F) %>% se()
    #print(paste0(pop," feature ",r," std error is ",pop_se))
    row_data <- cbind(pop,r,pop_mean,pop_se)
    colnames(row_data) <- c("population","ratio","avg_value","se_value")
    average_individual_counts_copies_synR <- rbind(average_individual_counts_copies_synR,row_data,stringsAsFactors=F)
  }
}
average_individual_counts_copies_synR$population = factor(average_individual_counts_copies_synR$population)
average_individual_counts_copies_synR$ratio = factor(average_individual_counts_copies_synR$ratio,levels=c("synonymous","missense","LoF"))

average_individual_counts_copies_synR$avg_value <- as.numeric(average_individual_counts_copies_synR$avg_value)
average_individual_counts_copies_synR$se_value <- as.numeric(average_individual_counts_copies_synR$se_value)
average_individual_counts_copies_synR

#Plot these means and standard errors:
average_individual_counts_copies_synR_ggplot <- ggplot(data=filter(average_individual_counts_copies_synR,ratio!="synonymous"), aes(population,avg_value,colour=population)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_wrap(. ~ ratio, scales = "free") +
  geom_point() +
  geom_errorbar(aes(ymin=avg_value-se_value, ymax=avg_value+se_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count \n relative to synonymous") +
  #scale_y_continuous(breaks = seq(0.8, 1.6, by = 0.2)) +
  #ylim(0.8,1.6) +
  ggtitle(paste0("Individuals ",type)) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
average_individual_counts_copies_synR_ggplot
ggsave(paste0("pop_average_raw_individual_counts_copies_synR_",type,".pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

#####Pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
pool_counts_gen0_40 <- read_tsv(paste0(wd_path,"counts_pool_gen0-40_",type,"_summary.txt")) %>% mutate(gen=ifelse(sample=="LBT0",000,ifelse(sample=="c1",005,ifelse(sample=="BT20",020,ifelse(sample=="BT30",030,040)))))
pool_counts_gen140 <- read_tsv(paste0(wd_path,"counts_pool_gen140_",type,"_summary.txt")) %>% mutate(gen="140" %>% as.double)

pool_counts <- rbind(pool_counts_gen0_40,pool_counts_gen140)

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$gen = factor(pool_counts$gen)
pool_counts <- pool_counts %>% arrange(population,gen)

pool_counts_copies_synR <- pool_counts %>% mutate(synonymous=synonymous_D/synonymous_D,missense=missense_D/synonymous_D,tolerated=tolerated_D/synonymous_D,deleterious=deleterious_D/synonymous_D,LoF=(stopgain_D+stoploss_D)/synonymous_D) %>% select(c(1,26:32)) %>% gather(ratio,value,-sample,-gen,-population,factor_key=T)
pool_counts_copies_synR

pool_counts_copies_synR <- pool_counts_copies_synR %>% mutate(group=ifelse(population=="Pb","Pb","lines"))
pool_counts_copies_synR$population = factor(pool_counts_copies_synR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts_copies_synR$ratio = factor(pool_counts_copies_synR$ratio,levels=c("synonymous","missense","tolerated","deleterious","LoF"))
#write_tsv(average_relativised_pool_counts_copies_synR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_synR_",type,".txt"))

#Humberto version (Pb):
Pb_pool_counts_copies_synR_ggplot <- ggplot(data=filter(pool_counts_copies_synR,ratio!="synonymous" & population=="Pb" & gen!=5), aes(gen,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_wrap(. ~ ratio, scales = "free", nrow=1) +
  geom_point(aes(colour=gen),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to synonymous") +
  #scale_y_continuous(breaks = seq(0.6, 1.2, by = 0.2)) +
  #ylim(0.5,1.4) +
  ggtitle(paste0("Pools ",type)) +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_pool_counts_copies_synR_ggplot
ggsave(paste0("Pb_pool_counts_copies_synR_",type,"_Humberto_version.pdf"), width=20, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
line_pool_counts_copies_synR_ggplot <- ggplot(data=filter(pool_counts_copies_synR,ratio!="synonymous" & group=="lines"), aes(ratio,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=gen)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to synonymous") +
  #scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  #ylim(0.7,1.05) +
  ggtitle(paste0("Pools ",type)) +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
line_pool_counts_copies_synR_ggplot
ggsave(paste0("lines_pool_counts_copies_synR_",type,"_Humberto_version.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

###Relative to 4fold:
####Relative to Pb:
#####Pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
pool_counts_gen0_40 <- read_tsv(paste0(wd_path,"counts_pool_gen0-40_",type,"_summary.txt")) %>% mutate(gen=ifelse(sample=="LBT0",000,ifelse(sample=="c1",005,ifelse(sample=="BT20",020,ifelse(sample=="BT30",030,040)))))
pool_counts_gen140 <- read_tsv(paste0(wd_path,"counts_pool_gen140_",type,"_summary.txt")) %>% mutate(gen="140" %>% as.double)

pool_counts <- rbind(pool_counts_gen0_40,pool_counts_gen140)

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$gen = factor(pool_counts$gen)
pool_counts <- pool_counts %>% arrange(population,gen)

pool_counts_copies_4FR <- pool_counts %>% mutate(fourfold=fourfold_D/fourfold_D,missense=missense_D/fourfold_D,tolerated=tolerated_D/fourfold_D,deleterious=deleterious_D/fourfold_D,LoF=(stopgain_D+stoploss_D)/fourfold_D) %>% select(c(1,29:35)) %>% gather(ratio,value,-sample,-gen,-population,factor_key=T)
pool_counts_copies_4FR

#pool_line_average_counts <- pool_counts %>% filter(population!="Pb") %>% group_by(gen) %>% summarise_if(is.numeric, mean, na.rm = TRUE) %>% mutate(population="lines")

#pool_line_average_counts_copies_synR <- pool_line_average_counts %>% mutate(synonymous=synonymous_D/synonymous_D,missense=missense_D/synonymous_D,tolerated=tolerated_D/synonymous_D,deleterious=deleterious_D/synonymous_D,LoF=(stopgain_D+stoploss_D)/synonymous_D) %>% select(c(1,26:31)) %>% gather(ratio,value,-gen,-population,factor_key=T)
#pool_line_average_counts_copies_synR

r_average_vector <- c()
for (r in unique(pool_counts_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_copies_4FR,r==ratio & population=="Pb" & gen==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_copies_4FR <- mutate(pool_counts_copies_4FR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_copies_4FR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_copies_4FR,population==pop),gen)),use.names=F)) {
  #species <- filter(relativised_pool_counts_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_copies_4FR,ratio==r & gen==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_copies_4FR,ratio==r & gen==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_copies_4FR <- rbind(average_relativised_pool_counts_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_copies_4FR <- average_relativised_pool_counts_copies_4FR %>% mutate(group=ifelse(population=="Pb","Pb","lines"))
average_relativised_pool_counts_copies_4FR$population = factor(average_relativised_pool_counts_copies_4FR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
average_relativised_pool_counts_copies_4FR$ratio = factor(average_relativised_pool_counts_copies_4FR$ratio,levels=c("fourfold","missense","tolerated","deleterious","LoF"))
average_relativised_pool_counts_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_copies_4FR$generation))
average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_copies_4FR
#write_tsv(average_relativised_pool_counts_copies_4FR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_4FR_",type,".txt"))


#Humberto version (Pb):
average_relativised_pool_counts_copies_4FR_ggplot <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR,ratio!="fourfold", ratio!="missense",population=="Pb" & generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_point(aes(colour=generation),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to 4-fold syn. and Pb-000") +
  scale_y_continuous(breaks = seq(0.6, 1.2, by = 0.2)) +
  ylim(0.5,1.4) +
  ggtitle("Pools, Pb") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
average_relativised_pool_counts_copies_4FR_ggplot
ggsave(paste0("Pb_relativised_pool_counts_copies_4FR_",type,"_Humberto_version.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
average_relativised_pool_counts_copies_4FR_ggplot <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR, ratio!="fourfold",ratio!="missense" & group=="lines"), aes(ratio,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=generation)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to 4-fold syn. and Pb-000") +
  scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  ylim(0.7,1.05) +
  ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
average_relativised_pool_counts_copies_4FR_ggplot
ggsave(paste0("lines_relativised_pool_counts_copies_4FR_",type,"_Humberto_version.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)


```

#####Individuals: MODIFY USING POOLS 0-140 BELOW
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "autosomes" #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
individual_counts <- read_tsv(paste0(wd_path,"counts_individual_",type,"_summary.txt"))
individual_counts <- mutate(individual_counts, population=gsub("-","",substr(sample,1,3)), individual=gsub("-","",substr(sample,nchar(sample)-1,nchar(sample))))
individual_counts$population = factor(individual_counts$population)
individual_counts$individual = as.numeric(individual_counts$individual)

individual_counts$generation <- as.numeric(individual_counts$generation)
individual_counts <- individual_counts %>% arrange(generation)
individual_counts$generation = factor(individual_counts$generation)
individual_counts <- individual_counts %>% arrange(population,generation)

individual_counts_copies_4FR <- individual_counts %>% mutate(fourfold=fourfold_D/fourfold_D,tolerated=tolerated_D/fourfold_D,deleterious=deleterious_D/fourfold_D,LoF=LoF_D/fourfold_D) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,-individual,factor_key=T)
individual_counts_copies_4FR

!!! RELATIVIZAR POR Pb-0 de los pools?

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_individual_counts_copies_4FR <- data_frame("population"=character(0),"ratio"=character(0),"avg_value"=character(0),"se_value"=character(0)) #next, create the empty dataframe

for (pop in unique(individual_counts_copies_4FR$population)) { #then loop over each population and feature to get the mean and standard error, and feed the dataframe
  print(pop)
  #species <- filter(relativised_individual_counts_copies_synR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
  for (r in unique(individual_counts_copies_4FR$ratio)) {
    print(r)
    pop_mean <- filter(individual_counts_copies_4FR,ratio==r & population==pop) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
    #print(paste0(pop," feature ",r," average is ",pop_mean))
    pop_se <- filter(individual_counts_copies_4FR,ratio==r & population==pop) %>% select(value) %>% unlist(.,use.names=F) %>% se()
    #print(paste0(pop," feature ",r," std error is ",pop_se))
    row_data <- cbind(pop,r,pop_mean,pop_se)
    colnames(row_data) <- c("population","ratio","avg_value","se_value")
    average_individual_counts_copies_4FR <- rbind(average_individual_counts_copies_4FR,row_data,stringsAsFactors=F)
  }
}
average_individual_counts_copies_4FR$population = factor(average_individual_counts_copies_4FR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
average_individual_counts_copies_4FR$ratio = factor(average_individual_counts_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
#average_individual_counts_copies_4FR$generation = as.factor(average_individual_counts_copies_4FR$generation)
average_individual_counts_copies_4FR$avg_value = as.numeric(average_individual_counts_copies_4FR$avg_value)
average_individual_counts_copies_4FR$se_value = as.numeric(average_individual_counts_copies_4FR$se_value)

#Plot these means and standard errors:
Pb_lines_average_individual_counts_copies_4FR_ggplot <- ggplot(data=filter(average_individual_counts_copies_4FR,ratio!="fourfold"), aes(population,avg_value,colour=population)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_wrap(. ~ ratio, scales = "free") +
  geom_point(aes(colour=population),size=2,position=position_dodge(0.8)) +
  geom_errorbar(aes(ymin=avg_value-se_value, ymax=avg_value+se_value), position=position_dodge(), width=0.5) +
  ylab("Derived count relative to 4FR") +
  #scale_y_continuous(breaks = seq(0.6, 1.0, by = 0.1)) +
  #ylim(0.6,1.1) +
  ggtitle(paste0("Individuals ",type)) +
  guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10,angle=90,hjust=1,vjust=0.5),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
Pb_lines_average_individual_counts_copies_4FR_ggplot
ggsave(paste0("Pb_lines_pop_average_raw_individual_counts_copies_4FR_",type,".pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```


#####Gen0-140 pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
pool_counts <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.txt")) 

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$generation <- as.numeric(gsub("gen","",pool_counts$generation))
pool_counts <- pool_counts %>% arrange(generation)
pool_counts$generation = factor(pool_counts$generation)
pool_counts <- pool_counts %>% arrange(population,generation)

pool_counts_copies_4FR <- pool_counts %>% mutate(fourfold=fourfold_D/fourfold_D,tolerated=tolerated_D/fourfold_D,deleterious=deleterious_D/fourfold_D,LoF=LoF_D/fourfold_D) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_copies_4FR <- mutate(pool_counts_copies_4FR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_copies_4FR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_copies_4FR,population==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_copies_4FR <- rbind(average_relativised_pool_counts_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_copies_4FR$population = factor(average_relativised_pool_counts_copies_4FR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
average_relativised_pool_counts_copies_4FR$ratio = factor(average_relativised_pool_counts_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_copies_4FR$generation))
average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_copies_4FR
#write_tsv(average_relativised_pool_counts_copies_4FR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_4FR_",type,".txt"))


#Humberto version (Pb):
Pb_relativised_pool_counts_copies_4FR_ggplot <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR,ratio!="fourfold",population=="Pb" & generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_point(aes(colour=generation),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Derived count\n relative to 4-fold syn. and Pb-000") +
  scale_y_continuous(breaks = seq(0.6, 1.0, by = 0.1)) +
  ylim(0.6,1.1) +
  ggtitle("Base population pools") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_relativised_pool_counts_copies_4FR_ggplot
ggsave(paste0("Pb_relativised_pool_counts_copies_4FR_",type,"_Humberto_version.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
lines_relativised_pool_counts_copies_4FR_ggplot <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR, ratio!="fourfold" & population!="Pb"), aes(ratio,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=generation)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Derived count\n relative to 4-fold syn. and Pb-000") +
  scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  ylim(0.7,1.05) +
  ggtitle("Line pools") +
  #guides(colour="none") +
  scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  labs(colour = "Generation") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        axis.title.x=element_blank(),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_text()
  )
lines_relativised_pool_counts_copies_4FR_ggplot
ggsave(paste0("lines_relativised_pool_counts_copies_4FR_",type,"_Humberto_version.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)


```

#####Gen0-140 pools Pb and line averages (combined version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
pool_counts <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.txt")) 

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$generation <- as.numeric(gsub("gen","",pool_counts$generation))
pool_counts <- pool_counts %>% arrange(generation)
pool_counts$generation = factor(pool_counts$generation)
pool_counts <- pool_counts %>% arrange(population,generation)

pool_counts_copies_4FR <- pool_counts %>% mutate(fourfold=fourfold_D/fourfold_D,tolerated=tolerated_D/fourfold_D,deleterious=deleterious_D/fourfold_D,LoF=LoF_D/fourfold_D) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_copies_4FR <- mutate(pool_counts_copies_4FR, Pb_relative_value=value/r_average_vector, group=ifelse(population=="Pb","Pb","lines"))
relativised_pool_counts_copies_4FR$group = factor(relativised_pool_counts_copies_4FR$group,levels=c("Pb","lines"))

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_copies_4FR$group)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_copies_4FR,group==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & group==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & group==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_copies_4FR <- rbind(average_relativised_pool_counts_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_copies_4FR$population = factor(average_relativised_pool_counts_copies_4FR$population,levels=c("Pb","lines"))
average_relativised_pool_counts_copies_4FR$ratio = factor(average_relativised_pool_counts_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_copies_4FR$generation))
average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_copies_4FR
#write_tsv(average_relativised_pool_counts_copies_4FR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_4FR_",type,".txt"))


#Combined version:
Pb_lines_relativised_pool_counts_copies_4FR_ggplot <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR,ratio!="fourfold", generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_errorbar(aes(ymin=avg_Pb_relative_value-2*se_Pb_relative_value, ymax=avg_Pb_relative_value+2*se_Pb_relative_value, colour=generation, group=population), position=position_dodge(0.8),size=0.5,width=0.5) +
  geom_point(aes(colour=generation,shape=population),size=2,position=position_dodge(0.8)) + 
  #ggtitle("Proportion of reads at different NM") +
  ylab("Derived count relative to Pb-000") +
  scale_y_continuous(breaks = seq(0.6, 1.0, by = 0.1)) +
  scale_shape_manual(values=c(1,16)) +
  ylim(0.6,1.1) +
  ggtitle("Pools") +
  guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
Pb_lines_relativised_pool_counts_copies_4FR_ggplot
ggsave(paste0("Pb_lines_relativised_pool_counts_copies_4FR_",type,".pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

#####Low rec. gen0-140 pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
pool_counts <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.low_recombination.txt")) 

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$generation <- as.numeric(gsub("gen","",pool_counts$generation))
pool_counts <- pool_counts %>% arrange(generation)
pool_counts$generation = factor(pool_counts$generation)
pool_counts <- pool_counts %>% arrange(population,generation)

pool_counts_copies_4FR <- pool_counts %>% mutate(fourfold=fourfold_D/fourfold_D,tolerated=tolerated_D/fourfold_D,deleterious=deleterious_D/fourfold_D,LoF=LoF_D/fourfold_D) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_copies_4FR <- mutate(pool_counts_copies_4FR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_copies_4FR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_copies_4FR,population==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_copies_4FR <- rbind(average_relativised_pool_counts_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_copies_4FR$population = factor(average_relativised_pool_counts_copies_4FR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
average_relativised_pool_counts_copies_4FR$ratio = factor(average_relativised_pool_counts_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_copies_4FR$generation))
average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_copies_4FR
#write_tsv(average_relativised_pool_counts_copies_4FR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_4FR_",type,".txt"))


#Humberto version (Pb):
Pb_relativised_pool_counts_copies_4FR_ggplot_lowrec <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR,ratio!="fourfold",population=="Pb" & generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_point(aes(colour=generation),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to 4-fold syn. and Pb-000") +
  scale_y_continuous(breaks = seq(0.6, 1.0, by = 0.1)) +
  ylim(0.6,1.1) +
  ggtitle("Pools, Pb, low recombination quartile") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_relativised_pool_counts_copies_4FR_ggplot_lowrec
ggsave(paste0("Pb_relativised_pool_counts_copies_4FR_",type,"_Humberto_version.low_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
lines_relativised_pool_counts_copies_4FR_ggplot_lowrec <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR, ratio!="fourfold" & population!="Pb"), aes(ratio,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=generation)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to 4-fold syn. and Pb-000") +
  scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  ylim(0.45,1.15) +
  ggtitle("Pools, lines (average), low recombination quartile") +
  #guides(colour="none") +
  scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
lines_relativised_pool_counts_copies_4FR_ggplot_lowrec
ggsave(paste0("lines_relativised_pool_counts_copies_4FR_",type,"_Humberto_version.low_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

#####High rec. gen0-140 pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
pool_counts <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.high_recombination.txt")) 

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$generation <- as.numeric(gsub("gen","",pool_counts$generation))
pool_counts <- pool_counts %>% arrange(generation)
pool_counts$generation = factor(pool_counts$generation)
pool_counts <- pool_counts %>% arrange(population,generation)

pool_counts_copies_4FR <- pool_counts %>% mutate(fourfold=fourfold_D/fourfold_D,tolerated=tolerated_D/fourfold_D,deleterious=deleterious_D/fourfold_D,LoF=LoF_D/fourfold_D) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_copies_4FR <- mutate(pool_counts_copies_4FR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_copies_4FR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_copies_4FR,population==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_copies_4FR <- rbind(average_relativised_pool_counts_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_copies_4FR$population = factor(average_relativised_pool_counts_copies_4FR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
average_relativised_pool_counts_copies_4FR$ratio = factor(average_relativised_pool_counts_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_copies_4FR$generation))
average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_copies_4FR
#write_tsv(average_relativised_pool_counts_copies_4FR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_4FR_",type,".txt"))


#Humberto version (Pb):
Pb_relativised_pool_counts_copies_4FR_ggplot_highrec <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR,ratio!="fourfold",population=="Pb" & generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_point(aes(colour=generation),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to 4-fold syn. and Pb-000") +
  scale_y_continuous(breaks = seq(0.6, 1.0, by = 0.1)) +
  ylim(0.6,1.1) +
  ggtitle("Pools, Pb, high recombination quartile") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_relativised_pool_counts_copies_4FR_ggplot_highrec
ggsave(paste0("Pb_relativised_pool_counts_copies_4FR_",type,"_Humberto_version.high_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
lines_relativised_pool_counts_copies_4FR_ggplot_highrec <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR, ratio!="fourfold" & population!="Pb"), aes(ratio,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=generation)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to 4-fold syn. and Pb-000") +
  scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  ylim(0.45,1.15) +
  ggtitle("Pools, lines (average), high recombination quartile") +
  #guides(colour="none") +
  scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
lines_relativised_pool_counts_copies_4FR_ggplot_highrec
ggsave(paste0("lines_relativised_pool_counts_copies_4FR_",type,"_Humberto_version.high_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

#####All rec. gen0-140 pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))

pool_counts_low <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.low_recombination.txt"))
pool_counts_high <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.high_recombination.txt"))


####First, process the low recombination dataset####
pool_counts_low <- left_join(pool_counts_low,codes_dictionary,by=c("sample"="old"))
pool_counts_low$population = factor(pool_counts_low$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts_low$generation <- as.numeric(gsub("gen","",pool_counts_low$generation))
pool_counts_low <- pool_counts_low %>% arrange(generation)
pool_counts_low$generation = factor(pool_counts_low$generation)
pool_counts_low <- pool_counts_low %>% arrange(population,generation)

pool_counts_low_copies_4FR <- pool_counts_low %>% mutate(fourfold=fourfold_D/fourfold_D,tolerated=tolerated_D/fourfold_D,deleterious=deleterious_D/fourfold_D,LoF=LoF_D/fourfold_D) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_low_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_low_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_low_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_low_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_low_copies_4FR <- mutate(pool_counts_low_copies_4FR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_low_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"recombination"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_low_copies_4FR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_low_copies_4FR,population==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_low_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_low_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_low_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_low_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,"low",r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","recombination","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_low_copies_4FR <- rbind(average_relativised_pool_counts_low_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_low_copies_4FR$population = factor(average_relativised_pool_counts_low_copies_4FR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
average_relativised_pool_counts_low_copies_4FR$ratio = factor(average_relativised_pool_counts_low_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_low_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_low_copies_4FR$generation))
average_relativised_pool_counts_low_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_low_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_low_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_low_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_low_copies_4FR


####Next, process the high recombination dataset####
pool_counts_high <- left_join(pool_counts_high,codes_dictionary,by=c("sample"="old"))
pool_counts_high$population = factor(pool_counts_high$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts_high$generation <- as.numeric(gsub("gen","",pool_counts_high$generation))
pool_counts_high <- pool_counts_high %>% arrange(generation)
pool_counts_high$generation = factor(pool_counts_high$generation)
pool_counts_high <- pool_counts_high %>% arrange(population,generation)

pool_counts_high_copies_4FR <- pool_counts_high %>% mutate(fourfold=fourfold_D/fourfold_D,tolerated=tolerated_D/fourfold_D,deleterious=deleterious_D/fourfold_D,LoF=LoF_D/fourfold_D) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_high_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_high_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_high_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_high_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_high_copies_4FR <- mutate(pool_counts_high_copies_4FR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_high_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"recombination"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_high_copies_4FR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_high_copies_4FR,population==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_high_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_high_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_high_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_high_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,"high",r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","recombination","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_high_copies_4FR <- rbind(average_relativised_pool_counts_high_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_high_copies_4FR$population = factor(average_relativised_pool_counts_high_copies_4FR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
average_relativised_pool_counts_high_copies_4FR$ratio = factor(average_relativised_pool_counts_high_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_high_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_high_copies_4FR$generation))
average_relativised_pool_counts_high_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_high_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_high_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_high_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_high_copies_4FR


####Then combine both datasets####
average_relativised_pool_counts_combined_copies_4FR <- rbind(average_relativised_pool_counts_low_copies_4FR,average_relativised_pool_counts_high_copies_4FR)
average_relativised_pool_counts_combined_copies_4FR$recombination = factor(average_relativised_pool_counts_combined_copies_4FR$recombination,levels=c("low","high"))

####Then plot the data####
#Humberto version (Pb):
Pb_relativised_pool_counts_copies_4FR_ggplot_rec <- ggplot(data=filter(average_relativised_pool_counts_combined_copies_4FR,ratio!="fourfold",population=="Pb" & generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(ratio ~ .) +  
  geom_line(aes(colour=recombination, group=recombination),alpha=0.5,size=0.5) +
  geom_point(aes(colour=recombination),size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Derived count\n relative to Pb-000 4-fold syn.") +
  scale_y_continuous(breaks = seq(0.6, 1.0, by = 0.1)) +
  ylim(0.6,1.1) +
  ggtitle("Base population pools") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_relativised_pool_counts_copies_4FR_ggplot_rec
ggsave(paste0("Pb_relativised_pool_counts_copies_4FR_",type,"_Humberto_version.all_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

#Vertical version (Pb):
Pb_relativised_pool_counts_copies_4FR_ggplot_recvert <- ggplot(data=filter(average_relativised_pool_counts_combined_copies_4FR,ratio!="fourfold",population=="Pb" & generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +  
  geom_line(aes(colour=recombination, group=recombination),alpha=0.5,size=0.5) +
  geom_point(aes(colour=recombination),size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Derived count\n relative to Pb-000 4-fold syn.") +
  scale_y_continuous(breaks = seq(0.6, 1.0, by = 0.1)) +
  ylim(0.6,1.1) +
  ggtitle("Base population pools") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_relativised_pool_counts_copies_4FR_ggplot_recvert
ggsave(paste0("Pb_relativised_pool_counts_copies_4FR_",type,"_Humberto_version.all_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
lines_relativised_pool_counts_copies_4FR_ggplot_rec <- ggplot(data=filter(average_relativised_pool_counts_combined_copies_4FR, ratio!="fourfold" & population!="Pb"), aes(ratio,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(generation ~ ., labeller = labeller(generation=c("40" = "Generation 40", "140" = "Generation 140"))) +
  #labels <- c(Female = "Women", Male = "Men")
  #sp + facet_grid(. ~ sex, labeller=labeller(sex = labels))
  geom_boxplot(aes(colour=recombination)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Derived count\n relative to Pb-000 4-fold syn.") +
  scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  ylim(0.45,1.15) +
  ggtitle("Line pools") +
  #guides(colour="none") +
  #scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  labs(colour = "Recombination") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        axis.title.x=element_blank(),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_text()
  )
lines_relativised_pool_counts_copies_4FR_ggplot_rec
ggsave(paste0("lines_relativised_pool_counts_copies_4FR_",type,"_Humberto_version.all_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

#####All rec. gen0-140 pools Pb and line averages (combined version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))

pool_counts_low <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.low_recombination.txt"))
pool_counts_high <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.high_recombination.txt"))


####First, process the low recombination dataset####
pool_counts_low <- left_join(pool_counts_low,codes_dictionary,by=c("sample"="old"))
pool_counts_low$population = factor(pool_counts_low$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts_low$generation <- as.numeric(gsub("gen","",pool_counts_low$generation))
pool_counts_low <- pool_counts_low %>% arrange(generation)
pool_counts_low$generation = factor(pool_counts_low$generation)
pool_counts_low <- pool_counts_low %>% arrange(population,generation)

pool_counts_low_copies_4FR <- pool_counts_low %>% mutate(fourfold=fourfold_D/fourfold_D,tolerated=tolerated_D/fourfold_D,deleterious=deleterious_D/fourfold_D,LoF=LoF_D/fourfold_D) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_low_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_low_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_low_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_low_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_low_copies_4FR <- mutate(pool_counts_low_copies_4FR, Pb_relative_value=value/r_average_vector, group=ifelse(population=="Pb","Pb","lines"))
relativised_pool_counts_low_copies_4FR$group = factor(relativised_pool_counts_low_copies_4FR$group,levels=c("Pb","lines"))

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_low_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"recombination"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_low_copies_4FR$group)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_low_copies_4FR,group==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_low_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_low_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_low_copies_4FR,ratio==r & generation==g & group==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_low_copies_4FR,ratio==r & generation==g & group==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,"low",r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","recombination","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_low_copies_4FR <- rbind(average_relativised_pool_counts_low_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_low_copies_4FR$population = factor(average_relativised_pool_counts_low_copies_4FR$population,levels=c("Pb","lines"))
average_relativised_pool_counts_low_copies_4FR$ratio = factor(average_relativised_pool_counts_low_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_low_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_low_copies_4FR$generation))
average_relativised_pool_counts_low_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_low_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_low_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_low_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_low_copies_4FR


####Next, process the high recombination dataset####
pool_counts_high <- left_join(pool_counts_high,codes_dictionary,by=c("sample"="old"))
pool_counts_high$population = factor(pool_counts_high$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts_high$generation <- as.numeric(gsub("gen","",pool_counts_high$generation))
pool_counts_high <- pool_counts_high %>% arrange(generation)
pool_counts_high$generation = factor(pool_counts_high$generation)
pool_counts_high <- pool_counts_high %>% arrange(population,generation)

pool_counts_high_copies_4FR <- pool_counts_high %>% mutate(fourfold=fourfold_D/fourfold_D,tolerated=tolerated_D/fourfold_D,deleterious=deleterious_D/fourfold_D,LoF=LoF_D/fourfold_D) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_high_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_high_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_high_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_high_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_high_copies_4FR <- mutate(pool_counts_high_copies_4FR, Pb_relative_value=value/r_average_vector, group=ifelse(population=="Pb","Pb","lines"))
relativised_pool_counts_high_copies_4FR$group = factor(relativised_pool_counts_high_copies_4FR$group,levels=c("Pb","lines"))

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_high_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"recombination"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_high_copies_4FR$group)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_high_copies_4FR,group==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_high_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_high_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_high_copies_4FR,ratio==r & generation==g & group==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_high_copies_4FR,ratio==r & generation==g & group==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,"high",r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","recombination","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_high_copies_4FR <- rbind(average_relativised_pool_counts_high_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_high_copies_4FR$population = factor(average_relativised_pool_counts_high_copies_4FR$population,levels=c("Pb","lines"))
average_relativised_pool_counts_high_copies_4FR$ratio = factor(average_relativised_pool_counts_high_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_high_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_high_copies_4FR$generation))
average_relativised_pool_counts_high_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_high_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_high_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_high_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_high_copies_4FR


####Then combine both datasets####
average_relativised_pool_counts_combined_copies_4FR <- rbind(average_relativised_pool_counts_low_copies_4FR,average_relativised_pool_counts_high_copies_4FR)
average_relativised_pool_counts_combined_copies_4FR$recombination = factor(average_relativised_pool_counts_combined_copies_4FR$recombination,levels=c("low","high"))


####Then plot the data####
#Combined version:
Pb_lines_relativised_pool_counts_copies_4FR_ggplot_recvert <- ggplot(data=filter(average_relativised_pool_counts_combined_copies_4FR,ratio!="fourfold", generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +  
  #geom_line(aes(colour=recombination, group=interaction(recombination, population)),alpha=0.5,size=0.5) +
  geom_point(aes(colour=recombination,shape=population),size=2,position=position_dodge(0.4)) +
  geom_errorbar(aes(ymin=avg_Pb_relative_value-2*se_Pb_relative_value, ymax=avg_Pb_relative_value+2*se_Pb_relative_value, colour=recombination,group=interaction(population,recombination)), position=position_dodge(0.4),size=0.5,width=0.5) +
  ylab("Derived count\n relative to Pb-000 4-fold syn.") +
  scale_y_continuous(breaks = seq(0.5, 1.0, by = 0.1)) +
  scale_shape_manual(values=c(1,16)) +
  ylim(0.4,1.1) +
  ggtitle("Pools") +
  #guides(colour="none") +
  labs(colour = "Recombination", shape = "Population") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_text()
  )
Pb_lines_relativised_pool_counts_copies_4FR_ggplot_recvert
ggsave(paste0("Pb_lines_relativised_pool_counts_copies_4FR_",type,".all_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

####Raw:
#####Pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)

type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
pool_counts_gen0_40 <- read_tsv(paste0(wd_path,"counts_pool_gen0-40_",type,"_summary.txt")) %>% mutate(gen=ifelse(sample=="LBT0",000,ifelse(sample=="c1",005,ifelse(sample=="BT20",020,ifelse(sample=="BT30",030,040)))))
pool_counts_gen140 <- read_tsv(paste0(wd_path,"counts_pool_gen140_",type,"_summary.txt")) %>% mutate(gen="140" %>% as.double)

pool_counts <- rbind(pool_counts_gen0_40,pool_counts_gen140)

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$gen = factor(pool_counts$gen)
pool_counts <- pool_counts %>% arrange(population,gen)

pool_counts_copies_4FR <- pool_counts %>% mutate(fourfold=fourfold_D/fourfold_D,missense=missense_D/fourfold_D,tolerated=tolerated_D/fourfold_D,deleterious=deleterious_D/fourfold_D,LoF=(stopgain_D+stoploss_D)/fourfold_D) %>% select(c(1,29:35)) %>% gather(ratio,value,-sample,-gen,-population,factor_key=T)
pool_counts_copies_4FR


pool_counts_copies_4FR <- pool_counts_copies_4FR %>% mutate(group=ifelse(population=="Pb","Pb","lines"))
pool_counts_copies_4FR$population = factor(pool_counts_copies_4FR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts_copies_4FR$ratio = factor(pool_counts_copies_4FR$ratio,levels=c("fourfold","missense","tolerated","deleterious","LoF"))
#write_tsv(average_relativised_pool_counts_copies_synR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_synR_",type,".txt"))

#Humberto version (Pb):
Pb_pool_counts_copies_4FR_ggplot <- ggplot(data=filter(pool_counts_copies_4FR,ratio!="fourfold", ratio!="missense", population=="Pb" & gen!=5), aes(gen,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_wrap(. ~ ratio, scales = "free", nrow=1) +
  geom_point(aes(colour=gen),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to 4-fold syn.") +
  #scale_y_continuous(breaks = seq(0.6, 1.2, by = 0.2)) +
  #ylim(0.5,1.4) +
  ggtitle("Pools, Pb") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_pool_counts_copies_4FR_ggplot
ggsave(paste0("Pb_pool_counts_copies_4FR_",type,"_Humberto_version.pdf"), width=17.5, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
line_pool_counts_copies_4FR_ggplot <- ggplot(data=filter(pool_counts_copies_4FR,ratio!="fourfold", ratio!="missense" & group=="lines"), aes(ratio,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=gen)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to 4-fold syn.") +
  #scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  #ylim(0.7,1.05) +
  ggtitle("Pools, lines (average)") +
  scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
line_pool_counts_copies_4FR_ggplot
ggsave(paste0("lines_pool_counts_copies_4FR_",type,"_Humberto_version.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

#####Gen0-140 pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)

type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
pool_counts <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.txt")) 

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$generation <- as.numeric(gsub("gen","",pool_counts$generation))
pool_counts <- pool_counts %>% arrange(generation)
pool_counts$generation = factor(pool_counts$generation)
pool_counts <- pool_counts %>% arrange(population,generation)

pool_counts_copies_4FR <- pool_counts %>% mutate(fourfold=fourfold_D/fourfold_D,tolerated=tolerated_D/fourfold_D,deleterious=deleterious_D/fourfold_D,LoF=LoF_D/fourfold_D) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_copies_4FR

#Humberto version (Pb):
Pb_raw_pool_counts_copies_4FR_ggplot <- ggplot(data=filter(pool_counts_copies_4FR,ratio!="fourfold", population=="Pb" & generation!=5), aes(generation,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_wrap(. ~ ratio, scales = "free", nrow=1) +
  geom_point(aes(colour=generation),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to 4-fold syn.") +
  #scale_y_continuous(breaks = seq(0.6, 1.2, by = 0.2)) +
  #ylim(0.5,1.4) +
  ggtitle("Pools, Pb") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_raw_pool_counts_copies_4FR_ggplot
ggsave(paste0("Pb_pool_counts_copies_4FR_",type,"_Humberto_version.pdf"), width=17.5, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
lines_raw_pool_counts_copies_4FR_ggplot <- ggplot(data=filter(pool_counts_copies_4FR,ratio!="fourfold", population!="Pb"), aes(ratio,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=generation)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to 4-fold syn.") +
  #scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  #ylim(0.7,1.05) +
  ggtitle("Pools, lines (average)") +
  scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
lines_raw_pool_counts_copies_4FR_ggplot
ggsave(paste0("lines_pool_counts_copies_4FR_",type,"_Humberto_version.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```


###Relative to introns:
####Relative to Pb:
#####Individuals:
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "autosomes" #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
individual_counts <- read_tsv(paste0(wd_path,"counts_individual_",type,"_summary.txt"))
individual_counts <- mutate(individual_counts, population=gsub("-","",substr(sample,1,3)))
individual_counts$population = factor(individual_counts$population)
individual_counts$population <- gsub('Pb', 'Pb-140', individual_counts$population)


individual_counts_copies_intrR <- individual_counts %>% mutate(synonymous=synonymous_D/intronic_D,missense=missense_D/intronic_D,LoF=(stopgain_D+stoploss_D)/intronic_D) %>% select(c(1,14:17)) %>% gather(ratio,value,-sample,-population,factor_key=T)
individual_counts_copies_intrR

r_average_vector <- c()
for (r in unique(individual_counts_copies_intrR$ratio)) {
  print(r)
  r_average <- filter(individual_counts_copies_intrR,r==ratio & population=="Pb-140") %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(individual_counts_copies_intrR,r==ratio))))
}
print(r_average_vector)

relativised_individual_counts_copies_intrR <- mutate(individual_counts_copies_intrR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_individual_counts_copies_intrR <- data_frame("population"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_individual_counts_copies_intrR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  #species <- filter(relativised_individual_counts_copies_intrR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
  for (r in unique(relativised_individual_counts_copies_intrR$ratio)) {
    print(r)
    pop_mean <- filter(relativised_individual_counts_copies_intrR,ratio==r & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
    #print(paste0(pop," feature ",r," average is ",pop_mean))
    pop_se <- filter(relativised_individual_counts_copies_intrR,ratio==r & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
    #print(paste0(pop," feature ",r," std error is ",pop_se))
    row_data <- cbind(pop,r,pop_mean,pop_se)
    colnames(row_data) <- c("population","ratio","avg_Pb_relative_value","se_Pb_relative_value")
    average_relativised_individual_counts_copies_intrR <- rbind(average_relativised_individual_counts_copies_intrR,row_data,stringsAsFactors=F)
  }
}
average_relativised_individual_counts_copies_intrR$population = factor(average_relativised_individual_counts_copies_intrR$population)
average_relativised_individual_counts_copies_intrR$ratio = factor(average_relativised_individual_counts_copies_intrR$ratio,levels=c("synonymous","missense","LoF"))

average_relativised_individual_counts_copies_intrR$avg_Pb_relative_value <- as.numeric(average_relativised_individual_counts_copies_intrR$avg_Pb_relative_value)
average_relativised_individual_counts_copies_intrR$se_Pb_relative_value <- as.numeric(average_relativised_individual_counts_copies_intrR$se_Pb_relative_value)
average_relativised_individual_counts_copies_intrR
write_tsv(average_relativised_individual_counts_copies_intrR,paste0(wd_path,"/pop_average_relativised_individual_counts_copies_intrR_",type,".txt"))

#Plot these means and standard errors:
average_relativised_individual_counts_copies_intrR_ggplot <- ggplot(data=average_relativised_individual_counts_copies_intrR, aes(population,avg_Pb_relative_value,colour=population)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_point() +
  geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count \n relative to introns and Pb-140") +
  scale_y_continuous(breaks = seq(0.8, 1.6, by = 0.2)) +
  ylim(0.8,1.6) +
  ggtitle(paste0("Individuals ",type)) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
average_relativised_individual_counts_copies_intrR_ggplot
ggsave(paste0("pop_average_relativised_individual_counts_copies_intrR_",type,".pdf"), width=21, height=12, units="cm", device="pdf", path=wd_path)

```

#####Pools:
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "gen0-40" #gen0-40 #gen140
subtype <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
pool_counts <- read_tsv(paste0(wd_path,"counts_pool_",type,"_",subtype,"_summary.txt"))

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old")) %>% arrange(population)
pool_counts$population = factor(pool_counts$population)

pool_counts_copies_intrR <- pool_counts %>% mutate(synonymous=synonymous_D/intronic_D,missense=missense_D/intronic_D,LoF=(stopgain_D+stoploss_D)/intronic_D) %>% select(c(1,20:23)) %>% gather(ratio,value,-sample,-population,factor_key=T)
pool_counts_copies_intrR

r_average_vector <- c()
for (r in unique(pool_counts_copies_intrR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_copies_intrR,r==ratio & population=="Pb-000") %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_copies_intrR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_copies_intrR <- mutate(pool_counts_copies_intrR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_copies_intrR <- data_frame("population"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_copies_intrR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  #species <- filter(relativised_pool_counts_copies_intrR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
  for (r in unique(relativised_pool_counts_copies_intrR$ratio)) {
    print(r)
    pop_mean <- filter(relativised_pool_counts_copies_intrR,ratio==r & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
    #print(paste0(pop," feature ",r," average is ",pop_mean))
    pop_se <- filter(relativised_pool_counts_copies_intrR,ratio==r & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
    #print(paste0(pop," feature ",r," std error is ",pop_se))
    row_data <- cbind(pop,r,pop_mean,pop_se)
    colnames(row_data) <- c("population","ratio","avg_Pb_relative_value","se_Pb_relative_value")
    average_relativised_pool_counts_copies_intrR <- rbind(average_relativised_pool_counts_copies_intrR,row_data,stringsAsFactors=F)
  }
}
average_relativised_pool_counts_copies_intrR$population = factor(average_relativised_pool_counts_copies_intrR$population)
average_relativised_pool_counts_copies_intrR$ratio = factor(average_relativised_pool_counts_copies_intrR$ratio,levels=c("synonymous","missense","LoF"))

average_relativised_pool_counts_copies_intrR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_intrR$avg_Pb_relative_value)
average_relativised_pool_counts_copies_intrR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_intrR$se_Pb_relative_value)
average_relativised_pool_counts_copies_intrR
write_tsv(average_relativised_pool_counts_copies_intrR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_intrR_",type,"_",subtype,".txt"))

#Plot these means and standard errors:
average_relativised_pool_counts_copies_intrR_ggplot <- ggplot(data=average_relativised_pool_counts_copies_intrR, aes(population,avg_Pb_relative_value,colour=population)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_point() +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average derived count\n relative to introns and Pb-000") +
  scale_y_continuous(breaks = seq(0.6, 1.6, by = 0.2)) +
  ylim(0.7,1.7) +
  ggtitle(paste0("Pools ",type," ",subtype)) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=8),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
average_relativised_pool_counts_copies_intrR_ggplot
ggsave(paste0("pop_average_relativised_pool_counts_copies_intrR_",type,"_",subtype,".pdf"), width=21, height=12, units="cm", device="pdf", path=wd_path)

```

##Site ratios:
###Relative to synonymous:
####Raw:
#####Individuals:
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "autosomes" #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
individual_counts <- read_tsv(paste0(wd_path,"counts_individual_",type,"_summary.txt"))
individual_counts <- mutate(individual_counts, population=gsub("-","",substr(sample,1,3)))
individual_counts$population = factor(individual_counts$population)
individual_counts$population <- gsub('Pb', 'Pb-140', individual_counts$population)


individual_counts_sites_synR <- individual_counts %>% mutate(synonymous=synonymous_V/synonymous_V,missense=missense_V/synonymous_V,LoF=(stopgain_V+stoploss_V)/synonymous_V) %>% select(c(1,14:17)) %>% gather(ratio,value,-sample,-population,factor_key=T)
individual_counts_sites_synR

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_individual_counts_sites_synR <- data_frame("population"=character(0),"ratio"=character(0),"avg_value"=character(0),"se_value"=character(0)) #next, create the empty dataframe

for (pop in unique(individual_counts_sites_synR$population)) { #then loop over each population and feature to get the mean and standard error, and feed the dataframe
  print(pop)
  #species <- filter(relativised_individual_counts_copies_synR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
  for (r in unique(individual_counts_sites_synR$ratio)) {
    print(r)
    pop_mean <- filter(individual_counts_sites_synR,ratio==r & population==pop) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
    #print(paste0(pop," feature ",r," average is ",pop_mean))
    pop_se <- filter(individual_counts_sites_synR,ratio==r & population==pop) %>% select(value) %>% unlist(.,use.names=F) %>% se()
    #print(paste0(pop," feature ",r," std error is ",pop_se))
    row_data <- cbind(pop,r,pop_mean,pop_se)
    colnames(row_data) <- c("population","ratio","avg_value","se_value")
    average_individual_counts_sites_synR <- rbind(average_individual_counts_sites_synR,row_data,stringsAsFactors=F)
  }
}
average_individual_counts_sites_synR$population = factor(average_individual_counts_sites_synR$population)
average_individual_counts_sites_synR$ratio = factor(average_individual_counts_sites_synR$ratio,levels=c("synonymous","missense","LoF"))

average_individual_counts_sites_synR$avg_value <- as.numeric(average_individual_counts_sites_synR$avg_value)
average_individual_counts_sites_synR$se_value <- as.numeric(average_individual_counts_sites_synR$se_value)
average_individual_counts_sites_synR

#Plot these means and standard errors:
average_individual_counts_sites_synR_ggplot <- ggplot(data=filter(average_individual_counts_sites_synR,ratio!="synonymous"), aes(population,avg_value,colour=population)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_wrap(. ~ ratio, scales = "free") +
  geom_point() +
  geom_errorbar(aes(ymin=avg_value-se_value, ymax=avg_value+se_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average number of sites with \n derived copies relative to synonymous") +
  #scale_y_continuous(breaks = seq(0.8, 1.6, by = 0.2)) +
  #ylim(0.8,1.6) +
  ggtitle(paste0("Individuals ",type)) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
average_individual_counts_sites_synR_ggplot
ggsave(paste0("pop_average_individual_counts_sites_synR_",type,".pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

###Relative to 4-fold:
####Relative to Pb:
#####Gen0-140 pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
pool_counts <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.txt")) 

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$generation <- as.numeric(gsub("gen","",pool_counts$generation))
pool_counts <- pool_counts %>% arrange(generation)
pool_counts$generation = factor(pool_counts$generation)
pool_counts <- pool_counts %>% arrange(population,generation)

pool_counts_copies_4FR <- pool_counts %>% mutate(fourfold=fourfold_V/fourfold_V,tolerated=tolerated_V/fourfold_V,deleterious=deleterious_V/fourfold_V,LoF=LoF_V/fourfold_V) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_copies_4FR <- mutate(pool_counts_copies_4FR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_copies_4FR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_copies_4FR,population==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_copies_4FR <- rbind(average_relativised_pool_counts_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_copies_4FR$population = factor(average_relativised_pool_counts_copies_4FR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
average_relativised_pool_counts_copies_4FR$ratio = factor(average_relativised_pool_counts_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_copies_4FR$generation))
average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_copies_4FR
#write_tsv(average_relativised_pool_counts_copies_4FR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_4FR_",type,".txt"))


#Humberto version (Pb):
Pb_relativised_pool_counts_sites_4FR_ggplot <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR,ratio!="fourfold",population=="Pb" & generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_point(aes(colour=generation),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Number of sites\n relative to 4-fold syn. and Pb-000") +
  scale_y_continuous(breaks = seq(0.7, 1, by = 0.1)) +
  ylim(0.6,1.1) +
  ggtitle("Base population pools") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_relativised_pool_counts_sites_4FR_ggplot
ggsave(paste0("Pb_relativised_pool_counts_sites_4FR_",type,"_Humberto_version.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
lines_relativised_pool_counts_sites_4FR_ggplot <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR, ratio!="fourfold" & population!="Pb"), aes(ratio,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=generation)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Number of sites\n relative to 4-fold syn. and Pb-000") +
  scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  ylim(0.4,1.05) +
  ggtitle("Line pools") +
  #guides(colour="none") +
  scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  labs(colour = "Generation") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        axis.title.x=element_blank(),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_text()
  )
lines_relativised_pool_counts_sites_4FR_ggplot
ggsave(paste0("lines_relativised_pool_counts_sites_4FR_",type,"_Humberto_version.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

#####Gen0-140 pools Pb and line averages (combined):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
pool_counts <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.txt")) 

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$generation <- as.numeric(gsub("gen","",pool_counts$generation))
pool_counts <- pool_counts %>% arrange(generation)
pool_counts$generation = factor(pool_counts$generation)
pool_counts <- pool_counts %>% arrange(population,generation)

pool_counts_copies_4FR <- pool_counts %>% mutate(fourfold=fourfold_V/fourfold_V,tolerated=tolerated_V/fourfold_V,deleterious=deleterious_V/fourfold_V,LoF=LoF_V/fourfold_V) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_copies_4FR <- mutate(pool_counts_copies_4FR, Pb_relative_value=value/r_average_vector, group=ifelse(population=="Pb","Pb","lines"))
relativised_pool_counts_copies_4FR$group = factor(relativised_pool_counts_copies_4FR$group,levels=c("Pb","lines"))

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_copies_4FR$group)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_copies_4FR,group==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & group==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & group==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_copies_4FR <- rbind(average_relativised_pool_counts_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_copies_4FR$population = factor(average_relativised_pool_counts_copies_4FR$population,levels=c("Pb","lines"))
average_relativised_pool_counts_copies_4FR$ratio = factor(average_relativised_pool_counts_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_copies_4FR$generation))
average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_copies_4FR
#write_tsv(average_relativised_pool_counts_copies_4FR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_4FR_",type,".txt"))


#Combined version (Pb):
Pb_lines_relativised_pool_counts_sites_4FR_ggplot <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR,ratio!="fourfold", generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_errorbar(aes(ymin=avg_Pb_relative_value-2*se_Pb_relative_value, ymax=avg_Pb_relative_value+2*se_Pb_relative_value, colour=generation,group=population), position=position_dodge(0.8),size=0.5,width=0.5) +
  geom_point(aes(colour=generation,shape=population),size=2,position=position_dodge(0.8)) + 
  #ggtitle("Proportion of reads at different NM") +
  ylab("Number of sites relative to Pb-000") +
  scale_y_continuous(breaks = seq(0.5, 1, by = 0.1)) +
  scale_shape_manual(values=c(1,16)) +
  ylim(0.5,1.1) +
  ggtitle("Pools") +
  guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
Pb_lines_relativised_pool_counts_sites_4FR_ggplot
ggsave(paste0("Pb_lines_relativised_pool_counts_sites_4FR_",type,".pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)


```

#####Low rec. gen0-140 pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
pool_counts <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.low_recombination.txt")) 

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$generation <- as.numeric(gsub("gen","",pool_counts$generation))
pool_counts <- pool_counts %>% arrange(generation)
pool_counts$generation = factor(pool_counts$generation)
pool_counts <- pool_counts %>% arrange(population,generation)

pool_counts_copies_4FR <- pool_counts %>% mutate(fourfold=fourfold_V/fourfold_V,tolerated=tolerated_V/fourfold_V,deleterious=deleterious_V/fourfold_V,LoF=LoF_V/fourfold_V) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_copies_4FR <- mutate(pool_counts_copies_4FR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_copies_4FR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_copies_4FR,population==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_copies_4FR <- rbind(average_relativised_pool_counts_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_copies_4FR$population = factor(average_relativised_pool_counts_copies_4FR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
average_relativised_pool_counts_copies_4FR$ratio = factor(average_relativised_pool_counts_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_copies_4FR$generation))
average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_copies_4FR
#write_tsv(average_relativised_pool_counts_copies_4FR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_4FR_",type,".txt"))


#Humberto version (Pb):
Pb_relativised_pool_counts_sites_4FR_ggplot_lowrec <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR,ratio!="fourfold",population=="Pb" & generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_point(aes(colour=generation),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average number of sites\n relative to 4-fold syn. and Pb-000") +
  scale_y_continuous(breaks = seq(0.7, 1, by = 0.1)) +
  ylim(0.6,1.1) +
  ggtitle("Pools, Pb, low recombination quartile") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_relativised_pool_counts_sites_4FR_ggplot_lowrec
ggsave(paste0("Pb_relativised_pool_counts_variants_4FR_",type,"_Humberto_version.low_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
lines_relativised_pool_counts_sites_4FR_ggplot_lowrec <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR, ratio!="fourfold" & population!="Pb"), aes(ratio,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=generation)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average number of sites\n relative to 4-fold syn. and Pb-000") +
  scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  ylim(0.3,1.05) +
  ggtitle("Pools, lines (average), low recombination quartile") +
  #guides(colour="none") +
  scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
lines_relativised_pool_counts_sites_4FR_ggplot_lowrec
ggsave(paste0("lines_relativised_pool_counts_variants_4FR_",type,"_Humberto_version.low_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

#####High rec. gen0-140 pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
pool_counts <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.high_recombination.txt")) 

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$generation <- as.numeric(gsub("gen","",pool_counts$generation))
pool_counts <- pool_counts %>% arrange(generation)
pool_counts$generation = factor(pool_counts$generation)
pool_counts <- pool_counts %>% arrange(population,generation)

pool_counts_copies_4FR <- pool_counts %>% mutate(fourfold=fourfold_V/fourfold_V,tolerated=tolerated_V/fourfold_V,deleterious=deleterious_V/fourfold_V,LoF=LoF_V/fourfold_V) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_copies_4FR <- mutate(pool_counts_copies_4FR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_copies_4FR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_copies_4FR,population==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_copies_4FR <- rbind(average_relativised_pool_counts_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_copies_4FR$population = factor(average_relativised_pool_counts_copies_4FR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
average_relativised_pool_counts_copies_4FR$ratio = factor(average_relativised_pool_counts_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_copies_4FR$generation))
average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_copies_4FR
#write_tsv(average_relativised_pool_counts_copies_4FR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_4FR_",type,".txt"))


#Humberto version (Pb):
Pb_relativised_pool_counts_sites_4FR_ggplot_highrec <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR,ratio!="fourfold",population=="Pb" & generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_point(aes(colour=generation),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average number of sites\n relative to 4-fold syn. and Pb-000") +
  scale_y_continuous(breaks = seq(0.7, 1, by = 0.1)) +
  ylim(0.6,1.1) +
  ggtitle("Pools, Pb, high recombination quartile") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_relativised_pool_counts_sites_4FR_ggplot_highrec
ggsave(paste0("Pb_relativised_pool_counts_variants_4FR_",type,"_Humberto_version.high_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
lines_relativised_pool_counts_sites_4FR_ggplot_highrec <- ggplot(data=filter(average_relativised_pool_counts_copies_4FR, ratio!="fourfold" & population!="Pb"), aes(ratio,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=generation)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average number of sites\n relative to 4-fold syn. and Pb-000") +
  scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  ylim(0.3,1.05) +
  ggtitle("Pools, lines (average), high recombination quartile") +
  #guides(colour="none") +
  scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
lines_relativised_pool_counts_sites_4FR_ggplot_highrec
ggsave(paste0("lines_relativised_pool_counts_variants_4FR_",type,"_Humberto_version.high_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

#####All rec. gen0-140 pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))

pool_counts_low <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.low_recombination.txt"))
pool_counts_high <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.high_recombination.txt"))


####First, process the low recombination dataset####
pool_counts_low <- left_join(pool_counts_low,codes_dictionary,by=c("sample"="old"))
pool_counts_low$population = factor(pool_counts_low$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts_low$generation <- as.numeric(gsub("gen","",pool_counts_low$generation))
pool_counts_low <- pool_counts_low %>% arrange(generation)
pool_counts_low$generation = factor(pool_counts_low$generation)
pool_counts_low <- pool_counts_low %>% arrange(population,generation)

pool_counts_low_copies_4FR <- pool_counts_low %>% mutate(fourfold=fourfold_V/fourfold_V,tolerated=tolerated_V/fourfold_V,deleterious=deleterious_V/fourfold_V,LoF=LoF_V/fourfold_V) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_low_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_low_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_low_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_low_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_low_copies_4FR <- mutate(pool_counts_low_copies_4FR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_low_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"recombination"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_low_copies_4FR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_low_copies_4FR,population==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_low_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_low_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_low_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_low_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,"low",r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","recombination","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_low_copies_4FR <- rbind(average_relativised_pool_counts_low_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_low_copies_4FR$population = factor(average_relativised_pool_counts_low_copies_4FR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
average_relativised_pool_counts_low_copies_4FR$ratio = factor(average_relativised_pool_counts_low_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_low_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_low_copies_4FR$generation))
average_relativised_pool_counts_low_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_low_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_low_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_low_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_low_copies_4FR


####Next, process the high recombination dataset####
pool_counts_high <- left_join(pool_counts_high,codes_dictionary,by=c("sample"="old"))
pool_counts_high$population = factor(pool_counts_high$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts_high$generation <- as.numeric(gsub("gen","",pool_counts_high$generation))
pool_counts_high <- pool_counts_high %>% arrange(generation)
pool_counts_high$generation = factor(pool_counts_high$generation)
pool_counts_high <- pool_counts_high %>% arrange(population,generation)

pool_counts_high_copies_4FR <- pool_counts_high %>% mutate(fourfold=fourfold_V/fourfold_V,tolerated=tolerated_V/fourfold_V,deleterious=deleterious_V/fourfold_V,LoF=LoF_V/fourfold_V) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_high_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_high_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_high_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_high_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_high_copies_4FR <- mutate(pool_counts_high_copies_4FR, Pb_relative_value=value/r_average_vector)

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_high_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"recombination"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_high_copies_4FR$population)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_high_copies_4FR,population==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_high_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_high_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_high_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_high_copies_4FR,ratio==r & generation==g & population==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,"high",r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","recombination","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_high_copies_4FR <- rbind(average_relativised_pool_counts_high_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_high_copies_4FR$population = factor(average_relativised_pool_counts_high_copies_4FR$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
average_relativised_pool_counts_high_copies_4FR$ratio = factor(average_relativised_pool_counts_high_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_high_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_high_copies_4FR$generation))
average_relativised_pool_counts_high_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_high_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_high_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_high_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_high_copies_4FR


####Then combine both datasets####
average_relativised_pool_counts_combined_copies_4FR <- rbind(average_relativised_pool_counts_low_copies_4FR,average_relativised_pool_counts_high_copies_4FR)
average_relativised_pool_counts_combined_copies_4FR$recombination = factor(average_relativised_pool_counts_combined_copies_4FR$recombination,levels=c("low","high"))

####Then plot the data####
#Humberto version (Pb):
Pb_relativised_pool_counts_sites_4FR_ggplot_rec <- ggplot(data=filter(average_relativised_pool_counts_combined_copies_4FR,ratio!="fourfold",population=="Pb" & generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(ratio ~ .) +
  geom_line(aes(colour=recombination, group=recombination),alpha=0.5,size=0.5) +
  geom_point(aes(colour=recombination),size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Number of sites\n relative to Pb-000 4-fold syn.") +
  scale_y_continuous(breaks = seq(0.7, 1, by = 0.1)) +
  ylim(0.6,1.1) +
  ggtitle("Base population pools") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_relativised_pool_counts_sites_4FR_ggplot_rec
ggsave(paste0("Pb_relativised_pool_counts_sites_4FR_",type,"_Humberto_version.all_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

#Vertical version (Pb):
Pb_relativised_pool_counts_sites_4FR_ggplot_recvert <- ggplot(data=filter(average_relativised_pool_counts_combined_copies_4FR,ratio!="fourfold",population=="Pb" & generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  geom_line(aes(colour=recombination, group=recombination),alpha=0.5,size=0.5) +
  geom_point(aes(colour=recombination),size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Number of sites\n relative to Pb-000 4-fold syn.") +
  scale_y_continuous(breaks = seq(0.7, 1, by = 0.1)) +
  ylim(0.6,1.1) +
  ggtitle("Base population pools") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_relativised_pool_counts_sites_4FR_ggplot_recvert
ggsave(paste0("Pb_relativised_pool_counts_sites_4FR_",type,"_Humberto_version.all_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
lines_relativised_pool_counts_sites_4FR_ggplot_rec <- ggplot(data=filter(average_relativised_pool_counts_combined_copies_4FR, ratio!="fourfold" & population!="Pb"), aes(ratio,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(generation ~ ., labeller = labeller(generation=c("40" = "Generation 40", "140" = "Generation 140"))) +
  geom_boxplot(aes(colour=recombination)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Number of sites\n relative to Pb-000 4-fold syn.") +
  scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  ylim(0.3,1.05) +
  ggtitle("Line pools") +
  #guides(colour="none") +
  #scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  labs(colour = "Recombination") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        axis.title.x=element_blank(),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_text()
  )
lines_relativised_pool_counts_sites_4FR_ggplot_rec
ggsave(paste0("lines_relativised_pool_counts_sites_4FR_",type,"_Humberto_version.all_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

#####All rec. gen0-140 pools Pb and line averages (combined version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))

pool_counts_low <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.low_recombination.txt"))
pool_counts_high <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.high_recombination.txt"))


####First, process the low recombination dataset####
pool_counts_low <- left_join(pool_counts_low,codes_dictionary,by=c("sample"="old"))
pool_counts_low$population = factor(pool_counts_low$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts_low$generation <- as.numeric(gsub("gen","",pool_counts_low$generation))
pool_counts_low <- pool_counts_low %>% arrange(generation)
pool_counts_low$generation = factor(pool_counts_low$generation)
pool_counts_low <- pool_counts_low %>% arrange(population,generation)

pool_counts_low_copies_4FR <- pool_counts_low %>% mutate(fourfold=fourfold_V/fourfold_V,tolerated=tolerated_V/fourfold_V,deleterious=deleterious_V/fourfold_V,LoF=LoF_V/fourfold_V) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_low_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_low_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_low_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_low_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_low_copies_4FR <- mutate(pool_counts_low_copies_4FR, Pb_relative_value=value/r_average_vector, group=ifelse(population=="Pb","Pb","lines"))
relativised_pool_counts_low_copies_4FR$group = factor(relativised_pool_counts_low_copies_4FR$group,levels=c("Pb","lines"))

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_low_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"recombination"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_low_copies_4FR$group)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_low_copies_4FR,group==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_low_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_low_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_low_copies_4FR,ratio==r & generation==g & group==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_low_copies_4FR,ratio==r & generation==g & group==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,"low",r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","recombination","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_low_copies_4FR <- rbind(average_relativised_pool_counts_low_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_low_copies_4FR$population = factor(average_relativised_pool_counts_low_copies_4FR$population,levels=c("Pb","lines"))
average_relativised_pool_counts_low_copies_4FR$ratio = factor(average_relativised_pool_counts_low_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_low_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_low_copies_4FR$generation))
average_relativised_pool_counts_low_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_low_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_low_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_low_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_low_copies_4FR


####Next, process the high recombination dataset####
pool_counts_high <- left_join(pool_counts_high,codes_dictionary,by=c("sample"="old"))
pool_counts_high$population = factor(pool_counts_high$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts_high$generation <- as.numeric(gsub("gen","",pool_counts_high$generation))
pool_counts_high <- pool_counts_high %>% arrange(generation)
pool_counts_high$generation = factor(pool_counts_high$generation)
pool_counts_high <- pool_counts_high %>% arrange(population,generation)

pool_counts_high_copies_4FR <- pool_counts_high %>% mutate(fourfold=fourfold_V/fourfold_V,tolerated=tolerated_V/fourfold_V,deleterious=deleterious_V/fourfold_V,LoF=LoF_V/fourfold_V) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_high_copies_4FR

r_average_vector <- c()
for (r in unique(pool_counts_high_copies_4FR$ratio)) {
  print(r)
  r_average <- filter(pool_counts_high_copies_4FR,r==ratio & population=="Pb" & generation==0) %>% select(value) %>% unlist(.,use.names=F) %>% mean()
  r_average_vector <- c(r_average_vector,rep(r_average,nrow(filter(pool_counts_high_copies_4FR,r==ratio))))
}
print(r_average_vector)

relativised_pool_counts_high_copies_4FR <- mutate(pool_counts_high_copies_4FR, Pb_relative_value=value/r_average_vector, group=ifelse(population=="Pb","Pb","lines"))
relativised_pool_counts_low_copies_4FR$group = factor(relativised_pool_counts_low_copies_4FR$group,levels=c("Pb","lines"))

#Obtain per population averages and standard errors:
se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function

average_relativised_pool_counts_high_copies_4FR <- data_frame("population"=character(0),"generation"=character(0),"recombination"=character(0),"ratio"=character(0),"avg_Pb_relative_value"=character(0),"se_Pb_relative_value"=character(0)) #next, create the empty dataframe

for (pop in unique(relativised_pool_counts_high_copies_4FR$group)) { #then loop over each population and feature to get the (relativised) mean and standard error, and feed the dataframe
  print(pop)
  for (g in unlist(unique(select(filter(relativised_pool_counts_high_copies_4FR,group==pop),generation)),use.names=F)) {
  #species <- filter(relativised_pool_counts_high_copies_4FR,ratio==r & population==pop) %>% select(species) %>% unlist(.,use.names=F) %>% unique()
    for (r in unique(relativised_pool_counts_high_copies_4FR$ratio)) {
      print(r)
      pop_mean <- filter(relativised_pool_counts_high_copies_4FR,ratio==r & generation==g & group==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% mean()
      #print(paste0(pop," feature ",r," average is ",pop_mean))
      pop_se <- filter(relativised_pool_counts_high_copies_4FR,ratio==r & generation==g & group==pop) %>% select(Pb_relative_value) %>% unlist(.,use.names=F) %>% se()
      #print(paste0(pop," feature ",r," std error is ",pop_se))
      row_data <- cbind(pop,g,"high",r,pop_mean,pop_se)
      colnames(row_data) <- c("population","generation","recombination","ratio","avg_Pb_relative_value","se_Pb_relative_value")
      average_relativised_pool_counts_high_copies_4FR <- rbind(average_relativised_pool_counts_high_copies_4FR,row_data,stringsAsFactors=F)
    }
  }
}
average_relativised_pool_counts_high_copies_4FR$population = factor(average_relativised_pool_counts_high_copies_4FR$population,levels=c("Pb","lines"))
average_relativised_pool_counts_high_copies_4FR$ratio = factor(average_relativised_pool_counts_high_copies_4FR$ratio,levels=c("fourfold","tolerated","deleterious","LoF"))
average_relativised_pool_counts_high_copies_4FR$generation <- as.factor(as.numeric(average_relativised_pool_counts_high_copies_4FR$generation))
average_relativised_pool_counts_high_copies_4FR$avg_Pb_relative_value <- as.numeric(average_relativised_pool_counts_high_copies_4FR$avg_Pb_relative_value)
average_relativised_pool_counts_high_copies_4FR$se_Pb_relative_value <- as.numeric(average_relativised_pool_counts_high_copies_4FR$se_Pb_relative_value)
average_relativised_pool_counts_high_copies_4FR


####Then combine both datasets####
average_relativised_pool_counts_combined_copies_4FR <- rbind(average_relativised_pool_counts_low_copies_4FR,average_relativised_pool_counts_high_copies_4FR)
average_relativised_pool_counts_combined_copies_4FR$recombination = factor(average_relativised_pool_counts_combined_copies_4FR$recombination,levels=c("low","high"))

####Then plot the data####
#Combined version (Pb):
Pb_lines_relativised_pool_counts_sites_4FR_ggplot_recvert <- ggplot(data=filter(average_relativised_pool_counts_combined_copies_4FR,ratio!="fourfold", generation!=5), aes(generation,avg_Pb_relative_value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_grid(. ~ ratio) +
  #geom_line(aes(colour=recombination, group=recombination),alpha=0.5,size=0.5) +
  geom_point(aes(colour=recombination,shape=population),size=2,position=position_dodge(0.4)) +
  geom_errorbar(aes(ymin=avg_Pb_relative_value-2*se_Pb_relative_value, ymax=avg_Pb_relative_value+2*se_Pb_relative_value, colour=recombination,group=interaction(population,recombination)), position=position_dodge(0.4),size=0.5,width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Number of sites\n relative to Pb-000 4-fold syn.") +
  scale_y_continuous(breaks = seq(0.5, 1, by = 0.1)) +
  scale_shape_manual(values=c(1,16)) +
  ylim(0.4,1.1) +
  ggtitle("Pools") +
  #guides(colour="none") +
  labs(colour = "Recombination", shape = "Population") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_text()
  )
Pb_lines_relativised_pool_counts_sites_4FR_ggplot_recvert
ggsave(paste0("Pb_lines_relativised_pool_counts_sites_4FR_",type,".all_recombination.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)


```

####Raw:
#####Gen0-140 pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
pool_counts <- read_tsv(paste0(wd_path,"counts_pool_gen0-140_",type,"_summary.txt")) 

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$generation <- as.numeric(gsub("gen","",pool_counts$generation))
pool_counts <- pool_counts %>% arrange(generation)
pool_counts$generation = factor(pool_counts$generation)
pool_counts <- pool_counts %>% arrange(population,generation)

pool_counts_copies_4FR <- pool_counts %>% mutate(fourfold=fourfold_V/fourfold_V,tolerated=tolerated_V/fourfold_V,deleterious=deleterious_V/fourfold_V,LoF=LoF_V/fourfold_V) %>% select(!contains(c("_V","_D","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_copies_4FR


#Humberto version (Pb):
Pb_raw_pool_counts_sites_4FR_ggplot <- ggplot(data=filter(pool_counts_copies_4FR,ratio!="fourfold", population=="Pb" & generation!=5), aes(generation,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_wrap(. ~ ratio, scales = "free", nrow=1) +
  geom_point(aes(colour=generation),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average number of sites\n relative to 4-fold syn.") +
  #scale_y_continuous(breaks = seq(0.6, 1.2, by = 0.2)) +
  #ylim(0.5,1.4) +
  ggtitle("Pools, Pb") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_raw_pool_counts_sites_4FR_ggplot
ggsave(paste0("Pb_pool_counts_variants_4FR_",type,"_Humberto_version.pdf"), width=17.5, height=12, units="cm", device="pdf", path=wd_path)


#Humberto version (lines):
lines_raw_pool_counts_sites_4FR_ggplot <- ggplot(data=filter(pool_counts_copies_4FR,ratio!="fourfold", population!="Pb"), aes(ratio,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=generation)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average number of sites\n relative to 4-fold syn.") +
  #scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  #ylim(0.7,1.05) +
  ggtitle("Pools, lines (average)") +
  scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
lines_raw_pool_counts_sites_4FR_ggplot
ggsave(paste0("lines_pool_counts_variants_4FR_",type,"_Humberto_version.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```


##Combine derived count and site ratios:
###Relative to 4-fold:
####Relative to Pb:
#####Gen0-140 pools Pb and line averages (Humberto version):
```{r}

library(grid)
library(gridExtra)
library(egg)

ggplot_combined_relativised <- grid.arrange(set_panel_size(Pb_relativised_pool_counts_sites_4FR_ggplot,width=unit(3.5,"cm"),height=unit(8,"cm")),set_panel_size(lines_relativised_pool_counts_sites_4FR_ggplot,width=unit(10,"cm"),height=unit(8,"cm")),set_panel_size(Pb_relativised_pool_counts_copies_4FR_ggplot,width=unit(3.5,"cm"),height=unit(8,"cm")),set_panel_size(lines_relativised_pool_counts_copies_4FR_ggplot,width=unit(10,"cm"),height=unit(8,"cm")), ncol=2)

ggsave("relativised_pool_counts_sites_and_counts_4FR_all_Humberto_version.pdf", width=31, height=23, units="cm", device="pdf", path="/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/",ggplot_combined_relativised)

```

#####Low rec. gen0-140 pools Pb and line averages (Humberto version):
```{r}

library(grid)
library(gridExtra)
library(egg)

ggplot_combined_relativised <- grid.arrange(set_panel_size(Pb_relativised_pool_counts_sites_4FR_ggplot_lowrec,width=unit(3.5,"cm"),height=unit(8,"cm")),set_panel_size(lines_relativised_pool_counts_sites_4FR_ggplot_lowrec,width=unit(10,"cm"),height=unit(8,"cm")),set_panel_size(Pb_relativised_pool_counts_copies_4FR_ggplot_lowrec,width=unit(3.5,"cm"),height=unit(8,"cm")),set_panel_size(lines_relativised_pool_counts_copies_4FR_ggplot_lowrec,width=unit(10,"cm"),height=unit(8,"cm")), ncol=2)

ggsave("relativised_pool_counts_sites_and_counts_4FR_all_Humberto_version.low_recombination.pdf", width=30, height=23, units="cm", device="pdf", path="/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/",ggplot_combined_relativised)

```

#####High rec. gen0-140 pools Pb and line averages (Humberto version):
```{r}

library(grid)
library(gridExtra)
library(egg)

ggplot_combined_relativised <- grid.arrange(set_panel_size(Pb_relativised_pool_counts_sites_4FR_ggplot_highrec,width=unit(3.5,"cm"),height=unit(8,"cm")),set_panel_size(lines_relativised_pool_counts_sites_4FR_ggplot_highrec,width=unit(10,"cm"),height=unit(8,"cm")),set_panel_size(Pb_relativised_pool_counts_copies_4FR_ggplot_highrec,width=unit(3.5,"cm"),height=unit(8,"cm")),set_panel_size(lines_relativised_pool_counts_copies_4FR_ggplot_highrec,width=unit(10,"cm"),height=unit(8,"cm")), ncol=2)

ggsave("relativised_pool_counts_sites_and_counts_4FR_all_Humberto_version.high_recombination.pdf", width=30, height=23, units="cm", device="pdf", path="/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/",ggplot_combined_relativised)

```

#####All rec. gen0-140 pools Pb and line averages (Humberto version):
```{r}

library(grid)
library(gridExtra)
library(egg)

ggplot_combined_relativised <- grid.arrange(set_panel_size(Pb_relativised_pool_counts_sites_4FR_ggplot_recvert,width=unit(3.5,"cm"),height=unit(8,"cm")),set_panel_size(lines_relativised_pool_counts_sites_4FR_ggplot_rec,width=unit(8,"cm"),height=unit(5,"cm")),set_panel_size(Pb_relativised_pool_counts_copies_4FR_ggplot_recvert,width=unit(3.5,"cm"),height=unit(8,"cm")),set_panel_size(lines_relativised_pool_counts_copies_4FR_ggplot_rec,width=unit(8,"cm"),height=unit(5,"cm")), ncol=2)

ggsave("relativised_pool_counts_sites_and_counts_4FR_all_Humberto_version.all_recombination.pdf", width=30, height=25, units="cm", device="pdf", path="/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/",ggplot_combined_relativised)

```

####Raw:
#####Gen0-140  pools Pb and line averages (Humberto version):
```{r}

library(grid)
library(gridExtra)
library(egg)

ggplot_combined_raw <- grid.arrange(set_panel_size(Pb_raw_pool_counts_sites_4FR_ggplot,width=unit(3.5,"cm"),height=unit(8,"cm")),set_panel_size(lines_raw_pool_counts_sites_4FR_ggplot,width=unit(10,"cm"),height=unit(8,"cm")),set_panel_size(Pb_raw_pool_counts_copies_4FR_ggplot,width=unit(3.5,"cm"),height=unit(8,"cm")),set_panel_size(lines_raw_pool_counts_copies_4FR_ggplot,width=unit(10,"cm"),height=unit(8,"cm")), ncol=2)

ggsave("raw_pool_counts_sites_and_counts_4FR_all_Humberto_version.pdf", width=32.5, height=23, units="cm", device="pdf", path="/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/",ggplot_combined_raw)

```

##Derived count proportions:
###Sites with 10 or more reads:
####Raw:
#####Pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
pool_counts_gen0_40 <- read_tsv(paste0(wd_path,"propD_pool_gen0-40_",type,"_d10_summary.txt")) %>% mutate(gen=ifelse(sample=="LBT0",000,ifelse(sample=="c1",005,ifelse(sample=="BT20",020,ifelse(sample=="BT30",030,040)))))
pool_counts_gen140 <- read_tsv(paste0(wd_path,"propD_pool_gen140_",type,"_d10_summary.txt")) %>% mutate(gen="140" %>% as.double)

pool_counts <- rbind(pool_counts_gen0_40,pool_counts_gen140)

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$gen = factor(pool_counts$gen)
pool_counts <- pool_counts %>% arrange(population,gen)

pool_counts_propD <- pool_counts %>% select("population","gen",contains("propD")) %>% gather(ratio,value,-gen,-population,factor_key=T)
pool_counts_propD$ratio <- gsub("_propD", "", pool_counts_propD$ratio)
pool_counts_propD

pool_counts_propD$population = factor(pool_counts_propD$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts_propD$ratio = factor(pool_counts_propD$ratio,levels=c("intronic", "synonymous","fourfold","missense","tolerated","deleterious","LoF"))

#Humberto version (Pb):
Pb_pool_propD_ggplot <- ggplot(data=filter(pool_counts_propD,population=="Pb" & gen!=5), aes(gen,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_wrap(. ~ ratio, nrow=1) +
  geom_point(aes(colour=gen),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average proportion\n of derived/total reads") +
  scale_y_continuous(breaks = seq(0, 0.5, by = 0.1)) +
  ylim(0,0.5) +
  ggtitle("Pb") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_pool_propD_ggplot
ggsave(paste0("Pb_pool_propD_",type,"_Humberto_version.pdf"), width=24, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
line_pool_propD_ggplot <- ggplot(data=filter(pool_counts_propD,population!="Pb"), aes(ratio,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=gen)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average proportion\n of derived/total reads") +
  #scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  #ylim(0.7,1.05) +
  ggtitle("lines") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
line_pool_propD_ggplot
ggsave(paste0("lines_pool_propD_",type,"_Humberto_version.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)


```

####Relative to synonymous:
#####Pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
pool_counts_gen0_40 <- read_tsv(paste0(wd_path,"counts_pool_gen0-40_",type,"_d10_summary.txt")) %>% mutate(gen=ifelse(sample=="LBT0",000,ifelse(sample=="c1",005,ifelse(sample=="BT20",020,ifelse(sample=="BT30",030,040)))))
pool_counts_gen140 <- read_tsv(paste0(wd_path,"counts_pool_gen140_",type,"_d10_summary.txt")) %>% mutate(gen="140" %>% as.double)

pool_counts <- rbind(pool_counts_gen0_40,pool_counts_gen140)

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$gen = factor(pool_counts$gen)
pool_counts <- pool_counts %>% arrange(population,gen)

pool_counts_propD <- pool_counts %>% mutate(intronic=intronic_propD/synonymous_propD,synonymous=synonymous_propD/synonymous_propD,fourfold=fourfold_propD/synonymous_propD,missense=missense_propD/synonymous_propD,tolerated=tolerated_propD/synonymous_propD,deleterious=deleterious_propD/synonymous_propD,stopgain=stopgain_propD/synonymous_propD,stoploss=stoploss_propD/synonymous_propD) %>% select(c(30,29,31:38)) %>% gather(ratio,value,-gen,-population,factor_key=T)
pool_counts_propD$ratio <- gsub("_propD", "", pool_counts_propD$ratio)
pool_counts_propD

pool_counts_propD$population = factor(pool_counts_propD$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts_propD$ratio = factor(pool_counts_propD$ratio,levels=c("intronic", "synonymous","fourfold","missense","tolerated","deleterious","stopgain","stoploss"))
#write_tsv(average_relativised_pool_counts_copies_synR,paste0(wd_path,"/pop_average_relativised_pool_counts_copies_synR_",type,".txt"))

#Humberto version (Pb):
Pb_pool_counts_propD_ggplot <- ggplot(data=filter(pool_counts_propD,population=="Pb" & gen!=5), aes(gen,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_wrap(. ~ ratio, nrow=1) +
  geom_point(aes(colour=gen),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average proportion\n of derived/total reads relative to syn.") +
  scale_y_continuous(breaks = seq(0.3, 1, by = 0.1)) +
  ylim(0.3,1) +
  ggtitle(paste0("Pools ",type)) +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_pool_counts_propD_ggplot
ggsave(paste0("Pb_pool_counts_propD_synR_",type,"_Humberto_version.pdf"), width=24, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
line_pool_counts_propD_ggplot <- ggplot(data=filter(pool_counts_propD,population!="Pb"), aes(ratio,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=gen)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average proportion of\n derived/total reads relative to syn.") +
  #scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  #ylim(0.7,1.05) +
  ggtitle(paste0("Pools ",type)) +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
line_pool_counts_propD_ggplot
ggsave(paste0("lines_pool_counts_propD_synR_",type,"_Humberto_version.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)


```

####Relative to 4fold:
#####Pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
pool_counts_gen0_40 <- read_tsv(paste0(wd_path,"propD_pool_gen0-40_",type,"_d10_summary.txt")) %>% mutate(gen=ifelse(sample=="LBT0",000,ifelse(sample=="c1",005,ifelse(sample=="BT20",020,ifelse(sample=="BT30",030,040)))))  
pool_counts_gen140 <- read_tsv(paste0(wd_path,"propD_pool_gen140_",type,"_d10_summary.txt")) %>% mutate(gen="140" %>% as.double)

pool_counts <- rbind(pool_counts_gen0_40,pool_counts_gen140)

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$gen = factor(pool_counts$gen)
pool_counts <- pool_counts %>% arrange(population,gen)

pool_counts_propD <- pool_counts %>% mutate(fourfold=fourfold_propD/fourfold_propD,tolerated=tolerated_propD/fourfold_propD,deleterious=deleterious_propD/fourfold_propD,LoF=LoF_propD/fourfold_propD) %>% select(-contains(c("_propD","_V","sample"))) %>% gather(ratio,value,-gen,-population,factor_key=T)
pool_counts_propD


#Humberto version (Pb):
Pb_pool_propD_ggplot <- ggplot(data=filter(pool_counts_propD,population=="Pb" & gen!=5), aes(gen,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_wrap(. ~ ratio, nrow=1) +
  geom_point(aes(colour=gen),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average proportion\n of derived/total reads relative to 4-fold") +
  scale_y_continuous(breaks = seq(0.3, 1, by = 0.1)) +
  ylim(0.3,1) +
  ggtitle("Pb") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_pool_propD_ggplot
ggsave(paste0("Pb_pool_propD_4FR_",type,"_Humberto_version.pdf"), width=24, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
line_pool_propD_ggplot <- ggplot(data=filter(pool_counts_propD,population!="Pb"), aes(ratio,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=gen)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average proportion of\n derived/total reads relative to 4-fold") +
  #scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  #ylim(0.7,1.05) +
  ggtitle("lines") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
line_pool_propD_ggplot
ggsave(paste0("lines_pool_propD_4FR_",type,"_Humberto_version.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

#####Gen0-140 pools Pb and line averages (Humberto version):
```{r Plot variant count results}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)

type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
pool_counts <- read_tsv(paste0(wd_path,"propD_pool_gen0-140_",type,"_d10_summary.txt"))

codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
pool_counts <- left_join(pool_counts,codes_dictionary,by=c("sample"="old"))
pool_counts$population = factor(pool_counts$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
pool_counts$generation <- as.numeric(gsub("gen","",pool_counts$generation))
pool_counts <- pool_counts %>% arrange(generation)
pool_counts$generation = factor(pool_counts$generation)
pool_counts <- pool_counts %>% arrange(population,generation)

pool_counts_propD <- pool_counts %>% mutate(fourfold=fourfold_propD/fourfold_propD,tolerated=tolerated_propD/fourfold_propD,deleterious=deleterious_propD/fourfold_propD,LoF=LoF_propD/fourfold_propD) %>% select(-contains(c("_propD","_V","sample"))) %>% gather(ratio,value,-generation,-population,factor_key=T)
pool_counts_propD


#Humberto version (Pb):
Pb_pool_propD_ggplot <- ggplot(data=filter(pool_counts_propD,population=="Pb" & generation!=5 & ratio!="fourfold"), aes(generation,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  facet_wrap(. ~ ratio, nrow=1) +
  geom_point(aes(colour=generation),alpha=0.5,size=2) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average proportion of\n derived/total reads relative to 4-fold") +
  scale_y_continuous(breaks = seq(0.3, 1, by = 0.1)) +
  ylim(0.3,1) +
  ggtitle("Base population") +
  #guides(colour="none") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        legend.position="none",
        legend.title=element_blank()
  )
Pb_pool_propD_ggplot
ggsave(paste0("Pb_pool_propD_4FR_",type,"_Humberto_version.pdf"), width=24, height=12, units="cm", device="pdf", path=wd_path)

#Humberto version (lines):
line_pool_propD_ggplot <- ggplot(data=filter(pool_counts_propD,population!="Pb" & ratio!="fourfold"), aes(ratio,value)) +
  #facet_wrap(feature ~ species,nrow=6,ncol=2,scales="free") +
  #facet_grid(. ~ population) +
  geom_boxplot(aes(colour=generation)) +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  #ggtitle("Proportion of reads at different NM") +
  ylab("Average proportion of\n derived/total reads relative to 4-fold") +
  #scale_y_continuous(breaks = seq(0.7, 1.1, by = 0.2)) +
  #ylim(0.7,1.05) +
  ggtitle("Lines (average)") +
  #guides(colour="none") +
  scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
line_pool_propD_ggplot
ggsave(paste0("lines_pool_propD_4FR_",type,"_Humberto_version.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)


#Combine both:
library(grid)
library(gridExtra)
library(egg)

ggplot_combined_propD <- grid.arrange(set_panel_size(Pb_pool_propD_ggplot,width=unit(3.5,"cm"),height=unit(8,"cm")),set_panel_size(line_pool_propD_ggplot,width=unit(10,"cm"),height=unit(8,"cm")), ncol=2)

ggsave("pool_propD_4FR_all_Humberto_version.pdf", width=30, height=12, units="cm", device="pdf", path="/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/",ggplot_combined_propD)

```
##SFS:
###Sites with 10 or more reads:
####Pools:
#####All sites, raw counts:
```{r}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
SFS_raw_data_gen0_40 <- read_tsv(paste0(wd_path,"SFS_pool_gen0-40_combined_",type,"_summary.txt")) %>% mutate(gen=ifelse(sample=="LBT0",000,ifelse(sample=="c1",005,ifelse(sample=="BT20",020,ifelse(sample=="BT30",030,040)))))
SFS_raw_data_gen140 <- read_tsv(paste0(wd_path,"SFS_pool_gen140_combined_",type,"_summary.txt")) %>% mutate(gen="140" %>% as.double)

SFS_raw_data_combined <- rbind(SFS_raw_data_gen0_40,SFS_raw_data_gen140)
SFS_raw_data_combined$gen = factor(SFS_raw_data_combined$gen)
SFS_raw_data_combined <- SFS_raw_data_combined %>% rowwise %>% mutate(total= sum(c_across(where(is.numeric)))) #row-wise sum of all numeric columns (the "af" columns; "gen" is excluded if its converted to a factor first.
codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
SFS_raw_data_combined <- left_join(SFS_raw_data_combined,codes_dictionary,by=c("sample"="old"))
SFS_raw_data_combined$population = factor(SFS_raw_data_combined$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
SFS_raw_data_combined$category = factor(SFS_raw_data_combined$category,levels=c("fourfold","tolerated","deleterious","LoF"))
SFS_raw_data_combined <- SFS_raw_data_combined %>% arrange(population,gen) %>% select(population,gen,category,contains("af_"))

SFS_raw_Pb_average <- SFS_raw_data_combined %>% filter(population=="Pb")
SFS_raw_lines_average <- SFS_raw_data_combined %>% filter(population!="Pb") %>% group_by(gen,category) %>% summarise_at(vars(contains("af_")), mean) %>% mutate(population="lines") %>% select(population,gen,category,contains("af_")) %>% ungroup()
SFS_raw_data_average <- rbind(SFS_raw_Pb_average,SFS_raw_lines_average)

SFS_raw_average_tidy <- SFS_raw_data_average %>% gather(bin,average,-population,-gen,-category,factor_key=T)
SFS_raw_average_tidy
SFS_raw_average_tidy$bin <- gsub('af_', '', SFS_raw_average_tidy$bin)
SFS_raw_average_tidy$bin <- as.factor(SFS_raw_average_tidy$bin)

se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function
SFS_raw_Pb_stderr <- SFS_raw_data_combined %>% filter(population=="Pb") %>% group_by(gen,category) %>% summarise_at(vars(contains("af_")), se) %>% mutate(population="Pb") %>% select(population,gen,category,contains("af_")) %>% ungroup()
SFS_raw_lines_stderr <- SFS_raw_data_combined %>% filter(population!="Pb") %>% group_by(gen,category) %>% summarise_at(vars(contains("af_")), se) %>% mutate(population="lines") %>% select(population,gen,category,contains("af_")) %>% ungroup()
SFS_raw_data_stderr <- rbind(SFS_raw_Pb_stderr,SFS_raw_lines_stderr)

SFS_raw_stderr_tidy <- SFS_raw_data_stderr %>% gather(bin,stderr,-population,-gen,-category,factor_key=T)
SFS_raw_stderr_tidy
SFS_raw_stderr_tidy$bin <- gsub('af_', '', SFS_raw_stderr_tidy$bin)
SFS_raw_stderr_tidy$bin <- as.factor(SFS_raw_stderr_tidy$bin)

SFS_raw_data_tidy <- left_join(SFS_raw_average_tidy,SFS_raw_stderr_tidy,by=c("population","gen","category","bin"))


#Pb:
SFS_raw_data_Pb_ggplot <- ggplot(data=filter(SFS_raw_data_tidy, population=="Pb", gen!=5), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2,scales="free_y") +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_col(aes(fill=gen), position="dodge") +
  #geom_errorbar(aes(ymin=avg_Pb_raw_value-se_Pb_raw_value, ymax=avg_Pb_raw_value+se_Pb_raw_value), position=position_dodge(), width=0.5) +
  ggtitle("SFS, Pb") +
  ylab("Number of sites") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  #scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
SFS_raw_data_Pb_ggplot
ggsave(paste0("Pb_",type,"_raw_SFS.pdf"), width=40, height=12, units="cm", device="pdf", path=wd_path)

#Lines:
SFS_raw_data_lines_ggplot <- ggplot(data=filter(SFS_raw_data_tidy, population=="lines"), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2,scales="free_y") +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_bar(aes(fill=gen), position=position_dodge(), stat="identity") +
  geom_errorbar(aes(x=bin,ymin=average-stderr, ymax=average+stderr, group=gen), width=0.5, position=position_dodge(0.9)) +
  ggtitle("SFS, line average") +
  ylab("Number of sites") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  #scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
SFS_raw_data_lines_ggplot
ggsave(paste0("lines_average_",type,"_raw_SFS.pdf"), width=30, height=12, units="cm", device="pdf", path=wd_path)

```

#####Only variable sites, raw counts:
```{r}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
SFS_raw_data_gen0_40 <- read_tsv(paste0(wd_path,"SFS_pool_gen0-40_combined_",type,"_summary.txt")) %>% mutate(gen=ifelse(sample=="LBT0",000,ifelse(sample=="c1",005,ifelse(sample=="BT20",020,ifelse(sample=="BT30",030,040)))))
SFS_raw_data_gen140 <- read_tsv(paste0(wd_path,"SFS_pool_gen140_combined_",type,"_summary.txt")) %>% mutate(gen="140" %>% as.double)

SFS_raw_data_combined <- rbind(SFS_raw_data_gen0_40,SFS_raw_data_gen140) %>% select(-af_000,-af_100)
SFS_raw_data_combined$gen = factor(SFS_raw_data_combined$gen)
SFS_raw_data_combined <- SFS_raw_data_combined %>% rowwise %>% mutate(total= sum(c_across(where(is.numeric)))) #row-wise sum of all numeric columns (the "af" columns; "gen" is excluded if its converted to a factor first.
codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
SFS_raw_data_combined <- left_join(SFS_raw_data_combined,codes_dictionary,by=c("sample"="old"))
SFS_raw_data_combined$population = factor(SFS_raw_data_combined$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
SFS_raw_data_combined$category = factor(SFS_raw_data_combined$category,levels=c("fourfold","tolerated","deleterious","LoF"))
SFS_raw_data_combined <- SFS_raw_data_combined %>% arrange(population,gen) %>% select(population,gen,category,contains("af_"))

SFS_raw_Pb_average <- SFS_raw_data_combined %>% filter(population=="Pb")
SFS_raw_lines_average <- SFS_raw_data_combined %>% filter(population!="Pb") %>% group_by(gen,category) %>% summarise_at(vars(contains("af_")), mean) %>% mutate(population="lines") %>% select(population,gen,category,contains("af_")) %>% ungroup()
SFS_raw_data_average <- rbind(SFS_raw_Pb_average,SFS_raw_lines_average)

SFS_raw_average_tidy <- SFS_raw_data_average %>% gather(bin,average,-population,-gen,-category,factor_key=T)
SFS_raw_average_tidy
SFS_raw_average_tidy$bin <- gsub('af_', '', SFS_raw_average_tidy$bin)
SFS_raw_average_tidy$bin <- as.factor(SFS_raw_average_tidy$bin)

se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function
SFS_raw_Pb_stderr <- SFS_raw_data_combined %>% filter(population=="Pb") %>% group_by(gen,category) %>% summarise_at(vars(contains("af_")), se) %>% mutate(population="Pb") %>% select(population,gen,category,contains("af_")) %>% ungroup()
SFS_raw_lines_stderr <- SFS_raw_data_combined %>% filter(population!="Pb") %>% group_by(gen,category) %>% summarise_at(vars(contains("af_")), se) %>% mutate(population="lines") %>% select(population,gen,category,contains("af_")) %>% ungroup()
SFS_raw_data_stderr <- rbind(SFS_raw_Pb_stderr,SFS_raw_lines_stderr)

SFS_raw_stderr_tidy <- SFS_raw_data_stderr %>% gather(bin,stderr,-population,-gen,-category,factor_key=T)
SFS_raw_stderr_tidy
SFS_raw_stderr_tidy$bin <- gsub('af_', '', SFS_raw_stderr_tidy$bin)
SFS_raw_stderr_tidy$bin <- as.factor(SFS_raw_stderr_tidy$bin)

SFS_raw_data_tidy <- left_join(SFS_raw_average_tidy,SFS_raw_stderr_tidy,by=c("population","gen","category","bin"))


#Pb:
SFS_raw_data_Pb_ggplot <- ggplot(data=filter(SFS_raw_data_tidy, population=="Pb", gen!=5), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2,scales="free_y") +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_col(aes(fill=gen), position="dodge") +
  #geom_errorbar(aes(ymin=avg_Pb_raw_value-se_Pb_raw_value, ymax=avg_Pb_raw_value+se_Pb_raw_value), position=position_dodge(), width=0.5) +
  ggtitle("SFS, Pb") +
  ylab("Number of sites") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  #scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
SFS_raw_data_Pb_ggplot
ggsave(paste0("Pb_",type,"_raw_variants_SFS.pdf"), width=40, height=12, units="cm", device="pdf", path=wd_path)

#Lines:
SFS_raw_data_lines_ggplot <- ggplot(data=filter(SFS_raw_data_tidy, population=="lines"), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2,scales="free_y") +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_bar(aes(fill=gen), position=position_dodge(), stat="identity") +
  geom_errorbar(aes(x=bin,ymin=average-stderr, ymax=average+stderr, group=gen), width=0.5, position=position_dodge(0.9)) +
  ggtitle("SFS, line average") +
  ylab("Number of sites") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  #scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
SFS_raw_data_lines_ggplot
ggsave(paste0("lines_average_",type,"_raw_variants_SFS.pdf"), width=30, height=12, units="cm", device="pdf", path=wd_path)

```

#####All sites, proportions:
```{r}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
SFS_raw_data_gen0_40 <- read_tsv(paste0(wd_path,"SFS_pool_gen0-40_combined_",type,"_summary.txt")) %>% mutate(gen=ifelse(sample=="LBT0",000,ifelse(sample=="c1",005,ifelse(sample=="BT20",020,ifelse(sample=="BT30",030,040)))))
SFS_raw_data_gen140 <- read_tsv(paste0(wd_path,"SFS_pool_gen140_combined_",type,"_summary.txt")) %>% mutate(gen="140" %>% as.double)

SFS_raw_data_combined <- rbind(SFS_raw_data_gen0_40,SFS_raw_data_gen140)
SFS_raw_data_combined$gen = factor(SFS_raw_data_combined$gen)
SFS_raw_data_combined <- SFS_raw_data_combined %>% rowwise %>% mutate(total= sum(c_across(where(is.numeric)))) #row-wise sum of all numeric columns (the "af" columns; "gen" is excluded if its converted to a factor first.
codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
SFS_raw_data_combined <- left_join(SFS_raw_data_combined,codes_dictionary,by=c("sample"="old"))
SFS_raw_data_combined$population = factor(SFS_raw_data_combined$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
SFS_raw_data_combined$category = factor(SFS_raw_data_combined$category,levels=c("fourfold","tolerated","deleterious","LoF"))
SFS_raw_data_combined <- SFS_raw_data_combined %>% arrange(population,gen)

SFS_relative_data_combined <- SFS_raw_data_combined %>% mutate_at(vars(contains("af_")),list(relative=~./total)) %>% select(population,gen,category,contains("_relative"))

SFS_relative_Pb_average <- SFS_relative_data_combined %>% filter(population=="Pb")
SFS_relative_lines_average <- SFS_relative_data_combined %>% filter(population!="Pb") %>% group_by(gen,category) %>% summarise_at(vars(contains("_relative")), mean) %>% mutate(population="lines") %>% select(population,gen,category,contains("_relative")) %>% ungroup()
SFS_relative_data_average <- rbind(SFS_relative_Pb_average,SFS_relative_lines_average)

SFS_relative_average_tidy <- SFS_relative_data_average %>% gather(bin,average,-population,-gen,-category,factor_key=T)
SFS_relative_average_tidy
SFS_relative_average_tidy$bin <- gsub('af_', '', gsub('_relative', '', SFS_relative_average_tidy$bin))
SFS_relative_average_tidy$bin <- as.factor(SFS_relative_average_tidy$bin)

se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function
SFS_relative_Pb_stderr <- SFS_relative_data_combined %>% filter(population=="Pb") %>% group_by(gen,category) %>% summarise_at(vars(contains("_relative")), se) %>% mutate(population="Pb") %>% select(population,gen,category,contains("_relative")) %>% ungroup()
SFS_relative_lines_stderr <- SFS_relative_data_combined %>% filter(population!="Pb") %>% group_by(gen,category) %>% summarise_at(vars(contains("_relative")), se) %>% mutate(population="lines") %>% select(population,gen,category,contains("_relative")) %>% ungroup()
SFS_relative_data_stderr <- rbind(SFS_relative_Pb_stderr,SFS_relative_lines_stderr)

SFS_relative_stderr_tidy <- SFS_relative_data_stderr %>% gather(bin,stderr,-population,-gen,-category,factor_key=T)
SFS_relative_stderr_tidy
SFS_relative_stderr_tidy$bin <- gsub('af_', '', gsub('_relative', '', SFS_relative_stderr_tidy$bin))
SFS_relative_stderr_tidy$bin <- as.factor(SFS_relative_stderr_tidy$bin)

SFS_relative_data_tidy <- left_join(SFS_relative_average_tidy,SFS_relative_stderr_tidy,by=c("population","gen","category","bin"))


#Pb:
SFS_relative_data_Pb_ggplot <- ggplot(data=filter(SFS_relative_data_tidy, population=="Pb", gen!=5), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2) +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_col(aes(fill=gen), position="dodge") +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  ggtitle("SFS, Pb") +
  ylab("Proportion") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  #scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
SFS_relative_data_Pb_ggplot
ggsave(paste0("Pb_",type,"_proportion_SFS.pdf"), width=39, height=12, units="cm", device="pdf", path=wd_path)

#Lines:
SFS_relative_data_lines_ggplot <- ggplot(data=filter(SFS_relative_data_tidy, population=="lines"), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2) +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_bar(aes(fill=gen), position=position_dodge(), stat="identity") +
  geom_errorbar(aes(x=bin,ymin=average-stderr, ymax=average+stderr, group=gen), width=0.5, position=position_dodge(0.9)) +
  ggtitle("SFS, line average") +
  ylab("Proportion") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  #scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
SFS_relative_data_lines_ggplot
ggsave(paste0("lines_average_",type,"_proportion_SFS.pdf"), width=29, height=12, units="cm", device="pdf", path=wd_path)

```

#####Only variable sites, proportions:
```{r}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/")
SFS_raw_data_gen0_40 <- read_tsv(paste0(wd_path,"SFS_pool_gen0-40_combined_",type,"_summary.txt")) %>% mutate(gen=ifelse(sample=="LBT0",000,ifelse(sample=="c1",005,ifelse(sample=="BT20",020,ifelse(sample=="BT30",030,040)))))
SFS_raw_data_gen140 <- read_tsv(paste0(wd_path,"SFS_pool_gen140_combined_",type,"_summary.txt")) %>% mutate(gen="140" %>% as.double)

SFS_raw_data_combined <- rbind(SFS_raw_data_gen0_40,SFS_raw_data_gen140) %>% select(-af_00,-af_10)
SFS_raw_data_combined$gen = factor(SFS_raw_data_combined$gen)
SFS_raw_data_combined <- SFS_raw_data_combined %>% rowwise %>% mutate(total= sum(c_across(where(is.numeric)))) #row-wise sum of all numeric columns (the "af" columns; "gen" is excluded if its converted to a factor first.
codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
SFS_raw_data_combined <- left_join(SFS_raw_data_combined,codes_dictionary,by=c("sample"="old"))
SFS_raw_data_combined$population = factor(SFS_raw_data_combined$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
SFS_raw_data_combined$category = factor(SFS_raw_data_combined$category,levels=c("fourfold","tolerated","deleterious","LoF"))
SFS_raw_data_combined <- SFS_raw_data_combined %>% arrange(population,gen)

SFS_relative_data_combined <- SFS_raw_data_combined %>% mutate_at(vars(contains("af_")),list(relative=~./total)) %>% select(population,gen,category,contains("_relative"))

SFS_relative_Pb_average <- SFS_relative_data_combined %>% filter(population=="Pb")
SFS_relative_lines_average <- SFS_relative_data_combined %>% filter(population!="Pb") %>% group_by(gen,category) %>% summarise_at(vars(contains("_relative")), mean) %>% mutate(population="lines") %>% select(population,gen,category,contains("_relative")) %>% ungroup()
SFS_relative_data_average <- rbind(SFS_relative_Pb_average,SFS_relative_lines_average)

SFS_relative_average_tidy <- SFS_relative_data_average %>% gather(bin,average,-population,-gen,-category,factor_key=T)
SFS_relative_average_tidy
SFS_relative_average_tidy$bin <- gsub('af_', '', gsub('_relative', '', SFS_relative_average_tidy$bin))
SFS_relative_average_tidy$bin <- as.factor(SFS_relative_average_tidy$bin)

se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function
SFS_relative_Pb_stderr <- SFS_relative_data_combined %>% filter(population=="Pb") %>% group_by(gen,category) %>% summarise_at(vars(contains("_relative")), se) %>% mutate(population="Pb") %>% select(population,gen,category,contains("_relative")) %>% ungroup()
SFS_relative_lines_stderr <- SFS_relative_data_combined %>% filter(population!="Pb") %>% group_by(gen,category) %>% summarise_at(vars(contains("_relative")), se) %>% mutate(population="lines") %>% select(population,gen,category,contains("_relative")) %>% ungroup()
SFS_relative_data_stderr <- rbind(SFS_relative_Pb_stderr,SFS_relative_lines_stderr)

SFS_relative_stderr_tidy <- SFS_relative_data_stderr %>% gather(bin,stderr,-population,-gen,-category,factor_key=T)
SFS_relative_stderr_tidy
SFS_relative_stderr_tidy$bin <- gsub('af_', '', gsub('_relative', '', SFS_relative_stderr_tidy$bin))
SFS_relative_stderr_tidy$bin <- as.factor(SFS_relative_stderr_tidy$bin)

SFS_relative_data_tidy <- left_join(SFS_relative_average_tidy,SFS_relative_stderr_tidy,by=c("population","gen","category","bin"))


#Pb:
SFS_relative_data_Pb_ggplot <- ggplot(data=filter(SFS_relative_data_tidy, population=="Pb", gen!=5), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2) +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_col(aes(fill=gen), position="dodge") +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  ggtitle("SFS, Pb") +
  ylab("Proportion") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  #scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
SFS_relative_data_Pb_ggplot
ggsave(paste0("Pb_",type,"_proportion_variants_SFS.pdf"), width=25, height=12, units="cm", device="pdf", path=wd_path)

#Lines:
SFS_relative_data_lines_ggplot <- ggplot(data=filter(SFS_relative_data_tidy, population=="lines"), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2) +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_bar(aes(fill=gen), position=position_dodge(), stat="identity") +
  geom_errorbar(aes(x=bin,ymin=average-stderr, ymax=average+stderr, group=gen), width=0.5, position=position_dodge(0.9)) +
  ggtitle("SFS, line average") +
  ylab("Proportion") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  #scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
SFS_relative_data_lines_ggplot
ggsave(paste0("lines_average_",type,"_proportion_variants_SFS.pdf"), width=15, height=12, units="cm", device="pdf", path=wd_path)

```

####Gen0-140 pools:
#####All sites, raw counts:
```{r}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
SFS_raw_data_combined <- read_tsv(paste0(wd_path,"SFS_pool_gen0-140_combined_",type,"_summary.txt"))

SFS_raw_data_combined <- SFS_raw_data_combined %>% rowwise %>% mutate(total= sum(c_across(where(is.numeric)))) #row-wise sum of all numeric columns (the "af" columns; "gen" is excluded if its converted to a factor first.
codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
SFS_raw_data_combined <- left_join(SFS_raw_data_combined,codes_dictionary,by=c("sample"="old"))
SFS_raw_data_combined$population = factor(SFS_raw_data_combined$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
SFS_raw_data_combined$generation <- as.numeric(gsub("gen","",SFS_raw_data_combined$generation))
SFS_raw_data_combined <- SFS_raw_data_combined %>% arrange(generation)
SFS_raw_data_combined$generation = factor(SFS_raw_data_combined$generation)
SFS_raw_data_combined$category = factor(SFS_raw_data_combined$category,levels=c("fourfold","tolerated","deleterious","LoF"))
SFS_raw_data_combined <- SFS_raw_data_combined %>% arrange(population,generation) %>% select(population,generation,category,contains("af_"))

SFS_raw_Pb_average <- SFS_raw_data_combined %>% filter(population=="Pb")
SFS_raw_lines_average <- SFS_raw_data_combined %>% filter(population!="Pb") %>% group_by(generation,category) %>% summarise_at(vars(contains("af_")), mean) %>% mutate(population="lines") %>% select(population,generation,category,contains("af_")) %>% ungroup()
SFS_raw_data_average <- rbind(SFS_raw_Pb_average,SFS_raw_lines_average)

SFS_raw_average_tidy <- SFS_raw_data_average %>% gather(bin,average,-population,-generation,-category,factor_key=T)
SFS_raw_average_tidy
SFS_raw_average_tidy$bin <- gsub('af_', '', SFS_raw_average_tidy$bin)
SFS_raw_average_tidy$bin <- as.factor(SFS_raw_average_tidy$bin)

se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function
SFS_raw_Pb_stderr <- SFS_raw_data_combined %>% filter(population=="Pb") %>% group_by(generation,category) %>% summarise_at(vars(contains("af_")), se) %>% mutate(population="Pb") %>% select(population,generation,category,contains("af_")) %>% ungroup()
SFS_raw_lines_stderr <- SFS_raw_data_combined %>% filter(population!="Pb") %>% group_by(generation,category) %>% summarise_at(vars(contains("af_")), se) %>% mutate(population="lines") %>% select(population,generation,category,contains("af_")) %>% ungroup()
SFS_raw_data_stderr <- rbind(SFS_raw_Pb_stderr,SFS_raw_lines_stderr)

SFS_raw_stderr_tidy <- SFS_raw_data_stderr %>% gather(bin,stderr,-population,-generation,-category,factor_key=T)
SFS_raw_stderr_tidy
SFS_raw_stderr_tidy$bin <- gsub('af_', '', SFS_raw_stderr_tidy$bin)
SFS_raw_stderr_tidy$bin <- as.factor(SFS_raw_stderr_tidy$bin)

SFS_raw_data_tidy <- left_join(SFS_raw_average_tidy,SFS_raw_stderr_tidy,by=c("population","generation","category","bin"))


#Pb:
SFS_raw_data_Pb_ggplot <- ggplot(data=filter(SFS_raw_data_tidy, population=="Pb", generation!=5), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2,scales="free_y") +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_col(aes(fill=generation), position="dodge") +
  #geom_errorbar(aes(ymin=avg_Pb_raw_value-se_Pb_raw_value, ymax=avg_Pb_raw_value+se_Pb_raw_value), position=position_dodge(), width=0.5) +
  ggtitle("Base population SFS") +
  ylab("Number of sites") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  #scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
SFS_raw_data_Pb_ggplot
ggsave(paste0("Pb_",type,"_raw_SFS.pdf"), width=40, height=12, units="cm", device="pdf", path=wd_path)

#Lines:
SFS_raw_data_lines_ggplot <- ggplot(data=filter(SFS_raw_data_tidy, population=="lines"), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2,scales="free_y") +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_bar(aes(fill=generation), position=position_dodge(), stat="identity") +
  geom_errorbar(aes(x=bin,ymin=average-stderr, ymax=average+stderr, group=generation), width=0.5, position=position_dodge(0.9)) +
  ggtitle("Lines (average) SFS") +
  ylab("Number of sites") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  scale_fill_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
SFS_raw_data_lines_ggplot
ggsave(paste0("lines_average_",type,"_raw_SFS.pdf"), width=30, height=12, units="cm", device="pdf", path=wd_path)

```

#####Only variable sites, raw counts:
```{r}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
SFS_raw_data_combined <- read_tsv(paste0(wd_path,"SFS_pool_gen0-140_combined_",type,"_summary.txt")) %>% select(-af_000,-af_100)

SFS_raw_data_combined <- SFS_raw_data_combined %>% rowwise %>% mutate(total= sum(c_across(where(is.numeric)))) #row-wise sum of all numeric columns (the "af" columns; "gen" is excluded if its converted to a factor first.
codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
SFS_raw_data_combined <- left_join(SFS_raw_data_combined,codes_dictionary,by=c("sample"="old"))
SFS_raw_data_combined$population = factor(SFS_raw_data_combined$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
SFS_raw_data_combined$generation <- as.numeric(gsub("gen","",SFS_raw_data_combined$generation))
SFS_raw_data_combined <- SFS_raw_data_combined %>% arrange(generation)
SFS_raw_data_combined$generation = factor(SFS_raw_data_combined$generation)
SFS_raw_data_combined$category = factor(SFS_raw_data_combined$category,levels=c("fourfold","tolerated","deleterious","LoF"))
SFS_raw_data_combined <- SFS_raw_data_combined %>% arrange(population,generation) %>% select(population,generation,category,contains("af_"))

SFS_raw_Pb_average <- SFS_raw_data_combined %>% filter(population=="Pb")
SFS_raw_lines_average <- SFS_raw_data_combined %>% filter(population!="Pb") %>% group_by(generation,category) %>% summarise_at(vars(contains("af_")), mean) %>% mutate(population="lines") %>% select(population,generation,category,contains("af_")) %>% ungroup()
SFS_raw_data_average <- rbind(SFS_raw_Pb_average,SFS_raw_lines_average)

SFS_raw_average_tidy <- SFS_raw_data_average %>% gather(bin,average,-population,-generation,-category,factor_key=T)
SFS_raw_average_tidy
SFS_raw_average_tidy$bin <- gsub('af_', '', SFS_raw_average_tidy$bin)
SFS_raw_average_tidy$bin <- as.factor(SFS_raw_average_tidy$bin)

se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function
SFS_raw_Pb_stderr <- SFS_raw_data_combined %>% filter(population=="Pb") %>% group_by(generation,category) %>% summarise_at(vars(contains("af_")), se) %>% mutate(population="Pb") %>% select(population,generation,category,contains("af_")) %>% ungroup()
SFS_raw_lines_stderr <- SFS_raw_data_combined %>% filter(population!="Pb") %>% group_by(generation,category) %>% summarise_at(vars(contains("af_")), se) %>% mutate(population="lines") %>% select(population,generation,category,contains("af_")) %>% ungroup()
SFS_raw_data_stderr <- rbind(SFS_raw_Pb_stderr,SFS_raw_lines_stderr)

SFS_raw_stderr_tidy <- SFS_raw_data_stderr %>% gather(bin,stderr,-population,-generation,-category,factor_key=T)
SFS_raw_stderr_tidy
SFS_raw_stderr_tidy$bin <- gsub('af_', '', SFS_raw_stderr_tidy$bin)
SFS_raw_stderr_tidy$bin <- as.factor(SFS_raw_stderr_tidy$bin)

SFS_raw_data_tidy <- left_join(SFS_raw_average_tidy,SFS_raw_stderr_tidy,by=c("population","generation","category","bin"))


#Pb:
SFS_raw_data_Pb_ggplot <- ggplot(data=filter(SFS_raw_data_tidy, population=="Pb", generation!=5), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2,scales="free_y") +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_col(aes(fill=generation), position="dodge") +
  #geom_errorbar(aes(ymin=avg_Pb_raw_value-se_Pb_raw_value, ymax=avg_Pb_raw_value+se_Pb_raw_value), position=position_dodge(), width=0.5) +
  ggtitle("Base population SFS") +
  ylab("Number of sites") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  #scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
SFS_raw_data_Pb_ggplot
ggsave(paste0("Pb_",type,"_raw_variants_SFS.pdf"), width=40, height=12, units="cm", device="pdf", path=wd_path)

#Lines:
SFS_raw_data_lines_ggplot <- ggplot(data=filter(SFS_raw_data_tidy, population=="lines"), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2,scales="free_y") +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_bar(aes(fill=generation), position=position_dodge(), stat="identity") +
  geom_errorbar(aes(x=bin,ymin=average-stderr, ymax=average+stderr, group=generation), width=0.5, position=position_dodge(0.9)) +
  ggtitle("Lines (average) SFS") +
  ylab("Number of sites") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  scale_fill_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,1,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_blank()
  )
SFS_raw_data_lines_ggplot
ggsave(paste0("lines_average_",type,"_raw_variants_SFS.pdf"), width=30, height=12, units="cm", device="pdf", path=wd_path)

```

#####All sites, proportions:
```{r}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
SFS_raw_data_combined <- read_tsv(paste0(wd_path,"SFS_pool_gen0-140_combined_",type,"_summary.txt"))

SFS_raw_data_combined <- SFS_raw_data_combined %>% rowwise %>% mutate(total= sum(c_across(where(is.numeric)))) #row-wise sum of all numeric columns (the "af" columns; "gen" is excluded if its converted to a factor first.
codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
SFS_raw_data_combined <- left_join(SFS_raw_data_combined,codes_dictionary,by=c("sample"="old"))
SFS_raw_data_combined$population = factor(SFS_raw_data_combined$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
SFS_raw_data_combined$generation <- as.numeric(gsub("gen","",SFS_raw_data_combined$generation))
SFS_raw_data_combined <- SFS_raw_data_combined %>% arrange(generation)
SFS_raw_data_combined$generation = factor(SFS_raw_data_combined$generation)
SFS_raw_data_combined$category = factor(SFS_raw_data_combined$category,levels=c("fourfold","tolerated","deleterious","LoF"))
SFS_raw_data_combined <- SFS_raw_data_combined %>% arrange(population,generation) %>% select(population,generation,category,contains("af_"),total)

SFS_relative_data_combined <- SFS_raw_data_combined %>% mutate_at(vars(contains("af_")),list(relative=~./total)) %>% select(population,generation,category,contains("_relative"))

SFS_relative_Pb_average <- SFS_relative_data_combined %>% filter(population=="Pb")
SFS_relative_lines_average <- SFS_relative_data_combined %>% filter(population!="Pb") %>% group_by(generation,category) %>% summarise_at(vars(contains("_relative")), mean) %>% mutate(population="lines") %>% select(population,generation,category,contains("_relative")) %>% ungroup()
SFS_relative_data_average <- rbind(SFS_relative_Pb_average,SFS_relative_lines_average)

SFS_relative_average_tidy <- SFS_relative_data_average %>% gather(bin,average,-population,-generation,-category,factor_key=T)
SFS_relative_average_tidy
SFS_relative_average_tidy$bin <- gsub('af_', '', gsub('_relative', '', SFS_relative_average_tidy$bin))
SFS_relative_average_tidy$bin <- as.factor(SFS_relative_average_tidy$bin)

se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function
SFS_relative_Pb_stderr <- SFS_relative_data_combined %>% filter(population=="Pb") %>% group_by(generation,category) %>% summarise_at(vars(contains("_relative")), se) %>% mutate(population="Pb") %>% select(population,generation,category,contains("_relative")) %>% ungroup()
SFS_relative_lines_stderr <- SFS_relative_data_combined %>% filter(population!="Pb") %>% group_by(generation,category) %>% summarise_at(vars(contains("_relative")), se) %>% mutate(population="lines") %>% select(population,generation,category,contains("_relative")) %>% ungroup()
SFS_relative_data_stderr <- rbind(SFS_relative_Pb_stderr,SFS_relative_lines_stderr)

SFS_relative_stderr_tidy <- SFS_relative_data_stderr %>% gather(bin,stderr,-population,-generation,-category,factor_key=T)
SFS_relative_stderr_tidy
SFS_relative_stderr_tidy$bin <- gsub('af_', '', gsub('_relative', '', SFS_relative_stderr_tidy$bin))
SFS_relative_stderr_tidy$bin <- as.factor(SFS_relative_stderr_tidy$bin)

SFS_relative_data_tidy <- left_join(SFS_relative_average_tidy,SFS_relative_stderr_tidy,by=c("population","generation","category","bin"))


#Pb:
SFS_relative_data_Pb_ggplot <- ggplot(data=filter(SFS_relative_data_tidy, population=="Pb", generation!=5), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2) +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_col(aes(fill=generation), position="dodge") +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  ggtitle("Base population SFS") +
  ylab("Proportion of sites") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  #scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  labs(fill = "Generation") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_text()
  )
SFS_relative_data_Pb_ggplot
ggsave(paste0("Pb_",type,"_proportion_SFS.pdf"), width=39, height=12, units="cm", device="pdf", path=wd_path)

#Lines:
SFS_relative_data_lines_ggplot <- ggplot(data=filter(SFS_relative_data_tidy, population=="lines"), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2) +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_bar(aes(fill=generation), position=position_dodge(), stat="identity") +
  geom_errorbar(aes(x=bin,ymin=average-stderr, ymax=average+stderr, group=generation), width=0.5, position=position_dodge(0.9)) +
  ggtitle("Lines (average) SFS") +
  ylab("Proportion of sites") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  scale_fill_manual(values=c(hue_pal()(5)[4:5])) +
  labs(fill = "Generation") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_text()
  )
SFS_relative_data_lines_ggplot
ggsave(paste0("lines_average_",type,"_proportion_SFS.pdf"), width=29, height=12, units="cm", device="pdf", path=wd_path)

```

#####Only variable sites, proportions:
```{r}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)

type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
SFS_raw_data_combined <- read_tsv(paste0(wd_path,"SFS_pool_gen0-140_combined_",type,"_summary.txt")) %>% select(-af_000,-af_100)

SFS_raw_data_combined <- SFS_raw_data_combined %>% rowwise %>% mutate(total= sum(c_across(where(is.numeric)))) #row-wise sum of all numeric columns (the "af" columns; "gen" is excluded if its converted to a factor first.
codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
SFS_raw_data_combined <- left_join(SFS_raw_data_combined,codes_dictionary,by=c("sample"="old"))
SFS_raw_data_combined$population = factor(SFS_raw_data_combined$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
SFS_raw_data_combined$generation <- as.numeric(gsub("gen","",SFS_raw_data_combined$generation))
SFS_raw_data_combined <- SFS_raw_data_combined %>% arrange(generation)
SFS_raw_data_combined$generation = factor(SFS_raw_data_combined$generation)
SFS_raw_data_combined$category = factor(SFS_raw_data_combined$category,levels=c("fourfold","tolerated","deleterious","LoF"))
SFS_raw_data_combined <- SFS_raw_data_combined %>% arrange(population,generation) %>% select(population,generation,category,contains("af_"),total)

SFS_relative_data_combined <- SFS_raw_data_combined %>% mutate_at(vars(contains("af_")),list(relative=~./total)) %>% select(population,generation,category,contains("_relative"))

SFS_relative_Pb_average <- SFS_relative_data_combined %>% filter(population=="Pb")
SFS_relative_lines_average <- SFS_relative_data_combined %>% filter(population!="Pb") %>% group_by(generation,category) %>% summarise_at(vars(contains("_relative")), mean) %>% mutate(population="lines") %>% select(population,generation,category,contains("_relative")) %>% ungroup()
SFS_relative_data_average <- rbind(SFS_relative_Pb_average,SFS_relative_lines_average)

SFS_relative_average_tidy <- SFS_relative_data_average %>% gather(bin,average,-population,-generation,-category,factor_key=T)
SFS_relative_average_tidy
SFS_relative_average_tidy$bin <- gsub('af_', '', gsub('_relative', '', SFS_relative_average_tidy$bin))
SFS_relative_average_tidy$bin <- as.factor(SFS_relative_average_tidy$bin)

se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function
SFS_relative_Pb_stderr <- SFS_relative_data_combined %>% filter(population=="Pb") %>% group_by(generation,category) %>% summarise_at(vars(contains("_relative")), se) %>% mutate(population="Pb") %>% select(population,generation,category,contains("_relative")) %>% ungroup()
SFS_relative_lines_stderr <- SFS_relative_data_combined %>% filter(population!="Pb") %>% group_by(generation,category) %>% summarise_at(vars(contains("_relative")), se) %>% mutate(population="lines") %>% select(population,generation,category,contains("_relative")) %>% ungroup()
SFS_relative_data_stderr <- rbind(SFS_relative_Pb_stderr,SFS_relative_lines_stderr)

SFS_relative_stderr_tidy <- SFS_relative_data_stderr %>% gather(bin,stderr,-population,-generation,-category,factor_key=T)
SFS_relative_stderr_tidy
SFS_relative_stderr_tidy$bin <- gsub('af_', '', gsub('_relative', '', SFS_relative_stderr_tidy$bin))
SFS_relative_stderr_tidy$bin <- as.factor(SFS_relative_stderr_tidy$bin)

SFS_relative_data_tidy <- left_join(SFS_relative_average_tidy,SFS_relative_stderr_tidy,by=c("population","generation","category","bin"))


#Pb:
SFS_relative_data_Pb_ggplot <- ggplot(data=filter(SFS_relative_data_tidy, population=="Pb", generation!=5), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2) +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_col(aes(fill=generation), position="dodge") +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  ggtitle("Base population SFS") +
  ylab("Proportion of sites") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  #scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  scale_y_continuous(breaks=seq(0,0.3,by=0.1)) +
  expand_limits(y = c(0, 0.35)) +
  labs(fill = "Generation") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_text()
  )
SFS_relative_data_Pb_ggplot
ggsave(paste0("Pb_",type,"_proportion_variants_SFS.pdf"), width=39, height=20, units="cm", device="pdf", path=wd_path)

#Lines:
SFS_relative_data_lines_ggplot <- ggplot(data=filter(SFS_relative_data_tidy, population=="lines"), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2) +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_bar(aes(fill=generation), position=position_dodge(), stat="identity") +
  geom_errorbar(aes(x=bin,ymin=average-stderr, ymax=average+stderr, group=generation), width=0.5, position=position_dodge(0.9)) +
  ggtitle("Lines (average) SFS") +
  ylab("Proportion of sites") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  scale_fill_manual(values=c(hue_pal()(5)[4:5])) +
  scale_y_continuous(breaks=seq(0,0.3,by=0.1)) +
  expand_limits(y = c(0, 0.35)) +
  labs(fill = "Generation") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(angle=90,hjust=1,vjust = 0.5,colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_text()
  )
SFS_relative_data_lines_ggplot
ggsave(paste0("lines_average_",type,"_proportion_variants_SFS.pdf"), width=29, height=20, units="cm", device="pdf", path=wd_path)

```

#####Only monomorphic sites, proportions:
```{r}

library(readr)
library(dplyr)
library(ggplot2)
library(tidyr)
library(scales)


type <- "all" #all #autosomes #Xchr

wd_path <- ("/Users/dani/Documents/dani/Post-docs/2022_contrato_Vigo/drosophila_load/Analisis_dani/counts/gen0-140/")
SFS_raw_data_combined <- read_tsv(paste0(wd_path,"SFS_pool_gen0-140_combined_",type,"_summary.txt"))

SFS_raw_data_combined <- SFS_raw_data_combined %>% rowwise %>% mutate(total= sum(c_across(where(is.numeric)))) #row-wise sum of all numeric columns (the "af" columns; "gen" is excluded if its converted to a factor first.
codes_dictionary <- read_tsv(paste0(wd_path,"pop_codes_dictionary.txt"),col_names=c("old","population"))
SFS_raw_data_combined <- left_join(SFS_raw_data_combined,codes_dictionary,by=c("sample"="old"))
SFS_raw_data_combined$population = factor(SFS_raw_data_combined$population,levels=c("Pb","L01","L16","L27","L29","L31","L33","L34","L36","L38","L39","L42","L45","L48","L51","L52","L54","L59","L61","L63"))
SFS_raw_data_combined$generation <- as.numeric(gsub("gen","",SFS_raw_data_combined$generation))
SFS_raw_data_combined <- SFS_raw_data_combined %>% arrange(generation)
SFS_raw_data_combined$generation = factor(SFS_raw_data_combined$generation)
SFS_raw_data_combined$category = factor(SFS_raw_data_combined$category,levels=c("fourfold","tolerated","deleterious","LoF"))
SFS_raw_data_combined <- SFS_raw_data_combined %>% arrange(population,generation) %>% select(population,generation,category,contains("af_"),total)

SFS_relative_data_combined <- SFS_raw_data_combined %>% mutate_at(vars(contains("af_")),list(relative=~./total)) %>% select(population,generation,category,contains("_relative"))

SFS_relative_Pb_average <- SFS_relative_data_combined %>% filter(population=="Pb")
SFS_relative_lines_average <- SFS_relative_data_combined %>% filter(population!="Pb") %>% group_by(generation,category) %>% summarise_at(vars(contains("_relative")), mean) %>% mutate(population="lines") %>% select(population,generation,category,contains("_relative")) %>% ungroup()
SFS_relative_data_average <- rbind(SFS_relative_Pb_average,SFS_relative_lines_average)

SFS_relative_average_tidy <- SFS_relative_data_average %>% gather(bin,average,-population,-generation,-category,factor_key=T)
SFS_relative_average_tidy
SFS_relative_average_tidy$bin <- gsub('af_', '', gsub('_relative', '', SFS_relative_average_tidy$bin))
SFS_relative_average_tidy$bin <- as.factor(SFS_relative_average_tidy$bin)

se <- function(x) sqrt(var(x)/length(x)) #first define the standard error function
SFS_relative_Pb_stderr <- SFS_relative_data_combined %>% filter(population=="Pb") %>% group_by(generation,category) %>% summarise_at(vars(contains("_relative")), se) %>% mutate(population="Pb") %>% select(population,generation,category,contains("_relative")) %>% ungroup()
SFS_relative_lines_stderr <- SFS_relative_data_combined %>% filter(population!="Pb") %>% group_by(generation,category) %>% summarise_at(vars(contains("_relative")), se) %>% mutate(population="lines") %>% select(population,generation,category,contains("_relative")) %>% ungroup()
SFS_relative_data_stderr <- rbind(SFS_relative_Pb_stderr,SFS_relative_lines_stderr)

SFS_relative_stderr_tidy <- SFS_relative_data_stderr %>% gather(bin,stderr,-population,-generation,-category,factor_key=T)
SFS_relative_stderr_tidy
SFS_relative_stderr_tidy$bin <- gsub('af_', '', gsub('_relative', '', SFS_relative_stderr_tidy$bin))
SFS_relative_stderr_tidy$bin <- as.factor(SFS_relative_stderr_tidy$bin)

SFS_relative_data_tidy <- left_join(SFS_relative_average_tidy,SFS_relative_stderr_tidy,by=c("population","generation","category","bin"))


#Pb:
SFS_relative_data_Pb_ggplot <- ggplot(data=filter(SFS_relative_data_tidy, population=="Pb" & generation!=5 & (bin=="000" | bin=="100")), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2) +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_col(aes(fill=generation), position="dodge") +
  #geom_errorbar(aes(ymin=avg_Pb_relative_value-se_Pb_relative_value, ymax=avg_Pb_relative_value+se_Pb_relative_value), position=position_dodge(), width=0.5) +
  ggtitle("Base population SFS") +
  ylab("Proportion of sites") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  #scale_colour_manual(values=c(hue_pal()(5)[4:5])) +
  scale_x_discrete(labels=c("000" = "0", "100" = "1")) +
  expand_limits(y = c(0, 0.70)) +
  labs(fill = "Generation") +
  theme_bw() +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_text()
  )
SFS_relative_data_Pb_ggplot
ggsave(paste0("Pb_",type,"_proportion_monomorphic_SFS.pdf"), width=12, height=20, units="cm", device="pdf", path=wd_path)

#Lines:
SFS_relative_data_lines_ggplot <- ggplot(data=filter(SFS_relative_data_tidy, population=="lines" & (bin=="000" | bin=="100")), aes(bin,average)) +
  facet_wrap(~category,nrow=2,ncol=2) +
  #facet_grid(. ~ population) +
  #geom_bar(stat="identity", colour="black", position="dodge")
  geom_bar(aes(fill=generation), position=position_dodge(), stat="identity") +
  geom_errorbar(aes(x=bin,ymin=average-stderr, ymax=average+stderr, group=generation), width=0.5, position=position_dodge(0.9)) +
  ggtitle("Lines (average) SFS") +
  ylab("Proportion of sites") +
  xlab("Derived allele frequency") +
  #ggtitle("Pools, lines (average)") +
  #guides(colour="none") +
  scale_x_discrete(labels=c("000" = "0", "100" = "1")) +
  expand_limits(y = c(0, 0.70)) +
  scale_fill_manual(values=c(hue_pal()(5)[4:5])) +
  theme_bw() +
  labs(fill = "Generation") +
  theme(text=element_text(size=12,face="bold"),
        rect=element_rect(size=1),
        axis.line=element_line(colour="black"),
        axis.title=element_text(size=16),
        axis.text=element_text(size=10,colour="black"),
        axis.text.x=element_text(colour="black",size=10),
        #axis.text.y=element_text(size=24,colour="black",margin=margin(t=0.5,unit="cm")),
        #axis.title.y=element_text(size=30,margin=margin(r=0.5,unit="cm")),
        panel.background=element_blank(),
        panel.border=element_rect(colour="black"),
        #panel.grid=element_blank(),
        #panel.grid.major=element_line(colour="grey", linetype="dashed", size=0.4),
        plot.margin=unit(c(0.5,0.5,0.5,0.5),"cm"),
        #plot.title=element_text(size=36, face="bold", margin=margin(b=0.5, unit="cm")),
        legend.background=element_rect(linetype="solid", colour="black", size=.5),
        #legend.justification=c(0,0),
        legend.key=element_rect(colour="white"),
        #legend.key.size=unit(1.3,"cm"),
        #legend.position="none",
        legend.title=element_text()
  )
SFS_relative_data_lines_ggplot
ggsave(paste0("lines_average_",type,"_proportion_monomorphic_SFS.pdf"), width=12, height=20, units="cm", device="pdf", path=wd_path)

```
